{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "# Treponema genome analysis workflow - Grillova et al. 2019"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## The environment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Tested software versions and operation system"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The workflows has been tested on Linux machine (Ubuntu 16.04) with Python 2.7.6, Python 3.4.3 using Conda 4.5.11, Jupyter notebook 5.7.2 and bash_kernel 0.7.1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Linux BioDA-server 4.4.0-148-generic #174-Ubuntu SMP Tue May 7 12:20:14 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux\n",
      "(.env) No LSB modules are available.\n",
      "Distributor ID:\tUbuntu\n",
      "Description:\tUbuntu 16.04.6 LTS\n",
      "Release:\t16.04\n",
      "Codename:\txenial\n",
      "(.env) (.env) /opt/install/dir/anaconda/bin/python3\n",
      "(.env) Python 3.7.1\n",
      "(.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/bin/conda\n",
      "(.env) conda 4.6.14\n",
      "(.env) Jupyter notebook 4.4.1\n",
      "(.env) The program 'pip3' is currently not installed. To run 'pip3' please ask your administrator to install the package 'python3-pip'\n",
      "bash_kernel\n",
      "(.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "uname -a\n",
    "lsb_release -a\n",
    "export PATH=/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/bin:$PATH # In case we use different version of Conda than the system-wide installation\n",
    "which python3\n",
    "python3 --version\n",
    "which conda\n",
    "conda --version\n",
    "echo \"Jupyter notebook\" `jupyter notebook --version`\n",
    "echo \"bash_kernel\" `pip3 show bash_kernel | grep Version`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## The analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "First, we can activate the environment and check "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(.env) (treponema) (.env) (treponema) (.env) \n",
      "     active environment : treponema\n",
      "    active env location : /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema\n",
      "            shell level : 1\n",
      "       user config file : /mnt/nfs/home/325073/000000-My_Documents/VM-home/.condarc\n",
      " populated config files : /mnt/nfs/home/325073/000000-My_Documents/VM-home/.condarc\n",
      "          conda version : 4.6.14\n",
      "    conda-build version : not installed\n",
      "         python version : 2.7.16.final.0\n",
      "       base environment : /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2  (read only)\n",
      "           channel URLs : https://conda.anaconda.org/conda-forge/linux-64\n",
      "                          https://conda.anaconda.org/conda-forge/noarch\n",
      "                          https://repo.anaconda.com/pkgs/main/linux-64\n",
      "                          https://repo.anaconda.com/pkgs/main/noarch\n",
      "                          https://repo.anaconda.com/pkgs/free/linux-64\n",
      "                          https://repo.anaconda.com/pkgs/free/noarch\n",
      "                          https://repo.anaconda.com/pkgs/r/linux-64\n",
      "                          https://repo.anaconda.com/pkgs/r/noarch\n",
      "                          https://conda.anaconda.org/r/linux-64\n",
      "                          https://conda.anaconda.org/r/noarch\n",
      "                          https://conda.anaconda.org/bioconda/linux-64\n",
      "                          https://conda.anaconda.org/bioconda/noarch\n",
      "          package cache : /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/pkgs\n",
      "                          /mnt/nfs/home/325073/000000-My_Documents/VM-home/.conda/pkgs\n",
      "       envs directories : /mnt/nfs/home/325073/000000-My_Documents/VM-home/.conda/envs\n",
      "                          /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs\n",
      "               platform : linux-64\n",
      "             user-agent : conda/4.6.14 requests/2.21.0 CPython/2.7.16 Linux/4.4.0-148-generic ubuntu/16.04.6 glibc/2.23\n",
      "                UID:GID : 100134:100006\n",
      "             netrc file : None\n",
      "           offline mode : False\n",
      "\n",
      "(treponema) (.env) # packages in environment at /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema:\n",
      "#\n",
      "# Name                    Version                   Build  Channel\n",
      "_r-mutex                  1.0.0               anacondar_1  \n",
      "asn1crypto                0.24.0                py27_1003    conda-forge\n",
      "augustus                  3.3             pl526hcfae127_4    bioconda\n",
      "backports                 1.0                        py_2    conda-forge\n",
      "backports.functools_lru_cache 1.5                        py_1    conda-forge\n",
      "backports_abc             0.5                        py_1    conda-forge\n",
      "bamtools                  2.4.1                         1    bioconda\n",
      "bbmap                     38.22                h14c3975_1    bioconda\n",
      "bcftools                  1.4.1                         0    bioconda\n",
      "besst                     2.2.8                      py_2    bioconda\n",
      "bioconductor-biobase      2.42.0           r351h14c3975_1    bioconda\n",
      "bioconductor-biocgenerics 0.28.0                   r351_1    bioconda\n",
      "bioconductor-biocparallel 1.16.6           r351h1c2f66e_0    bioconda\n",
      "bioconductor-biostrings   2.50.2           r351h14c3975_0    bioconda\n",
      "bioconductor-delayedarray 0.8.0            r351h14c3975_0    bioconda\n",
      "bioconductor-genomeinfodb 1.18.1                   r351_0    bioconda\n",
      "bioconductor-genomeinfodbdata 1.2.1                    r351_0    bioconda\n",
      "bioconductor-genomicalignments 1.18.1           r351h14c3975_0    bioconda\n",
      "bioconductor-genomicranges 1.34.0           r351h14c3975_0    bioconda\n",
      "bioconductor-iranges      2.16.0           r351h14c3975_0    bioconda\n",
      "bioconductor-noiseq       2.26.1                   r351_0    bioconda\n",
      "bioconductor-rsamtools    1.34.0           r351hf484d3e_0    bioconda\n",
      "bioconductor-rtracklayer  1.42.1           r351h9d9f1b6_1    bioconda\n",
      "bioconductor-s4vectors    0.20.1           r351h14c3975_0    bioconda\n",
      "bioconductor-summarizedexperiment 1.12.0                   r351_0    bioconda\n",
      "bioconductor-xvector      0.22.0           r351h14c3975_0    bioconda\n",
      "bioconductor-zlibbioc     1.28.0           r351h14c3975_0    bioconda\n",
      "blas                      1.0                         mkl  \n",
      "blast                     2.7.1                h4422958_6    bioconda\n",
      "boost                     1.67.0           py27h3e44d54_0    conda-forge\n",
      "boost-cpp                 1.67.0               h3a22d5f_0    conda-forge\n",
      "busco                     3.0.2                    py27_8    bioconda\n",
      "bwa                       0.7.15                        1    bioconda\n",
      "bwidget                   1.9.11                        1  \n",
      "bz2file                   0.98                       py_0    conda-forge\n",
      "bzip2                     1.0.6             h14c3975_1002    conda-forge\n",
      "ca-certificates           2019.6.16            hecc5488_0    conda-forge\n",
      "cairo                     1.14.12           h80bd089_1005    conda-forge\n",
      "certifi                   2019.3.9                 py27_0    anaconda\n",
      "cffi                      1.12.3           py27h8022711_0    conda-forge\n",
      "chardet                   3.0.4                 py27_1003    conda-forge\n",
      "circos                    0.69.6                        5    bioconda\n",
      "click                     7.0                        py_0    conda-forge\n",
      "colormath                 3.0.0                      py_2    conda-forge\n",
      "cryptography              2.7              py27h72c5cf5_0    conda-forge\n",
      "curl                      7.64.1               hf8cf82a_0    conda-forge\n",
      "cutadapt                  1.18             py27h14c3975_1    bioconda\n",
      "cycler                    0.10.0                     py_1    conda-forge\n",
      "dbus                      1.13.2               h714fa37_1  \n",
      "decorator                 4.4.0                      py_0    conda-forge\n",
      "enum34                    1.1.6                 py27_1001    conda-forge\n",
      "expat                     2.2.5             hf484d3e_1002    conda-forge\n",
      "fastqc                    0.11.8                        1    bioconda\n",
      "font-ttf-dejavu-sans-mono 2.37                 h6964260_0  \n",
      "fontconfig                2.13.1            he4413a7_1000    conda-forge\n",
      "freebayes                 0.9.21.26                     0    bioconda\n",
      "freetype                  2.10.0               he983fc9_0    conda-forge\n",
      "functools32               3.2.3.2                    py_3    conda-forge\n",
      "future                    0.17.1                py27_1000    conda-forge\n",
      "futures                   3.2.0                 py27_1000    conda-forge\n",
      "gatk                      3.7                      py27_1    bioconda\n",
      "gettext                   0.19.8.1          hc5be6a0_1002    conda-forge\n",
      "giflib                    5.1.9                h516909a_0    conda-forge\n",
      "glib                      2.56.2            had28632_1001    conda-forge\n",
      "glimmerhmm                3.0.4                h2d50403_2    bioconda\n",
      "gmp                       6.1.2             hf484d3e_1000    conda-forge\n",
      "gnutls                    3.5.19               h2a4e5f8_1    conda-forge\n",
      "graphite2                 1.3.13            hf484d3e_1000    conda-forge\n",
      "gsl                       2.2.1                h0c605f7_3  \n",
      "gst-plugins-base          1.14.0               hbbd80ab_1  \n",
      "gstreamer                 1.14.0               hb453b48_1  \n",
      "harfbuzz                  1.9.0             he243708_1001    conda-forge\n",
      "hmmer                     3.2.1                hf484d3e_1    bioconda\n",
      "htslib                    1.9                  h4da6232_3    bioconda\n",
      "icu                       58.2              hf484d3e_1000    conda-forge\n",
      "idna                      2.8                   py27_1000    conda-forge\n",
      "intel-openmp              2019.4                      243  \n",
      "ipaddress                 1.0.22                     py_1    conda-forge\n",
      "jinja2                    2.10.1                     py_0    conda-forge\n",
      "joblib                    0.13.2                     py_0    conda-forge\n",
      "jpeg                      9c                h14c3975_1001    conda-forge\n",
      "kiwisolver                1.1.0            py27hc9558a2_0    conda-forge\n",
      "kraken2                   2.0.8_beta      pl526h6bb024c_0    bioconda\n",
      "krb5                      1.16.3            h05b26f9_1001    conda-forge\n",
      "libcurl                   7.64.1               hda55be3_0    conda-forge\n",
      "libdeflate                1.2                  h14c3975_0    bioconda\n",
      "libedit                   3.1.20170329      hf8c457e_1001    conda-forge\n",
      "libffi                    3.2.1             he1b5a44_1006    conda-forge\n",
      "libgcc                    7.2.0                h69d50b8_2    conda-forge\n",
      "libgcc-ng                 9.1.0                hdf63c60_0  \n",
      "libgd                     2.2.5             h0d07dcb_1005    conda-forge\n",
      "libgfortran               3.0.0                         1    conda-forge\n",
      "libgfortran-ng            7.3.0                hdf63c60_0  \n",
      "libiconv                  1.15              h516909a_1005    conda-forge\n",
      "libidn2                   2.1.1                h14c3975_0    conda-forge\n",
      "libpng                    1.6.37               hed695b0_0    conda-forge\n",
      "libssh2                   1.8.2                h22169c7_2    conda-forge\n",
      "libstdcxx-ng              9.1.0                hdf63c60_0  \n",
      "libtiff                   4.0.10            h57b8799_1003    conda-forge\n",
      "libunistring              0.9.10               h14c3975_0    conda-forge\n",
      "libuuid                   2.32.1            h14c3975_1000    conda-forge\n",
      "libwebp                   1.0.2                h576950b_1    conda-forge\n",
      "libxcb                    1.13              h14c3975_1002    conda-forge\n",
      "libxml2                   2.9.9                h13577e0_0    conda-forge\n",
      "lp_solve                  5.5.2.5           h14c3975_1001    conda-forge\n",
      "lz4-c                     1.8.3             he1b5a44_1001    conda-forge\n",
      "lzstring                  1.0.4                   py_1001    conda-forge\n",
      "make                      4.2.1             h14c3975_2004    conda-forge\n",
      "markdown                  2.6.11                     py_0    conda-forge\n",
      "markupsafe                1.1.1            py27h14c3975_0    conda-forge\n",
      "mathstats                 0.2.6.5                    py_0    bioconda\n",
      "matplotlib                2.2.3            py27h8a2030e_1    conda-forge\n",
      "matplotlib-base           2.2.3            py27h60b886d_1    conda-forge\n",
      "metis                     5.1.0             hf484d3e_1003    conda-forge\n",
      "mkl                       2019.4                      243  \n",
      "mkl_fft                   1.0.13           py27h516909a_1    conda-forge\n",
      "mkl_random                1.0.4            py27hf2d7682_0    conda-forge\n",
      "multiqc                   1.7                        py_4    bioconda\n",
      "ncurses                   6.1               hf484d3e_1002    conda-forge\n",
      "nettle                    3.3                           0    conda-forge\n",
      "networkx                  2.2                        py_1    conda-forge\n",
      "numpy                     1.16.4           py27h7e9f1db_0  \n",
      "numpy-base                1.16.4           py27hde5b4d6_0  \n",
      "openblas                  0.2.20                        8    conda-forge\n",
      "openjdk                   8.0.192           h14c3975_1003    conda-forge\n",
      "openssl                   1.1.1                h7b6447c_0    anaconda\n",
      "pango                     1.40.14           hf0c64fd_1003    conda-forge\n",
      "pcre                      8.43                 he6710b0_0  \n",
      "perl                      5.26.2            h516909a_1006    conda-forge\n",
      "perl-app-cpanminus        1.7044                  pl526_1    bioconda\n",
      "perl-archive-tar          2.32                    pl526_0    bioconda\n",
      "perl-autoloader           5.74                    pl526_2    bioconda\n",
      "perl-carp                 1.38                    pl526_3    bioconda\n",
      "perl-clone                0.41            pl526h14c3975_1    bioconda\n",
      "perl-compress-raw-bzip2   2.086           pl526hf484d3e_0    bioconda\n",
      "perl-compress-raw-zlib    2.086           pl526h6bb024c_1    bioconda\n",
      "perl-config-general       2.63                    pl526_0    bioconda\n",
      "perl-constant             1.33                    pl526_1    bioconda\n",
      "perl-cpan-meta            2.150010                pl526_0    bioconda\n",
      "perl-cpan-meta-requirements 2.140                   pl526_0    bioconda\n",
      "perl-cpan-meta-yaml       0.018                   pl526_0    bioconda\n",
      "perl-data-dumper          2.173                   pl526_0    bioconda\n",
      "perl-dbi                  1.642                   pl526_0    bioconda\n",
      "perl-digest-perl-md5      1.9                     pl526_1    bioconda\n",
      "perl-dynaloader           1.25                    pl526_1    bioconda\n",
      "perl-encode               2.88                    pl526_1    bioconda\n",
      "perl-exporter             5.72                    pl526_1    bioconda\n",
      "perl-exporter-tiny        1.002001                pl526_0    bioconda\n",
      "perl-extutils-cbuilder    0.280230                pl526_1    bioconda\n",
      "perl-extutils-makemaker   7.36                    pl526_1    bioconda\n",
      "perl-extutils-manifest    1.72                    pl526_0    bioconda\n",
      "perl-extutils-parsexs     3.35                    pl526_0    bioconda\n",
      "perl-file-path            2.16                    pl526_0    bioconda\n",
      "perl-file-temp            0.2304                  pl526_2    bioconda\n",
      "perl-font-ttf             1.06                    pl526_0    bioconda\n",
      "perl-gd                   2.71            pl526he860b03_0    bioconda\n",
      "perl-getopt-long          2.50                    pl526_1    bioconda\n",
      "perl-io-compress          2.086           pl526hf484d3e_0    bioconda\n",
      "perl-io-string            1.08                    pl526_3    bioconda\n",
      "perl-io-zlib              1.10                    pl526_2    bioconda\n",
      "perl-ipc-cmd              1.02                    pl526_0    bioconda\n",
      "perl-json-pp              4.02                    pl526_0    bioconda\n",
      "perl-list-moreutils       0.428                   pl526_1    bioconda\n",
      "perl-list-moreutils-xs    0.428                   pl526_0    bioconda\n",
      "perl-locale-maketext-simple 0.21                    pl526_2    bioconda\n",
      "perl-math-bezier          0.01                    pl526_1    bioconda\n",
      "perl-math-round           0.07                    pl526_1    bioconda\n",
      "perl-math-vecstat         0.08                    pl526_1    bioconda\n",
      "perl-module-build         0.4224                  pl526_3    bioconda\n",
      "perl-module-corelist      5.20190524              pl526_0    bioconda\n",
      "perl-module-implementation 0.09                    pl526_2    bioconda\n",
      "perl-module-load          0.32                    pl526_1    bioconda\n",
      "perl-module-load-conditional 0.68                    pl526_2    bioconda\n",
      "perl-module-metadata      1.000036                pl526_0    bioconda\n",
      "perl-module-runtime       0.016                   pl526_1    bioconda\n",
      "perl-number-format        1.75                    pl526_3    bioconda\n",
      "perl-params-check         0.38                    pl526_1    bioconda\n",
      "perl-params-validate      1.29            pl526h14c3975_1    bioconda\n",
      "perl-parent               0.236                   pl526_1    bioconda\n",
      "perl-pathtools            3.75            pl526h14c3975_1    bioconda\n",
      "perl-perl-ostype          1.010                   pl526_1    bioconda\n",
      "perl-readonly             2.05                    pl526_0    bioconda\n",
      "perl-regexp-common        2017060201              pl526_0    bioconda\n",
      "perl-scalar-list-utils    1.50            pl526h14c3975_0    bioconda\n",
      "perl-set-intspan          1.19                    pl526_1    bioconda\n",
      "perl-statistics-basic     1.6611                  pl526_2    bioconda\n",
      "perl-svg                  2.84                    pl526_0    bioconda\n",
      "perl-text-abbrev          1.02                    pl526_0    bioconda\n",
      "perl-text-format          0.59                    pl526_2    bioconda\n",
      "perl-text-parsewords      3.30                    pl526_0    bioconda\n",
      "perl-time-hires           1.9760          pl526h14c3975_1    bioconda\n",
      "perl-try-tiny             0.30                    pl526_1    bioconda\n",
      "perl-version              0.9924                  pl526_0    bioconda\n",
      "perl-xml-parser           2.44            pl526h4e0c4b3_7    bioconda\n",
      "perl-xsloader             0.24                    pl526_0    bioconda\n",
      "perl-yaml                 1.29                    pl526_0    bioconda\n",
      "picard                    2.9.2                         2    bioconda\n",
      "pigz                      2.3.4                         0    conda-forge\n",
      "pip                       19.1.1                   py27_0    conda-forge\n",
      "pixman                    0.34.0            h14c3975_1003    conda-forge\n",
      "popt                      1.16                          1    bioconda\n",
      "pthread-stubs             0.4               h14c3975_1001    conda-forge\n",
      "pycparser                 2.19                     py27_1    conda-forge\n",
      "pyopenssl                 19.0.0                   py27_0    conda-forge\n",
      "pyparsing                 2.4.0                      py_0    conda-forge\n",
      "pyqt                      5.6.0           py27h13b7fb3_1008    conda-forge\n",
      "pysam                     0.11.2.2                 py27_1    bioconda\n",
      "pysocks                   1.7.0                    py27_0    conda-forge\n",
      "python                    2.7.15            h721da81_1008    conda-forge\n",
      "python-dateutil           2.8.0                      py_0    conda-forge\n",
      "pytz                      2019.1                     py_0    conda-forge\n",
      "pyyaml                    5.1.1            py27h516909a_0    conda-forge\n",
      "qt                        5.6.3                h8bf5577_3  \n",
      "qualimap                  2.2.2b                        1    bioconda\n",
      "quast                     5.0.2           py27pl526ha92aebf_0    bioconda\n",
      "r                         3.5.1                 r351_1000    conda-forge\n",
      "r-assertthat              0.2.1            r351h6115d3f_0    conda-forge\n",
      "r-backports               1.1.4            r351hcdcec82_0    conda-forge\n",
      "r-base                    3.5.1                h391c2eb_5    conda-forge\n",
      "r-bh                      1.69.0_1         r351h6115d3f_0    conda-forge\n",
      "r-bitops                  1.0_6           r351h96ca727_1002    conda-forge\n",
      "r-boot                    1.3_22                   r351_0    conda-forge\n",
      "r-class                   7.3_15          r351h96ca727_1000    conda-forge\n",
      "r-cli                     1.1.0            r351h6115d3f_0    conda-forge\n",
      "r-cluster                 2.0.9            r351h9bbef5b_0    conda-forge\n",
      "r-codetools               0.2_16          r351h6115d3f_1000    conda-forge\n",
      "r-colorspace              1.4_1            r351hcdcec82_0    conda-forge\n",
      "r-crayon                  1.3.4           r351h6115d3f_1001    conda-forge\n",
      "r-digest                  0.6.19           r351h0357c0b_0    conda-forge\n",
      "r-fansi                   0.4.0           r351h96ca727_1000    conda-forge\n",
      "r-foreign                 0.8_71          r351h96ca727_1002    conda-forge\n",
      "r-formatr                 1.7               r35h6115d3f_0    conda-forge\n",
      "r-futile.logger           1.4.3           r351h6115d3f_1001    conda-forge\n",
      "r-futile.options          1.0.1           r351h6115d3f_1000    conda-forge\n",
      "r-getopt                  1.20.3                   r351_0    conda-forge\n",
      "r-ggplot2                 3.2.0             r35h6115d3f_0    conda-forge\n",
      "r-glue                    1.3.1            r351hcdcec82_0    conda-forge\n",
      "r-gtable                  0.3.0            r351h6115d3f_0    conda-forge\n",
      "r-kernsmooth              2.23_15         r351ha65eedd_1002    conda-forge\n",
      "r-labeling                0.3             r351h6115d3f_1001    conda-forge\n",
      "r-lambda.r                1.2.3           r351h6115d3f_1000    conda-forge\n",
      "r-lattice                 0.20_38         r351h96ca727_1000    conda-forge\n",
      "r-lazyeval                0.2.2            r351hcdcec82_0    conda-forge\n",
      "r-magrittr                1.5             r351h6115d3f_1001    conda-forge\n",
      "r-mass                    7.3_51.4         r351hcdcec82_0    conda-forge\n",
      "r-matrix                  1.2_17           r351hcdcec82_0    conda-forge\n",
      "r-matrixstats             0.54.0          r351h96ca727_1000    conda-forge\n",
      "r-mgcv                    1.8_28           r351hcdcec82_0    conda-forge\n",
      "r-munsell                 0.5.0           r351h6115d3f_1001    conda-forge\n",
      "r-nlme                    3.1_140          r351h9bbef5b_0    conda-forge\n",
      "r-nnet                    7.3_12          r351h96ca727_1002    conda-forge\n",
      "r-optparse                1.6.1            r351h6115d3f_0    conda-forge\n",
      "r-pillar                  1.4.1                h6115d3f_0    conda-forge\n",
      "r-pkgconfig               2.0.2           r351h6115d3f_1001    conda-forge\n",
      "r-plyr                    1.8.4           r351h29659fb_1002    conda-forge\n",
      "r-r6                      2.4.0            r351h6115d3f_0    conda-forge\n",
      "r-rcolorbrewer            1.1_2           r351h6115d3f_1001    conda-forge\n",
      "r-rcpp                    1.0.1            r351h0357c0b_0    conda-forge\n",
      "r-rcurl                   1.95_4.12        r351hcdcec82_0    conda-forge\n",
      "r-recommended             3.5.1                 r351_1001    conda-forge\n",
      "r-reshape2                1.4.3           r351h29659fb_1003    conda-forge\n",
      "r-rlang                   0.3.4            r351hcdcec82_0    conda-forge\n",
      "r-rpart                   4.1_15           r351hcdcec82_0    conda-forge\n",
      "r-scales                  1.0.0           r351h29659fb_1001    conda-forge\n",
      "r-snow                    0.4_3           r351h6115d3f_1000    conda-forge\n",
      "r-spatial                 7.3_11          r351h96ca727_1002    conda-forge\n",
      "r-stringi                 1.4.3            r351h0357c0b_0    conda-forge\n",
      "r-stringr                 1.4.0            r351h6115d3f_0    conda-forge\n",
      "r-survival                2.44_1.1         r351hcdcec82_0    conda-forge\n",
      "r-tibble                  2.1.3             r35hcdcec82_0    conda-forge\n",
      "r-utf8                    1.1.4           r351h96ca727_1000    conda-forge\n",
      "r-vctrs                   0.1.0           r351h96ca727_1000    conda-forge\n",
      "r-viridislite             0.3.0           r351h6115d3f_1001    conda-forge\n",
      "r-withr                   2.1.2           r351h6115d3f_1000    conda-forge\n",
      "r-xml                     3.98_1.20         r35hcdcec82_0    conda-forge\n",
      "r-zeallot                 0.1.0           r351h6115d3f_1000    conda-forge\n",
      "readline                  7.0               hf8c457e_1001    conda-forge\n",
      "reaper                    16.098               ha92aebf_2    bioconda\n",
      "requests                  2.22.0                   py27_0    conda-forge\n",
      "rsync                     3.1.3             h84994c4_1002    conda-forge\n",
      "samtools                  1.4.1                         0    bioconda\n",
      "scipy                     1.2.1            py27h7c811a0_0  \n",
      "seqtk                     1.3                  h84994c4_1    bioconda\n",
      "setuptools                41.0.1                   py27_0    conda-forge\n",
      "simplejson                3.8.1                    py27_0    bioconda\n",
      "singledispatch            3.4.0.3               py27_1000    conda-forge\n",
      "sip                       4.18.1          py27hf484d3e_1000    conda-forge\n",
      "six                       1.12.0                py27_1000    conda-forge\n",
      "snpeff                    4.3.1t                        2    bioconda\n",
      "spades                    3.13.0                        0    bioconda\n",
      "spectra                   0.0.11                     py_1    conda-forge\n",
      "sqlite                    3.28.0               h8b20d00_0    conda-forge\n",
      "subprocess32              3.5.4            py27h516909a_0    conda-forge\n",
      "suitesparse               5.2.0                h9e4a6bb_0  \n",
      "tbb                       2019.7               hc9558a2_0    conda-forge\n",
      "tk                        8.6.9             hed695b0_1002    conda-forge\n",
      "tktable                   2.10                 h14c3975_0  \n",
      "tornado                   5.1.1           py27h14c3975_1000    conda-forge\n",
      "urllib3                   1.24.3                   py27_0    conda-forge\n",
      "vcflib                    1.0.0_rc2            h56106d0_2    bioconda\n",
      "vcftools                  0.1.15               he941832_2    bioconda\n",
      "virtualenv                16.0.0                   py27_0    anaconda\n",
      "wget                      1.20.1               h90d6eec_0    conda-forge\n",
      "wheel                     0.33.4                   py27_0    conda-forge\n",
      "xopen                     0.7.0                      py_0    bioconda\n",
      "xorg-kbproto              1.0.7             h14c3975_1002    conda-forge\n",
      "xorg-libice               1.0.9             h516909a_1004    conda-forge\n",
      "xorg-libsm                1.2.3             h84519dc_1000    conda-forge\n",
      "xorg-libx11               1.6.7             h14c3975_1000    conda-forge\n",
      "xorg-libxau               1.0.9                h14c3975_0    conda-forge\n",
      "xorg-libxdmcp             1.1.3                h516909a_0    conda-forge\n",
      "xorg-libxext              1.3.4                h516909a_0    conda-forge\n",
      "xorg-libxrender           0.9.10            h516909a_1002    conda-forge\n",
      "xorg-renderproto          0.11.1            h14c3975_1002    conda-forge\n",
      "xorg-xextproto            7.3.0             h14c3975_1002    conda-forge\n",
      "xorg-xproto               7.0.31            h14c3975_1007    conda-forge\n",
      "xz                        5.2.4             h14c3975_1001    conda-forge\n",
      "yaml                      0.1.7             h14c3975_1001    conda-forge\n",
      "zlib                      1.2.11            h14c3975_1004    conda-forge\n",
      "zstd                      1.4.0                h3b9ef0a_0    conda-forge\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "#export PATH=/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/bin:$PATH # In case we use different version of Conda than the system-wide installation\n",
    "source activate treponema\n",
    "# In case we want to export the system settings and software versions\n",
    "conda info\n",
    "conda list\n",
    "# In case we modified the environment want to export it\n",
    "#conda env export > treponema_mod.yml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Running the workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Input variables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "First, we have to setup few variables which will be used throughout the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "INPUT_DIR=\"/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/data/raw\" # Directory with the raw paired input files in fastq.gz - first pair has to be named xxx_R1.fastq.gz and the second xxx_R2.fastq.gz otherwise you would have to change the input suffixes for the individual steps\n",
    "OUTPUT_DIR=\"/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/results\" # Directory where we want to save the results\n",
    "\n",
    "REFERENCE=\"/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/data/references/SS14.fa.gz\" # Bacteria reference genome\n",
    "HOST_GENOME=\"/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/data/references/GCF_000001405.36_GRCh38.p10_genomic.fna.gz\" # Host genome reference sequence; for human - ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.36_GRCh38.p10/GCF_000001405.36_GRCh38.p10_genomic.fna.gz\n",
    "\n",
    "THREADS=12 # Number of threads we will use in the analysis\n",
    "\n",
    "ADAPTER_R1=\"CTGTCTCTTATACACATCT\" # R1 3' adapeter, if any\n",
    "ADAPTER_R2=$ADAPTER_R1 # R2 3' adapter, if any"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Initial quality check"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "It is always a good idea to run an initial quality check on your data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/data/raw/CW56_S1_R1_001.fastq.gz\n",
      "/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/data/raw/CW56_S1_R2_001.fastq.gz\n",
      "(treponema) (.env) (treponema) (.env) Started analysis of CW56_S1_R1_001.fastq.gz\n",
      "Started analysis of CW56_S1_R2_001.fastq.gz\n",
      "Approx 5% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 5% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 10% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 10% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 15% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 15% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 20% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 20% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 25% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 25% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 30% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 30% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 35% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 35% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 40% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 40% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 45% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 45% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 50% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 50% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 55% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 55% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 60% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 60% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 65% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 65% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 70% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 70% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 75% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 75% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 80% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 80% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 85% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 85% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 90% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 90% complete for CW56_S1_R2_001.fastq.gz\n",
      "Approx 95% complete for CW56_S1_R1_001.fastq.gz\n",
      "Approx 95% complete for CW56_S1_R2_001.fastq.gz\n",
      "Analysis complete for CW56_S1_R1_001.fastq.gz\n",
      "Unable to revert mtime: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/fonts\n",
      "Unable to revert mtime: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/fonts/modern\n",
      "Unable to revert mtime: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/fonts/symbols\n",
      "Analysis complete for CW56_S1_R2_001.fastq.gz\n",
      "(treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  configs = yaml.load(f)\n",
      "/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:51: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  sp = yaml.load(f)\n",
      "[INFO   ]         multiqc : This is MultiQC v1.7\n",
      "[INFO   ]         multiqc : Template    : default\n",
      "[INFO   ]         multiqc : Searching '/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/results/qc/fastqc/raw/'\n",
      "[INFO   ]          fastqc : Found 2 reports\n",
      "[INFO   ]         multiqc : Compressing plot data\n",
      "[INFO   ]         multiqc : Report      : ../../home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/results/qc/fastqc/raw/multiqc_report.html\n",
      "[INFO   ]         multiqc : Data        : ../../home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/results/qc/fastqc/raw/multiqc_data\n",
      "[INFO   ]         multiqc : MultiQC complete\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir -p $OUTPUT_DIR/qc/fastqc/raw\n",
    "\n",
    "ls $INPUT_DIR/*.gz\n",
    "\n",
    "fastqc --threads $THREADS --outdir $OUTPUT_DIR/qc/fastqc/raw $INPUT_DIR/*.gz\n",
    "\n",
    "multiqc --outdir $OUTPUT_DIR/qc/fastqc/raw $OUTPUT_DIR/qc/fastqc/raw/ "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Read preprocessing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Since we are working with DNA data and aiming for the results including the polymorphisms we should perform a careful preprocessing to remove the adapter sequences and perform quality trimming."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Now I am processing CW56_S1_R1_001.fastq.gz as first in a pair and CW56_S1_R2_001.fastq.gz as a second in a pair.\n",
      "Done processing CW56_S1_R1_001.fastq.gz and CW56_S1_R2_001.fastq.gz\n",
      "(treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  configs = yaml.load(f)\n",
      "/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:51: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  sp = yaml.load(f)\n",
      "[INFO   ]         multiqc : This is MultiQC v1.7\n",
      "[INFO   ]         multiqc : Template    : default\n",
      "[INFO   ]         multiqc : Searching '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/cutadapt/'\n",
      "[INFO   ]        cutadapt : Found 1 reports\n",
      "[INFO   ]         multiqc : Compressing plot data\n",
      "[INFO   ]         multiqc : Report      : ../results/qc/cutadapt/multiqc_report.html\n",
      "[INFO   ]         multiqc : Data        : ../results/qc/cutadapt/multiqc_data\n",
      "[INFO   ]         multiqc : MultiQC complete\n",
      "(treponema) (.env) (treponema) (.env) Started analysis of CW56_S1_R1.trimmed.fastq.gz\n",
      "Started analysis of CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 5% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 5% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 10% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 10% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 15% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 15% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 20% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 20% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 25% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 25% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 30% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 30% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 35% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 35% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 40% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 40% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 45% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 45% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 50% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 50% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 55% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 55% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 60% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 60% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 65% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 65% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 70% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 70% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 75% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 75% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 80% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 80% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 85% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 85% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 90% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 90% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Approx 95% complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Approx 95% complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "Analysis complete for CW56_S1_R1.trimmed.fastq.gz\n",
      "Analysis complete for CW56_S1_R2.trimmed.fastq.gz\n",
      "(treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  configs = yaml.load(f)\n",
      "/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:51: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  sp = yaml.load(f)\n",
      "[INFO   ]         multiqc : This is MultiQC v1.7\n",
      "[INFO   ]         multiqc : Template    : default\n",
      "[INFO   ]         multiqc : Searching '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/fastqc/preprocessed/'\n",
      "[INFO   ]          fastqc : Found 2 reports\n",
      "[INFO   ]         multiqc : Compressing plot data\n",
      "[INFO   ]         multiqc : Report      : ../results/qc/fastqc/preprocessed/multiqc_report.html\n",
      "[INFO   ]         multiqc : Data        : ../results/qc/fastqc/preprocessed/multiqc_data\n",
      "[INFO   ]         multiqc : MultiQC complete\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir -p $OUTPUT_DIR/data/preprocessed\n",
    "mkdir $OUTPUT_DIR/qc/cutadapt\n",
    "mkdir $OUTPUT_DIR/qc/fastqc/preprocessed\n",
    "\n",
    "cd $INPUT_DIR/\n",
    "\n",
    "for sample in *R1*.gz\n",
    "do\n",
    "    FORWARD=$sample\n",
    "    extension=\"${FORWARD##*R1}\"\n",
    "    REVERSE=${FORWARD%R1*}R2${extension}\n",
    "\n",
    "    # Input file check\n",
    "    if [ ! -f ${FORWARD} ]; then\n",
    "    echo \"Input file not found! First in pair.\"; echo ${FORWARD}\n",
    "    fi\n",
    "    if [ ! -f ${REVERSE} ]; then\n",
    "    echo \"Input file not found! Second in pair.\"; echo ${REVERSE}\n",
    "    fi\n",
    "\n",
    "    echo \"Now I am processing ${FORWARD} as first in a pair and ${REVERSE} as a second in a pair.\"\n",
    "\n",
    "    cutadapt -a $ADAPTER_R1 -A $ADAPTER_R2 \\\n",
    "    --times 1 --quality-cutoff 15,15 --trim-n \\\n",
    "    --error-rate 0.10 -O 3 --minimum-length 35 --max-n 0 \\\n",
    "    --output $OUTPUT_DIR/data/preprocessed/${FORWARD%$extension}.trimmed.fastq.gz \\\n",
    "    --paired-output $OUTPUT_DIR/data/preprocessed/${REVERSE%$extension}.trimmed.fastq.gz \\\n",
    "    $FORWARD $REVERSE &>$OUTPUT_DIR/qc/cutadapt/${FORWARD%_R1${extension}}.cutadapt.out\n",
    "\n",
    "    echo \"Done processing $FORWARD and $REVERSE\"\n",
    "done\n",
    "\n",
    "multiqc --outdir $OUTPUT_DIR/qc/cutadapt $OUTPUT_DIR/qc/cutadapt/\n",
    "\n",
    "fastqc --threads $THREADS --outdir $OUTPUT_DIR/qc/fastqc/preprocessed $OUTPUT_DIR/data/preprocessed/*.gz\n",
    "\n",
    "multiqc --outdir $OUTPUT_DIR/qc/fastqc/preprocessed $OUTPUT_DIR/qc/fastqc/preprocessed/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Host genome contamination removal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Removal of the host genome DNA before the analysis speeds up the analysis and we will be working with much smaller files as well. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "As is usuall, we have to generate a host genome DNA reference index. This is run just once for one reference and one BBMap version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/bbmap_index\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) java -Djava.library.path=/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/opt/bbmap-38.22-1/jni/ -ea -Xmx20g -cp /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/opt/bbmap-38.22-1/current/ align2.BBMap build=1 overwrite=true fastareadlen=500 ref=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/GCF_000001405.36_GRCh38.p10_genomic.fna.gz -Xmx20g path=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/bbmap_index/GCF_000001405.36_GRCh38.p10_genomic.fna.gz\n",
      "Executing align2.BBMap [build=1, overwrite=true, fastareadlen=500, ref=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/GCF_000001405.36_GRCh38.p10_genomic.fna.gz, -Xmx20g, path=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/bbmap_index/GCF_000001405.36_GRCh38.p10_genomic.fna.gz]\n",
      "Version 38.22\n",
      "\n",
      "No output file.\n",
      "Writing reference.\n",
      "Executing dna.FastaToChromArrays2 [/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/GCF_000001405.36_GRCh38.p10_genomic.fna.gz, 1, writeinthread=false, genscaffoldinfo=true, retain, waitforwriting=false, gz=true, maxlen=536670912, writechroms=true, minscaf=1, midpad=300, startpad=8000, stoppad=8000, nodisk=false]\n",
      "\n",
      "Set genScaffoldInfo=true\n",
      "Writing chunk 1\n",
      "Writing chunk 2\n",
      "Writing chunk 3\n",
      "Writing chunk 4\n",
      "Writing chunk 5\n",
      "Writing chunk 6\n",
      "Writing chunk 7\n",
      "Set genome to 1\n",
      "\n",
      "Loaded Reference:\t0.052 seconds.\n",
      "Loading index for chunk 1-7, build 1\n",
      "No index available; generating from reference genome: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/bbmap_index/GCF_000001405.36_GRCh38.p10_genomic.fna.gz/ref/index/1/chr1-3_index_k13_c2_b1.block\n",
      "No index available; generating from reference genome: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/bbmap_index/GCF_000001405.36_GRCh38.p10_genomic.fna.gz/ref/index/1/chr4-7_index_k13_c2_b1.block\n",
      "Indexing threads started for block 0-3\n",
      "Indexing threads started for block 4-7\n",
      "Indexing threads finished for block 0-3\n",
      "Indexing threads finished for block 4-7\n",
      "Generated Index:\t477.441 seconds.\n",
      "Finished Writing:\t19.012 seconds.\n",
      "No reads to process; quitting.\n",
      "\n",
      "Total time:     \t551.026 seconds.\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir $(dirname $HOST_GENOME)/bbmap_index\n",
    "echo $(dirname $HOST_GENOME)/bbmap_index\n",
    "\n",
    "# If the host genome index does not exist create it\n",
    "if [ ! -d \"$(dirname $HOST_GENOME)/bbmap_index/$(basename $HOST_GENOME)\" ]; then\n",
    "    mkdir $(dirname $HOST_GENOME)/bbmap_index/$(basename $HOST_GENOME)\n",
    "    bbmap.sh ref=$HOST_GENOME -Xmx20g path=$(dirname $HOST_GENOME)/bbmap_index/$(basename $HOST_GENOME)\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Once the index is done we can launch the host-genome removal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Now I am processing CW56_S1_R1.trimmed.fastq.gz as first in a pair and CW56_S1_R2.trimmed.fastq.gz as a second in a pair with reference /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/GCF_000001405.36_GRCh38.p10_genomic.fna.gz.\n",
      "java -ea -Xmx12g -cp /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/opt/bbmap-38.22-1/current/ jgi.ReformatReads -Xmx12g verifypaired in=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/data/preprocessed/CW56_S1.clean.fastq.gz out1=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/data/preprocessed/CW56_S1_R1.clean.fastq.gz out2=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/data/preprocessed/CW56_S1_R2.clean.fastq.gz\n",
      "Executing jgi.ReformatReads [-Xmx12g, verifypaired, in=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/data/preprocessed/CW56_S1.clean.fastq.gz, out1=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/data/preprocessed/CW56_S1_R1.clean.fastq.gz, out2=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/data/preprocessed/CW56_S1_R2.clean.fastq.gz]\n",
      "\n",
      "Input is being processed as paired\n",
      "Input:                  \t9905136 reads          \t1399685204 bases\n",
      "Output:                 \t9905136 reads (100.00%) \t1399685204 bases (100.00%)\n",
      "\n",
      "Time:                         \t15.561 seconds.\n",
      "Reads Processed:       9905k \t636.52k reads/sec\n",
      "Bases Processed:       1399m \t89.95m bases/sec\n",
      "Names appear to be correctly paired.\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir $OUTPUT_DIR/qc/bbmap\n",
    "\n",
    "cd $OUTPUT_DIR/data/preprocessed/\n",
    "\n",
    "for sample in *R1*trimmed.fastq.gz\n",
    "do\n",
    "    FORWARD=$sample\n",
    "    extension=\"${FORWARD##*R1}\"\n",
    "    REVERSE=${FORWARD%R1*}R2${extension}\n",
    "\n",
    "    # Input file check\n",
    "    if [ ! -f ${FORWARD} ]; then\n",
    "        echo \"Input file not found! First in pair.\"; echo ${FORWARD}\n",
    "    fi\n",
    "    if [ ! -f ${REVERSE} ]; then\n",
    "        echo \"Input file not found! Second in pair.\"; echo ${REVERSE}\n",
    "    fi\n",
    "\n",
    "    echo \"Now I am processing ${FORWARD} as first in a pair and ${REVERSE} as a second in a pair with reference $HOST_GENOME.\"\n",
    "\n",
    "    # Start mapping\n",
    "    bbmap.sh threads=$THREADS -Xmx25g minid=0.95 maxindel=3 bandwidthratio=0.16 \\\n",
    "    bandwidth=12 quickmatch fast minhits=2 path=$(dirname $HOST_GENOME)/bbmap_index/$(basename $HOST_GENOME) unpigz pigz \\\n",
    "    in=${FORWARD} in2=${REVERSE} outu=$OUTPUT_DIR/data/preprocessed/${FORWARD%_R1${extension}}.clean.fastq.gz \\\n",
    "    outm=$OUTPUT_DIR/data/preprocessed/${FORWARD%_R1${extension}}.dirty.fastq.gz &>$OUTPUT_DIR/qc/bbmap/${FORWARD%_R1${extension}}.bbmap.out # qtrim=rl trimq=10 untrim  # We already have preprocessed data, no need for this\n",
    "\n",
    "    # De-interleave\n",
    "    reformat.sh -Xmx12g verifypaired in=$OUTPUT_DIR/data/preprocessed/${FORWARD%_R1${extension}}.clean.fastq.gz \\\n",
    "    out1=$OUTPUT_DIR/data/preprocessed/${FORWARD%${extension}}.clean.fastq.gz out2=$OUTPUT_DIR/data/preprocessed/${REVERSE%${extension}}.clean.fastq.gz\n",
    "\n",
    "    # Remove the host genome mapped reads (usefull for mapping precision)\n",
    "    rm $OUTPUT_DIR/data/preprocessed/${FORWARD%_R1${extension}}.clean.fastq.gz\n",
    "    rm $OUTPUT_DIR/data/preprocessed/${FORWARD%_R1${extension}}.dirty.fastq.gz\n",
    "done\n",
    "\n",
    "mkdir $OUTPUT_DIR/qc/fastqc/clean\n",
    "\n",
    "fastqc --threads $THREADS --outdir $OUTPUT_DIR/qc/fastqc/clean $OUTPUT_DIR/data/preprocessed/*.clean.fastq.gz\n",
    "\n",
    "multiqc --outdir $OUTPUT_DIR/qc/fastqc/clean $OUTPUT_DIR/qc/fastqc/clean/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Bacteria contamination scan"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Before we start with the alignment we can quickly scan for possible bacterial contamination in our dataset. This scan uses default StrainSeeker database which is most likely outdated but StrainSeeker offers a possibility to generate your [own index](http://bioinfo.ut.ee/strainseeker/index.php?r=site/page&view=manual#database) for the scan with their builder script. One advantage over tools such as [Kraken2](https://ccb.jhu.edu/software/kraken2/) (a great tool) is that is consumes much less RAM. However, the latest releases of MiniKraken2 could be used as well.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Before we can use StrainSeeker we have to download the database to scan. We can use the database provided at the StrainSeeker webpage or create our own. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "STRAINSEEKER_DB=/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/data/references/strainseekerdb\n",
    "\n",
    "mkdir $STRAINSEEKER_DB\n",
    "cd $STRAINSEEKER_DB/\n",
    "wget http://bioinfo.ut.ee/strainseeker/executables/ss_db_w32_4324.tar.gz\n",
    "tar xvzf ss_db_w32_4324.tar.gz\n",
    "\n",
    "STRAINSEEKER_DB=$STRAINSEEKER_DB/ss_db_w32_4324"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Once we have the database we can start the scan. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/strainseeker: File exists\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Working on sample CW56_S1_R1.clean.fastq.gz\n",
      "Subsampling\n",
      "Scanning\n",
      "Read tree info file...\n",
      "Getting k-mer counts from sample...\n",
      "Searching...\n",
      "------\n",
      "Reading tree info file: 0.0s\n",
      "Get k-mer counts from sample: 34.0s\n",
      "Search process: 9.0s\n",
      "Cleanup: 0.0s\n",
      "Database root-leaf count: 878\n",
      "TOTAL RUN TIME: 43.0s\n",
      "------\n",
      "Input files:\n",
      "CW56_S1_R1.clean.fastq.gz.sub\n",
      "Results saved in: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/strainseeker/CW56_S1_R1.clean.seeker.txt\n",
      "Working on sample CW56_S1_R2.clean.fastq.gz\n",
      "Subsampling\n",
      "Scanning\n",
      "Read tree info file...\n",
      "Getting k-mer counts from sample...\n",
      "Searching...\n",
      "------\n",
      "Reading tree info file: 0.0s\n",
      "Get k-mer counts from sample: 33.0s\n",
      "Search process: 8.0s\n",
      "Cleanup: 0.0s\n",
      "Database root-leaf count: 878\n",
      "TOTAL RUN TIME: 41.0s\n",
      "------\n",
      "Input files:\n",
      "CW56_S1_R2.clean.fastq.gz.sub\n",
      "Results saved in: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/strainseeker/CW56_S1_R2.clean.seeker.txt\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir $OUTPUT_DIR/qc/strainseeker\n",
    "\n",
    "cd $OUTPUT_DIR/data/preprocessed/\n",
    "\n",
    "for sample in *.clean.fastq.gz\n",
    "do \n",
    "    echo \"Working on sample $sample\"\n",
    "\n",
    "    echo \"Subsampling\"\n",
    "    seqtk sample -s100 $sample 1000000 > $sample.sub # Subsample fastq to 1M\n",
    "\n",
    "    echo \"Scanning\"\n",
    "    seeker.pl -i $sample.sub -d $STRAINSEEKER_DB -o $OUTPUT_DIR/qc/strainseeker/${sample%.fastq.gz}.seeker.txt\n",
    "\n",
    "    rm $sample.sub\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Bacteria contamination scan - Kraken2 (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "If you have enough resources you can use [Kraken2](https://ccb.jhu.edu/software/kraken2/). It scans the most recent bacterial, viral and fungal databases (or their subselection) and evaluates the possible distributions of individual species. If you decided to use Kraken2, you have to build the Kraken2 database. Please note this might take a while and will consume quite a lof of [RAM](https://ccb.jhu.edu/software/kraken2/index.shtml?t=manual#kraken-2-databases). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) 06182019\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "KRAKEN2_DB=\"/mnt/nfs/home/323639/000000-My_Documents/VM-home/projects/honza/linda/pipeline/data/references/kraken2db\" # Directory with stored StrainSeeker database\n",
    "mkdir $KRAKEN2_DB\n",
    "\n",
    "DATE=$(date +'%m%d%Y') # Save current date -> database version\n",
    "echo $DATE\n",
    "\n",
    "kraken2-build --standard --threads $THREADS --db $KRAKEN2_DB/$DATE\n",
    "\n",
    "KRAKEN2_DB=$KRAKEN2_DB/06182019"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Once we have the database we can start the scan. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(.env) (.env) (.env) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "THREADS=12\n",
    "OUTPUT_DIR=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results\n",
    "KRAKEN2_DB=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/kraken2db/06182019\n",
    "REFERENCE=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Now I am processing CW56_S1_R1.clean.fastq.gz as first in a pair and CW56_S1_R2.clean.fastq.gz as a second in a pair with database /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/kraken2db/06182019.\n",
      "Loading database information... done.\n",
      "4952568 sequences (1399.69 Mbp) processed in 39.405s (7541.1 Kseq/m, 2131.24 Mbp/m).\n",
      "  3344833 sequences classified (67.54%)\n",
      "  1607735 sequences unclassified (32.46%)\n",
      "[1] 60912\n",
      "[2] 60913\n",
      "[3] 60914\n",
      "[4] 60915\n",
      "(treponema) (.env) (treponema) (.env) [1]   Done                    gzip $output\n",
      "[3]-  Done                    gzip $output\n",
      "[4]+  Done                    gzip $output\n",
      "[2]+  Done                    gzip $output\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir -p $OUTPUT_DIR/qc/kraken2/reads\n",
    "\n",
    "cd $OUTPUT_DIR/data/preprocessed/\n",
    "\n",
    "for sample in *R1*.clean.fastq.gz\n",
    "do\n",
    "    FORWARD=$sample\n",
    "    extension=\"${FORWARD##*R1}\"\n",
    "    REVERSE=${FORWARD%R1*}R2${extension}\n",
    "\n",
    "    # Input file check\n",
    "    if [ ! -f ${FORWARD} ]; then\n",
    "        echo \"Input file not found! First in pair.\"; echo ${FORWARD}\n",
    "    fi\n",
    "    if [ ! -f ${REVERSE} ]; then\n",
    "        echo \"Input file not found! Second in pair.\"; echo ${REVERSE}\n",
    "    fi\n",
    "\n",
    "    echo \"Now I am processing ${FORWARD} as first in a pair and ${REVERSE} as a second in a pair with database $KRAKEN2_DB.\"\n",
    "    \n",
    "    kraken2 --use-names --paired --threads $THREADS --gzip-compressed --classified-out $OUTPUT_DIR/qc/kraken2/reads/${FORWARD%R1*}.classified-out#.fastq --unclassified-out $OUTPUT_DIR/qc/kraken2/reads/${FORWARD%R1*}.unclassified-out#.fastq --output $OUTPUT_DIR/qc/kraken2/reads/${FORWARD%R1*}.kraken.txt --report $OUTPUT_DIR/qc/kraken2/reads/${FORWARD%R1*}.kraken.report.txt --db $KRAKEN2_DB $FORWARD $REVERSE\n",
    "\n",
    "    for output in $OUTPUT_DIR/qc/kraken2/reads/*classified*.fastq\n",
    "    do\n",
    "        gzip $output &\n",
    "    done \n",
    "done\n",
    "\n",
    "wait\n",
    "\n",
    "# Get only species lines\n",
    "for report in $OUTPUT_DIR/qc/kraken2/reads/*.kraken.report.txt\n",
    "do \n",
    "    grep -P 'unclassified|root|\\tS\\t' $report | sort -k1,1nr > ${report%.kraken.report.txt*}.S.kraken.report.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Bacteria reference genome alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "With host genome DNA cleaned data we can proceed to the alignment to the reference. Please note we apply few filterings already at this step, mainly the minimal mapping quality, pairing of the reads and read duplication."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) [bwa_index] Pack FASTA... 0.01 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.25 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.01 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.01 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.12 sec\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: bwa index /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa\n",
      "[main] Real time: 0.399 sec; CPU: 0.396 sec\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Now I am processing CW56_S1_R1.clean.fastq.gz as first in a pair and CW56_S1_R2.clean.fastq.gz as a second in a pair with reference /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (83, 132, 235)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 539)\n",
      "[M::mem_pestat] mean and std.dev: (168.32, 124.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 691)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (166, 234, 335)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 673)\n",
      "[M::mem_pestat] mean and std.dev: (257.99, 128.35)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 842)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (52, 100, 277)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 727)\n",
      "[M::mem_pestat] mean and std.dev: (113.93, 89.19)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 952)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (68, 184, 272)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 680)\n",
      "[M::mem_pestat] mean and std.dev: (186.00, 113.58)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 884)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (168, 237, 340)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 684)\n",
      "[M::mem_pestat] mean and std.dev: (261.61, 129.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 856)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (54, 104, 203)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 501)\n",
      "[M::mem_pestat] mean and std.dev: (114.17, 97.68)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 650)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (75, 146, 224)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 522)\n",
      "[M::mem_pestat] mean and std.dev: (163.36, 93.22)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 671)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (167, 237, 341)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 689)\n",
      "[M::mem_pestat] mean and std.dev: (261.79, 131.00)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 863)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (52, 73, 246)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 634)\n",
      "[M::mem_pestat] mean and std.dev: (125.24, 99.03)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 828)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (88, 152, 235)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 529)\n",
      "[M::mem_pestat] mean and std.dev: (168.62, 111.45)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 676)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (167, 236, 340)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 686)\n",
      "[M::mem_pestat] mean and std.dev: (261.27, 131.10)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 859)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (78, 186, 424)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 1116)\n",
      "[M::mem_pestat] mean and std.dev: (260.95, 255.05)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1462)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (76, 123, 193)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 427)\n",
      "[M::mem_pestat] mean and std.dev: (121.47, 66.53)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 544)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (167, 237, 338)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 680)\n",
      "[M::mem_pestat] mean and std.dev: (260.75, 130.33)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 851)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (59, 100, 252)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 638)\n",
      "[M::mem_pestat] mean and std.dev: (116.82, 82.52)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 831)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (123, 160, 261)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 537)\n",
      "[M::mem_pestat] mean and std.dev: (170.52, 69.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 675)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (169, 237, 339)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 679)\n",
      "[M::mem_pestat] mean and std.dev: (261.64, 128.86)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 849)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (56, 104, 364)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 980)\n",
      "[M::mem_pestat] mean and std.dev: (185.73, 211.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 1288)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (68, 140, 287)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 725)\n",
      "[M::mem_pestat] mean and std.dev: (178.38, 134.36)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 944)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (166, 234, 333)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 667)\n",
      "[M::mem_pestat] mean and std.dev: (256.88, 126.64)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 834)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (80, 175, 309)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 767)\n",
      "[M::mem_pestat] mean and std.dev: (155.17, 101.11)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 996)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (49, 119, 184)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 454)\n",
      "[M::mem_pestat] mean and std.dev: (125.05, 104.55)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 589)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (164, 232, 330)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 662)\n",
      "[M::mem_pestat] mean and std.dev: (254.62, 125.54)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 828)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (80, 104, 277)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 671)\n",
      "[M::mem_pestat] mean and std.dev: (115.71, 71.29)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 868)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (59, 118, 161)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 365)\n",
      "[M::mem_pestat] mean and std.dev: (106.88, 60.97)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 467)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (164, 232, 329)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 659)\n",
      "[M::mem_pestat] mean and std.dev: (253.99, 125.16)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 824)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (85, 104, 288)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 694)\n",
      "[M::mem_pestat] mean and std.dev: (144.52, 98.83)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 897)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (90, 163, 266)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 618)\n",
      "[M::mem_pestat] mean and std.dev: (187.00, 122.20)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 794)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (164, 232, 330)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 662)\n",
      "[M::mem_pestat] mean and std.dev: (254.19, 125.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 828)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (72, 120, 252)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 612)\n",
      "[M::mem_pestat] mean and std.dev: (146.19, 112.39)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 792)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (72, 183, 291)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 729)\n",
      "[M::mem_pestat] mean and std.dev: (199.52, 164.44)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 948)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (165, 233, 330)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 660)\n",
      "[M::mem_pestat] mean and std.dev: (255.10, 125.32)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 825)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (84, 194, 892)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2508)\n",
      "[M::mem_pestat] mean and std.dev: (307.31, 371.91)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3316)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FF...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (94, 173, 260)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 592)\n",
      "[M::mem_pestat] mean and std.dev: (187.25, 114.43)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 758)\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation FR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (172, 241, 343)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 685)\n",
      "[M::mem_pestat] mean and std.dev: (265.37, 129.88)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 856)\n",
      "[M::mem_pestat] skip orientation RF as there are not enough pairs\n",
      "[M::mem_pestat] analyzing insert size distribution for orientation RR...\n",
      "[M::mem_pestat] (25, 50, 75) percentile: (78, 100, 995)\n",
      "[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 2829)\n",
      "[M::mem_pestat] mean and std.dev: (338.50, 467.50)\n",
      "[M::mem_pestat] low and high boundaries for proper pairs: (1, 3746)\n",
      "[M::mem_pestat] skip orientation FF\n",
      "[M::mem_pestat] skip orientation RR\n",
      "[main] Version: 0.7.15-r1140\n",
      "[main] CMD: bwa mem -t 12 -T 20 -v 1 -M -R @RG\\tID:1\\tLB:CW56_S1_R1\\tPL:Illumina\\tSM:CW56_S1_R1\\tPU:CW56_S1_R1 /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa CW56_S1_R1.clean.fastq.gz CW56_S1_R2.clean.fastq.gz\n",
      "[main] Real time: 64.832 sec; CPU: 672.536 sec\n",
      "Mapping finished\n",
      "Starting post-alignment processing and basic filtering\n",
      "[1] 34474\n",
      "[1]+  Done                    samtools flagstat $i > $OUTPUT_DIR/qc/alignment_stats/${i%.*}.flagstat\n",
      "Filtering finished\n",
      "Starting indel realignment\n",
      "[Wed Jun 26 20:01:18 CEST 2019] picard.sam.CreateSequenceDictionary REFERENCE=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa OUTPUT=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.dict    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 26 20:01:18 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "[Wed Jun 26 20:01:18 CEST 2019] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.00 minutes.\n",
      "Runtime.totalMemory()=514850816\n",
      "INFO  20:01:20,172 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  20:01:20,175 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.7-0-gcfedb67, Compiled 2016/12/12 11:21:18 \n",
      "INFO  20:01:20,175 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  20:01:20,175 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  20:01:20,176 HelpFormatter - [Wed Jun 26 20:01:20 CEST 2019] Executing on Linux 4.4.0-148-generic amd64 \n",
      "INFO  20:01:20,176 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_192-b01 \n",
      "INFO  20:01:20,179 HelpFormatter - Program Args: -T RealignerTargetCreator --num_threads 12 -R /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa -I CW56_S1_.SS14.filt.bam -o /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results//mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14/alignment/CW56_S1_.SS14.filt.forIndelRealigner.intervals \n",
      "INFO  20:01:20,187 HelpFormatter - Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01. \n",
      "INFO  20:01:20,188 HelpFormatter - Date/Time: 2019/06/26 20:01:20 \n",
      "INFO  20:01:20,188 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  20:01:20,188 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  20:01:20,193 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  20:01:20,293 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  20:01:20,301 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,325 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.02 \n",
      "INFO  20:01:20,349 MicroScheduler - Running the GATK in parallel mode with 12 total threads, 1 CPU thread(s) for each of 12 data thread(s), of 70 processors available on this machine \n",
      "INFO  20:01:20,480 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  20:01:20,514 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  20:01:20,514 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  20:01:20,514 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  20:01:20,515 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  20:01:20,522 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,526 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,527 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,529 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,530 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,532 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,533 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,535 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,536 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,538 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,539 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,542 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,542 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,545 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,545 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,548 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,549 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,551 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,551 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,555 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:20,556 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:20,559 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.00 \n",
      "INFO  20:01:35,196 ProgressMeter -            done   1139569.0    14.0 s      12.0 s      100.0%    14.0 s       0.0 s \n",
      "INFO  20:01:35,197 ProgressMeter - Total runtime 14.68 secs, 0.24 min, 0.00 hours \n",
      "INFO  20:01:35,198 MicroScheduler - 68773 reads were filtered out during the traversal out of approximately 3985412 total reads (1.73%) \n",
      "INFO  20:01:35,198 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  20:01:35,198 MicroScheduler -   -> 0 reads (0.00% of total) failing BadMateFilter \n",
      "INFO  20:01:35,199 MicroScheduler -   -> 0 reads (0.00% of total) failing DuplicateReadFilter \n",
      "INFO  20:01:35,199 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  20:01:35,199 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  20:01:35,200 MicroScheduler -   -> 0 reads (0.00% of total) failing MappingQualityUnavailableFilter \n",
      "INFO  20:01:35,200 MicroScheduler -   -> 68773 reads (1.73% of total) failing MappingQualityZeroFilter \n",
      "INFO  20:01:35,200 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  20:01:35,200 MicroScheduler -   -> 0 reads (0.00% of total) failing Platform454Filter \n",
      "INFO  20:01:35,201 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were no warn messages.\n",
      "------------------------------------------------------------------------------------------\n",
      "INFO  20:01:37,046 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  20:01:37,049 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.7-0-gcfedb67, Compiled 2016/12/12 11:21:18 \n",
      "INFO  20:01:37,049 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  20:01:37,049 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  20:01:37,049 HelpFormatter - [Wed Jun 26 20:01:37 CEST 2019] Executing on Linux 4.4.0-148-generic amd64 \n",
      "INFO  20:01:37,050 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_192-b01 \n",
      "INFO  20:01:37,053 HelpFormatter - Program Args: -I CW56_S1_.SS14.filt.bam -R /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa -T IndelRealigner -LOD 2.5 --consensusDeterminationModel USE_SW -targetIntervals /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results//mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14/alignment/CW56_S1_.SS14.filt.forIndelRealigner.intervals -o /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results//mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14/alignment/CW56_S1_.SS14.filt.indelRealigned.bam \n",
      "INFO  20:01:37,062 HelpFormatter - Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01. \n",
      "INFO  20:01:37,062 HelpFormatter - Date/Time: 2019/06/26 20:01:37 \n",
      "INFO  20:01:37,062 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  20:01:37,062 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  20:01:37,146 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  20:01:37,228 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  20:01:37,235 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  20:01:37,257 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.02 \n",
      "INFO  20:01:37,405 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  20:01:37,410 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  20:01:37,410 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  20:01:37,411 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  20:01:37,411 ProgressMeter -        Location |     reads | elapsed |     reads | completed | runtime |   runtime \n",
      "INFO  20:01:37,441 ReadShardBalancer$1 - Loading BAM index data \n",
      "INFO  20:01:37,442 ReadShardBalancer$1 - Done loading BAM index data \n",
      "INFO  20:02:07,438 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:47339    100002.0    30.0 s       5.0 m        4.2%    12.0 m      11.5 m \n",
      "INFO  20:02:37,442 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:99854    300040.0    60.0 s       3.3 m        8.8%    11.4 m      10.4 m \n",
      "INFO  20:03:08,044 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:154105    500052.0    90.0 s       3.0 m       13.5%    11.1 m       9.6 m \n",
      "INFO  20:03:38,047 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:208584    700069.0   120.0 s       2.9 m       18.3%    10.9 m       8.9 m \n",
      "INFO  20:04:08,117 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:266472    900110.0     2.5 m       2.8 m       23.4%    10.7 m       8.2 m \n",
      "INFO  20:04:38,119 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:337140   1200136.0     3.0 m       2.5 m       29.6%    10.1 m       7.1 m \n",
      "INFO  20:05:08,132 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:406650   1400155.0     3.5 m       2.5 m       35.7%     9.8 m       6.3 m \n",
      "INFO  20:05:38,134 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:459487   1600160.0     4.0 m       2.5 m       40.3%     9.9 m       5.9 m \n",
      "INFO  20:06:08,137 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:527497   1800170.0     4.5 m       2.5 m       46.3%     9.7 m       5.2 m \n",
      "INFO  20:06:38,139 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:581445   2000180.0     5.0 m       2.5 m       51.0%     9.8 m       4.8 m \n",
      "INFO  20:07:09,058 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:651910   2200216.0     5.5 m       2.5 m       57.2%     9.6 m       4.1 m \n",
      "INFO  20:07:39,061 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:727807   2500230.0     6.0 m       2.4 m       63.9%     9.4 m       3.4 m \n",
      "INFO  20:08:09,063 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:788117   2700239.0     6.5 m       2.4 m       69.2%     9.4 m       2.9 m \n",
      "INFO  20:08:39,065 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:854587   2900253.0     7.0 m       2.4 m       75.0%     9.4 m       2.3 m \n",
      "INFO  20:09:09,084 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:915591   3100266.0     7.5 m       2.4 m       80.3%     9.4 m     110.0 s \n",
      "INFO  20:09:39,086 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:976066   3300269.0     8.0 m       2.4 m       85.7%     9.4 m      80.0 s \n",
      "INFO  20:10:09,134 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:1033915   3500297.0     8.5 m       2.4 m       90.7%     9.4 m      52.0 s \n",
      "INFO  20:10:39,143 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:1096107   3700305.0     9.0 m       2.4 m       96.2%     9.4 m      21.0 s \n",
      "INFO  20:10:59,696 ProgressMeter -            done   3960618.0     9.4 m       2.4 m      100.0%     9.4 m       0.0 s \n",
      "INFO  20:10:59,697 ProgressMeter - Total runtime 562.29 secs, 9.37 min, 0.16 hours \n",
      "INFO  20:10:59,701 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 3960618 total reads (0.00%) \n",
      "INFO  20:10:59,702 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  20:10:59,702 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were no warn messages.\n",
      "------------------------------------------------------------------------------------------\n",
      "Finished indel realignment\n",
      "Starting read duplicate removal\n",
      "[Wed Jun 26 20:11:01 CEST 2019] picard.sam.markduplicates.MarkDuplicates INPUT=[CW56_S1_.SS14.filt.indelRealigned.bam] OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.bam METRICS_FILE=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14/qc/picard_dup/CW56_S1_.SS14.filt.indelRealigned.dedupStats.txt REMOVE_DUPLICATES=true OPTICAL_DUPLICATE_PIXEL_DISTANCE=100    MAX_SEQUENCES_FOR_DISK_READ_ENDS_MAP=50000 MAX_FILE_HANDLES_FOR_READ_ENDS_MAP=8000 SORTING_COLLECTION_SIZE_RATIO=0.25 TAG_DUPLICATE_SET_MEMBERS=false REMOVE_SEQUENCING_DUPLICATES=false TAGGING_POLICY=DontTag ASSUME_SORTED=false DUPLICATE_SCORING_STRATEGY=SUM_OF_BASE_QUALITIES PROGRAM_RECORD_ID=MarkDuplicates PROGRAM_GROUP_NAME=MarkDuplicates READ_NAME_REGEX=<optimized capture of last three ':' separated fields as numeric values> VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 26 20:11:01 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "INFO\t2019-06-26 20:11:01\tMarkDuplicates\tStart of doWork freeMemory: 507089912; totalMemory: 514850816; maxMemory: 954728448\n",
      "INFO\t2019-06-26 20:11:01\tMarkDuplicates\tReading input file and constructing read end information.\n",
      "INFO\t2019-06-26 20:11:01\tMarkDuplicates\tWill retain up to 3459161 data points before spilling to disk.\n",
      "INFO\t2019-06-26 20:11:08\tMarkDuplicates\tRead     1,000,000 records.  Elapsed time: 00:00:06s.  Time for last 1,000,000:    6s.  Last read position: gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:266,273\n",
      "INFO\t2019-06-26 20:11:08\tMarkDuplicates\tTracking 190 as yet unmatched pairs. 190 records in RAM.\n",
      "INFO\t2019-06-26 20:11:14\tMarkDuplicates\tRead     2,000,000 records.  Elapsed time: 00:00:12s.  Time for last 1,000,000:    5s.  Last read position: gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:570,398\n",
      "INFO\t2019-06-26 20:11:14\tMarkDuplicates\tTracking 162 as yet unmatched pairs. 162 records in RAM.\n",
      "INFO\t2019-06-26 20:11:26\tMarkDuplicates\tRead     3,000,000 records.  Elapsed time: 00:00:24s.  Time for last 1,000,000:   11s.  Last read position: gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:872,643\n",
      "INFO\t2019-06-26 20:11:26\tMarkDuplicates\tTracking 210 as yet unmatched pairs. 210 records in RAM.\n",
      "INFO\t2019-06-26 20:11:34\tMarkDuplicates\tRead 3960618 records. 0 pairs never matched.\n",
      "INFO\t2019-06-26 20:11:39\tMarkDuplicates\tAfter buildSortedReadEndLists freeMemory: 720305248; totalMemory: 885522432; maxMemory: 954728448\n",
      "INFO\t2019-06-26 20:11:39\tMarkDuplicates\tWill retain up to 29835264 duplicate indices before spilling to disk.\n",
      "INFO\t2019-06-26 20:11:39\tMarkDuplicates\tTraversing read pair information and detecting duplicates.\n",
      "INFO\t2019-06-26 20:11:40\tMarkDuplicates\tTraversing fragment information and detecting duplicates.\n",
      "INFO\t2019-06-26 20:11:41\tMarkDuplicates\tSorting list of duplicate records.\n",
      "INFO\t2019-06-26 20:11:41\tMarkDuplicates\tAfter generateDuplicateIndexes freeMemory: 636283840; totalMemory: 883425280; maxMemory: 954728448\n",
      "INFO\t2019-06-26 20:11:41\tMarkDuplicates\tMarking 3091308 records as duplicates.\n",
      "INFO\t2019-06-26 20:11:41\tMarkDuplicates\tFound 106 optical duplicate clusters.\n",
      "INFO\t2019-06-26 20:11:41\tMarkDuplicates\tReads are assumed to be ordered by: coordinate\n",
      "INFO\t2019-06-26 20:12:03\tMarkDuplicates\tBefore output close freeMemory: 887147816; totalMemory: 896532480; maxMemory: 954728448\n",
      "INFO\t2019-06-26 20:12:03\tMarkDuplicates\tAfter output close freeMemory: 880329064; totalMemory: 889716736; maxMemory: 954728448\n",
      "[Wed Jun 26 20:12:03 CEST 2019] picard.sam.markduplicates.MarkDuplicates done. Elapsed time: 1.03 minutes.\n",
      "Runtime.totalMemory()=889716736\n",
      "bash: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results//mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14/alignment_stats/CW56_S1_.SS14.filt.indelRealigned.dedup.flagstat: No such file or directory\n",
      "Finished read duplicate removal\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Prepare reference indexes\n",
    "if [[ ${REFERENCE##*.} == \"gz\" ]] # Uncompress the reference if it is in gz archive\n",
    "then\n",
    "    gunzip -c $REFERENCE > ${REFERENCE%.gz*}\n",
    "    REFERENCE=${REFERENCE%.gz*}\n",
    "fi\n",
    "\n",
    "if [[ `echo $REFERENCE | grep \".fna$\"` != \"\" ]] || [[ `echo $REFERENCE | grep \".fasta$\"` != \"\" ]] # Make a link to the original to have \".fa\" suffix\n",
    "then\n",
    "    ln -s $REFERENCE ${REFERENCE%.*}.fa\n",
    "fi\n",
    "\n",
    "bwa index $REFERENCE\n",
    "samtools faidx $REFERENCE\n",
    "\n",
    "cd $OUTPUT_DIR/data/preprocessed\n",
    "\n",
    "OUTPUT_DIR=${OUTPUT_DIR}/$(basename $REFERENCE .fa} # Adjust output directory to reflect the used reference genome\n",
    "\n",
    "mkdir -p $OUTPUT_DIR/alignment\n",
    "mkdir -p $OUTPUT_DIR/qc/alignment_stats\n",
    "\n",
    "for sample in *R1*clean.fastq.gz\n",
    "do\n",
    "    FORWARD=$sample\n",
    "    extension=\"${FORWARD##*R1}\"\n",
    "    REVERSE=${FORWARD%R1*}R2${extension}\n",
    "\n",
    "    # Input file check\n",
    "    if [ ! -f ${FORWARD} ]; then\n",
    "        echo \"Input file not found! First in pair.\"; echo ${FORWARD}\n",
    "    fi\n",
    "    if [ ! -f ${REVERSE} ]; then\n",
    "        echo \"Input file not found! Second in pair.\"; echo ${REVERSE}\n",
    "    fi\n",
    "\n",
    "    # Start mapping\n",
    "    echo \"Now I am processing ${FORWARD} as first in a pair and ${REVERSE} as a second in a pair with reference $REFERENCE\"\n",
    "\n",
    "    bwa mem -t $THREADS -T 20 -v 1 -M -R \"@RG\\tID:1\\tLB:${sample%%.*}\\tPL:Illumina\\tSM:${sample%%.*}\\tPU:${sample%%.*}\" $REFERENCE ${FORWARD} ${REVERSE} | samtools view -F 4 -@ $THREADS -b - | samtools sort -@ $THREADS - > $OUTPUT_DIR/alignment/${FORWARD%R1*}.$(basename $REFERENCE .fa).bam \n",
    "\n",
    "    echo \"Mapping finished\"\n",
    "\n",
    "    echo \"Starting post-alignment processing and basic filtering\"\n",
    "    \n",
    "    cd $OUTPUT_DIR/alignment\n",
    "\n",
    "    i=${FORWARD%R1*}.$(basename $REFERENCE .fa).bam\n",
    "\n",
    "    samtools index -@ $THREADS $i # Index BAM files\n",
    "    samtools flagstat $i > $OUTPUT_DIR/qc/alignment_stats/${i%.*}.flagstat &\n",
    "    samtools view -@ $THREADS -h -F 12 -f 2 -F 256 -b $i | samtools sort -n -@ $THREADS - | samtools fixmate -O bam - - | samtools sort -@ $THREADS - > ${i%.*}.filt.bam # -F 2048 = supplementary alignment, \"chimeric/non-linear alignments\"; -q $MAPQ\n",
    "    samtools index -@ $THREADS ${i%.*}.filt.bam\n",
    "    samtools flagstat ${i%.*}.filt.bam > $OUTPUT_DIR/qc/alignment_stats/${i%.*}.filt.flagstat\n",
    "\n",
    "    echo \"Filtering finished\"\n",
    "\n",
    "    echo \"Starting indel realignment\"\n",
    "    # Indel realignment\n",
    "    # Create input seqence dictionary and index reference\n",
    "    rm ${REFERENCE%.*}.dict\n",
    "    picard CreateSequenceDictionary R=$REFERENCE O=${REFERENCE%.*}.dict\n",
    "\n",
    "    # Realign\n",
    "    i=${i%.*}.filt.bam\n",
    "\n",
    "    gatk -T RealignerTargetCreator --num_threads $THREADS -R $REFERENCE -I ${i} -o $OUTPUT_DIR/alignment/${i%.*}.forIndelRealigner.intervals # Prepare intervals\n",
    "    gatk -I ${i} -R $REFERENCE -T IndelRealigner -LOD 2.5 --consensusDeterminationModel USE_SW -targetIntervals $OUTPUT_DIR/alignment/${i%.*}.forIndelRealigner.intervals -o $OUTPUT_DIR/alignment/${i%.*}.indelRealigned.bam # Run re-alignment\n",
    "\n",
    "    samtools index -@ $THREADS ${i%.*}.indelRealigned.bam\n",
    "\n",
    "    rm $i\n",
    "    rm $i*\n",
    "    rm $OUTPUT_DIR/alignment/${i%.*}.forIndelRealigner.intervals\n",
    "\n",
    "    echo \"Finished indel realignment\"\n",
    "\n",
    "    echo \"Starting read duplicate removal\"\n",
    "\n",
    "    # Remove duplicates\n",
    "    i=${i%.*}.indelRealigned.bam\n",
    "\n",
    "    mkdir $OUTPUT_DIR/qc/picard_dup\n",
    "    \n",
    "    picard MarkDuplicates INPUT=$i OUTPUT=${i%.*}.dedup.bam METRICS_FILE=$OUTPUT_DIR/qc/picard_dup/${i%.*}.dedupStats.txt REMOVE_DUPLICATES=true OPTICAL_DUPLICATE_PIXEL_DISTANCE=100 # Taggs ALL duplicates, PCR and optical and remove them\n",
    "\n",
    "    samtools index -@ $THREADS ${i%.*}.dedup.bam\n",
    "    samtools flagstat ${i%.*}.dedup.bam > $OUTPUT_DIR/alignment_stats/${i%.*}.dedup.flagstat\n",
    "\n",
    "    rm $i\n",
    "    rm $i*\n",
    "    \n",
    "    echo \"Finished read duplicate removal\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Post-alignment filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "The most important part of the processing is the **post-alignment filtering**. `BWA MEM` is known to be very sensitive but not very specific. Right now, we have a lot of alignments which in reality do not belong to our reference. This is due to several factors used in BWA MEM, such as minimal length of alignment (default: 19, can be adjusted by `-k [INT]`), allowance of extensive soft-clipping, allowed hard-clipping and allowing a lot of mismatches. \n",
    "\n",
    "The simplest way to see the alignment artifacts and cross-mappings is to check the reference alignment coverage and check highly uneven coverage peaks. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "We apply the following filters:\n",
    "\n",
    "1. To many mismatches\n",
    "    * max. 0.05% mismatches and\n",
    "    * max. 5 mismatches\n",
    "2. Very short alignments\n",
    "     * min 35 bp mapping (measured on the read, not the reference)\n",
    "3. Too much softclipped \n",
    "     * max 0.05% soft-clipped\n",
    "4. Supplementary/chimeric reads \n",
    "    * `samtools -F 2048` flag\n",
    "5. Too much hardclipped\n",
    "     * max 0.00% hard-clipped (no hard-clipping allowed)\n",
    "6. MAPQ 10 \n",
    "    * very often repetitive alignments\n",
    "7. MAPQ 40 \n",
    "    * get only high quality alignment\n",
    "8. Singletons \n",
    "    * remove singleton reads (not paired) after all the mapping filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) mkdir: cannot create directory /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp: File exists\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) Processing sample CW56_S1_.SS14.filt.indelRealigned.dedup.bam\n",
      "                                                                                \b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\bDone! (0:24)\n",
      "851379 kept\n",
      "17931 failed\n",
      "                                                                                \b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\bDone! (0:22)\n",
      "843691 kept\n",
      "7688 failed\n",
      "Number of too much mismatched reads is 23501 CW56_S1_.SS14.filt.indelRealigned.dedup.mm.fail.txt.tmp for sample  CW56_S1_.SS14.filt.indelRealigned.dedup.bam\n",
      "[Wed Jun 19 02:43:08 CEST 2019] picard.sam.FilterSamReads INPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.bam FILTER=includeReadList [OUTPUT SAM/BAM will contain reads that are supplied in the READ_LIST_FILE file] READ_LIST_FILE=CW56_S1_.SS14.filt.indelRealigned.dedup.mm.fail.txt.tmp OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.mm.filtOut.bam    WRITE_READS_FILES=true VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 19 02:43:08 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "INFO\t2019-06-19 02:43:12\tFilterSamReads\tFiltering [presorted=true] CW56_S1_.SS14.filt.indelRealigned.dedup.bam -> OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.mm.filtOut.bam [sortorder=coordinate]\n",
      "INFO\t2019-06-19 02:43:15\tFilterSamReads\t47,002 SAMRecords written to CW56_S1_.SS14.filt.indelRealigned.dedup.mm.filtOut.bam\n",
      "[Wed Jun 19 02:43:15 CEST 2019] picard.sam.FilterSamReads done. Elapsed time: 0.12 minutes.\n",
      "Runtime.totalMemory()=708837376\n",
      "[1] 2165\n",
      "[1]+  Done                    samtools view -h -@ $THREADS ${sample%.*}.mm2.bam | perl -slane '$l = 0; $F[5] =~ s/(\\d+)[MX=DN]/$l+=$1/eg; print if $l < $MIN_LENGTH_MAPPED or /^@/' -- -MIN_LENGTH_MAPPED=$MIN_LENGTH_MAPPED | samtools view -@ $THREADS -b - > ${sample%.*}.short.filtOut.bam\n",
      "                                                                                \b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\b\bDone! (0:19)\n",
      "Wrote: 840236 reads\n",
      "Altered: 13046\n",
      "Unmapped: 0\n",
      "Number of too much soft clipped reads is 10155 CW56_S1_.SS14.filt.indelRealigned.dedup.read_names_to_remove_highSoftClip.txt for sample  CW56_S1_.SS14.filt.indelRealigned.dedup.bam\n",
      "[1] 2258\n",
      "[Wed Jun 19 02:43:48 CEST 2019] picard.sam.FilterSamReads INPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.short.bam FILTER=excludeReadList [OUTPUT bam will contain reads that are *not* supplied in the READ_LIST_FILE file] READ_LIST_FILE=CW56_S1_.SS14.filt.indelRealigned.dedup.read_names_to_remove_highSoftClip.txt OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.scf.bam    WRITE_READS_FILES=true VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 19 02:43:48 CEST 2019] picard.sam.FilterSamReads INPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.short.bam FILTER=includeReadList [OUTPUT SAM/BAM will contain reads that are supplied in the READ_LIST_FILE file] READ_LIST_FILE=CW56_S1_.SS14.filt.indelRealigned.dedup.read_names_to_remove_highSoftClip.txt OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.scf.filtOut.bam    WRITE_READS_FILES=true VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 19 02:43:48 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "[Wed Jun 19 02:43:48 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "INFO\t2019-06-19 02:43:52\tFilterSamReads\tFiltering [presorted=true] CW56_S1_.SS14.filt.indelRealigned.dedup.short.bam -> OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.scf.bam [sortorder=coordinate]\n",
      "INFO\t2019-06-19 02:43:52\tFilterSamReads\tFiltering [presorted=true] CW56_S1_.SS14.filt.indelRealigned.dedup.short.bam -> OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.scf.filtOut.bam [sortorder=coordinate]\n",
      "INFO\t2019-06-19 02:43:55\tFilterSamReads\t18,480 SAMRecords written to CW56_S1_.SS14.filt.indelRealigned.dedup.scf.filtOut.bam\n",
      "[Wed Jun 19 02:43:55 CEST 2019] picard.sam.FilterSamReads done. Elapsed time: 0.10 minutes.\n",
      "Runtime.totalMemory()=708313088\n",
      "INFO\t2019-06-19 02:44:00\tFilterSamReads\t821,756 SAMRecords written to CW56_S1_.SS14.filt.indelRealigned.dedup.scf.bam\n",
      "WARNING: BAM index file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.scf.bam.bai is older than BAM /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.scf.bam\n",
      "[Wed Jun 19 02:44:03 CEST 2019] picard.sam.FilterSamReads done. Elapsed time: 0.24 minutes.\n",
      "Runtime.totalMemory()=709885952\n",
      "[1]+  Done                    picard FilterSamReads I=${sample%.*}.short.bam O=${sample%.*}.scf.filtOut.bam READ_LIST_FILE=${sample%.*}.read_names_to_remove_highSoftClip.txt FILTER=includeReadList\n",
      "[1] 2468\n",
      "[1]+  Done                    samtools view -@ $THREADS -h -f 2048 -b ${sample%.*}.scf.bam > ${sample%.*}.sup.filtOut.bam\n",
      "[1] 2500\n",
      "[1]+  Done                    samtools view -@ $THREADS -h ${sample%.*}.sup.bam | awk '$6 ~ /H/{print}' | samtools view -bh - > ${sample%.*}.hcf.filtOut.bam\n",
      "[1] 2532\n",
      "[2] 2563\n",
      "[1]-  Done                    samtools view -@ $THREADS -h ${sample%.*}.sup.bam | awk -v var=\"$MAPQ\" '$5 < var || $1 ~ /^@/' | samtools view -b - > ${sample%.*}.MAPQ${MAPQ}.filtOut.bam\n",
      "[2]+  Done                    samtools view -@ $THREADS -h ${sample%.*}.MAPQ${MAPQ}.bam | awk -v var=\"$MAPQ_FINAL\" '$5 < var || $1 ~ /^@/' | samtools view -b - > ${sample%.*}.MAPQ${MAPQ_FINAL}.filtOut.bam\n",
      "Number of singleton reads after filtering is 19963 CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.singleAfterFilt.txt for sample  CW56_S1_.SS14.filt.indelRealigned.dedup.bam\n",
      "[1] 2611\n",
      "[Wed Jun 19 02:44:32 CEST 2019] picard.sam.FilterSamReads INPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam FILTER=excludeReadList [OUTPUT bam will contain reads that are *not* supplied in the READ_LIST_FILE file] READ_LIST_FILE=CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.singleAfterFilt.txt OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.tmp    WRITE_READS_FILES=true VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 19 02:44:32 CEST 2019] picard.sam.FilterSamReads INPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam FILTER=includeReadList [OUTPUT SAM/BAM will contain reads that are supplied in the READ_LIST_FILE file] READ_LIST_FILE=CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.singleAfterFilt.txt OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.singletons.filtOut.bam    WRITE_READS_FILES=true VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 19 02:44:32 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "[Wed Jun 19 02:44:32 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "WARNING: BAM index file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.bai is older than BAM /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "WARNING: BAM index file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.bai is older than BAM /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "WARNING: BAM index file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.bai is older than BAM /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "WARNING: BAM index file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.bai is older than BAM /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "INFO\t2019-06-19 02:44:36\tFilterSamReads\tFiltering [presorted=true] CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam -> OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.tmp [sortorder=coordinate]\n",
      "WARNING: BAM index file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.bai is older than BAM /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "WARNING: BAM index file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.bai is older than BAM /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "INFO\t2019-06-19 02:44:36\tFilterSamReads\tFiltering [presorted=true] CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam -> OUTPUT=CW56_S1_.SS14.filt.indelRealigned.dedup.singletons.filtOut.bam [sortorder=coordinate]\n",
      "INFO\t2019-06-19 02:44:38\tFilterSamReads\t19,963 SAMRecords written to CW56_S1_.SS14.filt.indelRealigned.dedup.singletons.filtOut.bam\n",
      "[Wed Jun 19 02:44:38 CEST 2019] picard.sam.FilterSamReads done. Elapsed time: 0.11 minutes.\n",
      "Runtime.totalMemory()=708837376\n",
      "INFO\t2019-06-19 02:44:44\tFilterSamReads\t786,004 SAMRecords written to CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam.tmp\n",
      "[Wed Jun 19 02:44:46 CEST 2019] picard.sam.FilterSamReads done. Elapsed time: 0.24 minutes.\n",
      "Runtime.totalMemory()=709361664\n",
      "[1]+  Done                    picard FilterSamReads I=${sample%.*}.MAPQ${MAPQ_FINAL}.bam O=${sample%.*}.singletons.filtOut.bam READ_LIST_FILE=${sample%.*}.MAPQ${MAPQ_FINAL}.singleAfterFilt.txt FILTER=includeReadList\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Java memory size is set to 1200M\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.2.2-dev\n",
      "Built on 2017-08-28 08:37\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 32\n",
      "Max memory (Mb): 1118\n",
      "Wed Jun 19 02:45:53 CEST 2019\t\tWARNING\tOutput folder already exists, the results will be saved there\n",
      "\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 400\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 400 windows...\n",
      "Processed 100 out of 400 windows...\n",
      "Processed 150 out of 400 windows...\n",
      "Processed 200 out of 400 windows...\n",
      "Processed 250 out of 400 windows...\n",
      "Processed 300 out of 400 windows...\n",
      "Processed 350 out of 400 windows...\n",
      "Processed 400 out of 400 windows...\n",
      "Total processed windows:400\n",
      "Number of reads: 869310\n",
      "Number of valid reads: 869310\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 869310\n",
      "Num mapped first of pair: 434655\n",
      "Num mapped second of pair: 434655\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 8\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 116157204\n",
      "referenceSize: 1139569\n",
      "numberOfSequencedBases: 116151139\n",
      "numberOfAs: 27241454\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 8\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing PDF report...\n",
      "PDF file created successfully \n",
      "\n",
      "Finished\n",
      "Java memory size is set to 1200M\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.2.2-dev\n",
      "Built on 2017-08-28 08:37\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 32\n",
      "Max memory (Mb): 1118\n",
      "Wed Jun 19 02:46:04 CEST 2019\t\tWARNING\tOutput folder already exists, the results will be saved there\n",
      "\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 400\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 400 windows...\n",
      "Processed 100 out of 400 windows...\n",
      "Processed 150 out of 400 windows...\n",
      "Processed 200 out of 400 windows...\n",
      "Processed 250 out of 400 windows...\n",
      "Processed 300 out of 400 windows...\n",
      "Processed 350 out of 400 windows...\n",
      "Processed 400 out of 400 windows...\n",
      "Total processed windows:400\n",
      "Number of reads: 869310\n",
      "Number of valid reads: 869310\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 869310\n",
      "Num mapped first of pair: 434655\n",
      "Num mapped second of pair: 434655\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 8\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 116157204\n",
      "referenceSize: 1139569\n",
      "numberOfSequencedBases: 116151139\n",
      "numberOfAs: 27241454\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 8\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "(treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  configs = yaml.load(f)\n",
      "/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:51: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  sp = yaml.load(f)\n",
      "[INFO   ]         multiqc : This is MultiQC v1.7\n",
      "[INFO   ]         multiqc : Template    : default\n",
      "[INFO   ]         multiqc : Searching '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/qualimap/'\n",
      "\u001b[?25lSearching 52 files..  [####################################]  100%          \u001b[?25h\n",
      "[INFO   ]        qualimap : Found 1 BamQC reports\n",
      "[INFO   ]         multiqc : Compressing plot data\n",
      "[INFO   ]         multiqc : Report      : ../qc/qualimap/multiqc_report.html\n",
      "[INFO   ]         multiqc : Data        : ../qc/qualimap/multiqc_data\n",
      "[INFO   ]         multiqc : MultiQC complete\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) rm: cannot remove '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/*.bam.tmp': No such file or directory\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) mv: cannot move '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/filtered' to a subdirectory of itself, '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/filtered/filt/filtered'\n",
      "(treponema) (.env) mv: cannot stat '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/filt/*.MAPQ40.bam*': No such file or directory\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "MAX_PERC_OF_MM=0.05 # Maximum percentage of mismatches compared to the read length - bad if we have to error-prone reads but helps to remove false-positives\n",
    "MAX_NUMBER_OF_MM=5 # Maximum number of mismatches\n",
    "MIN_LENGTH_MAPPED=35 # Remove mappings that mapped with too few bases (remove excesive soft-clipping)\n",
    "MAX_SOFTCLIP=0.05 # Maximal percentage of the reads allowed to be soft-clipped\n",
    "MAX_HARDCLIP=0.00 # Maximal percentage of the reads allowed to be hard-clipped\n",
    "MAPQ=10 # Minimal MAPQ for (probably) repetitive regions\n",
    "MAPQ_FINAL=40 # Minimal MAPQ for final results\n",
    "\n",
    "cd $OUTPUT_DIR/alignment\n",
    "\n",
    "mkdir $OUTPUT_DIR/tmp\n",
    "\n",
    "# Filter mappings\n",
    "for sample in *.filt.indelRealigned.dedup.bam\n",
    "do \n",
    "    echo \"Processing sample $sample\"\n",
    "\n",
    "    samtools index -@ $THREADS $sample\n",
    "\n",
    "    # 1) Remove too many mismatches\n",
    "    bamutils filter $sample ${sample%.*}.mm1.bam -failed ${sample%.*}.mm1.fail.txt -maximum_mismatch_ratio $MAX_PERC_OF_MM # Filter by perc. of mismatches; It's more filtering on edit distance = how many nucleotides have to be changed to get exactly the reference sequence; indels are counted as many times as they are \"long\"; error in the source https://github.com/ngsutils/ngsutils/pull/18\n",
    "    bamutils filter ${sample%.*}.mm1.bam ${sample%.*}.mm2.bam -failed ${sample%.*}.mm2.fail.txt -mismatch $MAX_NUMBER_OF_MM # Filter by num. of mismatches\n",
    "\n",
    "    cat ${sample%.*}.mm1.fail.txt ${sample%.*}.mm2.fail.txt > ${sample%.*}.mm.fail.txt # Merge mapping filtered by mismatches\n",
    "\n",
    "    # Use extracted filtered read name and filter it out from the original bam but ONLY when the mismatches filtering had some results\n",
    "    if [ -s ${sample%.*}.mm.fail.txt ]\n",
    "    then\n",
    "        cat ${sample%.*}.mm.fail.txt | awk '{print $1}' | sort -T $OUTPUT_DIR/tmp --parallel=$THREADS | uniq > ${sample%.*}.mm.fail.txt.tmp\n",
    "        echo \"Number of too much mismatched reads is\" `wc -l ${sample%.*}.mm.fail.txt.tmp` \"for sample \" $sample\n",
    "        picard FilterSamReads I=$sample O=${sample%.*}.mm.filtOut.bam READ_LIST_FILE=${sample%.*}.mm.fail.txt.tmp FILTER=includeReadList # Include reads\n",
    "        rm ${sample%.*}.mm.fail.txt.tmp\n",
    "    else\n",
    "        echo \"There are none to much mismatched reads with \" $MAX_PERC_OF_MM \" % and \" $MAX_NUMBER_OF_MM \" mismatches for sample \" $sample\". Nothing to report.\"\n",
    "    fi\n",
    "\n",
    "    # 2) Remove too short alignment\n",
    "    # Filter out mappings that mapped with length shorter than MIN_LENGTH_MAPPED https://www.biostars.org/p/12406/ - this filters the mapping by calculating the length of mapping on the read itself; example 106S8M1D21M14S with MIN_LENGTH_MAPPED=30 is filtered out because 8M+21M=29\n",
    "    # Another filtering by mapped read length could be taken from here https://www.biostars.org/p/151510/ - this filters the mapping by calculating the length of mapping on the reference; example 106S8M1D21M14S with MIN_LENGTH_MAPPED=30 is NOT filtered out because 8M+1D+21M=30, it is filtered out when set to MIN_LENGTH_MAPPED=31\n",
    "    #samjs.jar -e '!record.readUnmappedFlag  && record.cigar.referenceLength  >= $MIN_LENGTH_MAPPED '$sample | samtools view -@ $THREADS -b -o ${sample%.*}.short.bam -\n",
    "    samtools view -h -@ $THREADS ${sample%.*}.mm2.bam | perl -slane '$l = 0; $F[5] =~ s/(\\d+)[MX=DN]/$l+=$1/eg; print if $l < $MIN_LENGTH_MAPPED or /^@/' -- -MIN_LENGTH_MAPPED=$MIN_LENGTH_MAPPED | samtools view -@ $THREADS -b - > ${sample%.*}.short.filtOut.bam &\n",
    "    samtools view -h -@ $THREADS ${sample%.*}.mm2.bam | perl -slane '$l = 0; $F[5] =~ s/(\\d+)[MX=DN]/$l+=$1/eg; print if $l >= $MIN_LENGTH_MAPPED or /^@/' -- -MIN_LENGTH_MAPPED=$MIN_LENGTH_MAPPED | samtools view -@ $THREADS -b - > ${sample%.*}.short.bam\n",
    "\n",
    "    # 3) Too much softclipped\n",
    "    # Running it on the whole file and saving it creates error in BAM validation; we have to just take the read names from here and remove them from the alignment\n",
    "    bamutils removeclipping ${sample%.*}.short.bam ${sample%.*}.scf.bam.tmp\n",
    "    samtools view -@ $THREADS ${sample%.*}.scf.bam.tmp | grep 'ZC:f:' | awk '{for (i=1;i<=NF;i++){if ($sample ~/ZC:f:/) {print $1, $sample}}}' | sed 's/ZC:f://' | awk -v MAX_SOFTCLIP=\"$MAX_SOFTCLIP\" ' $2 > MAX_SOFTCLIP {print $1}' > ${sample%.*}.read_names_to_remove_highSoftClip.txt.tmp # Get only softclipped reads above the threshold\n",
    "    rm ${sample%.*}.scf.bam.tmp # Remove temporary BAMs\n",
    "    cat ${sample%.*}.read_names_to_remove_highSoftClip.txt.tmp | cut -f 1 | sort -T $OUTPUT_DIR/tmp --parallel=$THREADS | uniq > ${sample%.*}.read_names_to_remove_highSoftClip.txt # Get unique read names with softclipping!\n",
    "    #cat ${sample%.*}.read_names_to_remove_highSoftClip.txt | cut -f 1 | sort -T $OUTPUT_DIR/tmp --parallel=$THREADS | uniq -d > tmp; mv tmp ${sample%.*}.read_names_to_remove_highSoftClip.txt # Get unique read names with softclipping if whole pair failed the filtering!\n",
    "    rm ${sample%.*}.read_names_to_remove_highSoftClip.txt.tmp\n",
    "\n",
    "    # Use extracted filtered read name and filter it out from the original bam but ONLY when the soft clipping filtering had some results\n",
    "    if [ -s ${sample%.*}.read_names_to_remove_highSoftClip.txt ]\n",
    "    then\n",
    "        echo \"Number of too much soft clipped reads is\" `wc -l ${sample%.*}.read_names_to_remove_highSoftClip.txt` \"for sample \" $sample\n",
    "        picard FilterSamReads I=${sample%.*}.short.bam O=${sample%.*}.scf.filtOut.bam READ_LIST_FILE=${sample%.*}.read_names_to_remove_highSoftClip.txt FILTER=includeReadList & # Include reads\n",
    "        picard FilterSamReads I=${sample%.*}.short.bam O=${sample%.*}.scf.bam READ_LIST_FILE=${sample%.*}.read_names_to_remove_highSoftClip.txt FILTER=excludeReadList # Exclude reads\n",
    "    else\n",
    "        echo \"There are none to much soft clipped reads with \" $MAX_SOFTCLIP \" % soft clipping for sample \" $sample. Continue without filtering.\n",
    "        mv ${sample%.*}.short.bam ${sample%.*}.scf.bam\n",
    "    fi\n",
    "\n",
    "    samtools index -@ $THREADS ${sample%.*}.scf.bam\n",
    "\n",
    "    # 4) Supplementary/chimeric alignment\n",
    "    samtools view -@ $THREADS -h -f 2048 -b ${sample%.*}.scf.bam > ${sample%.*}.sup.filtOut.bam & # get -F 2048 \"chimeric/non-linear alignments\" - might be interesting for translocations; possible overlap with hardclipping \n",
    "    samtools view -@ $THREADS -h -F 2048 -b ${sample%.*}.scf.bam > ${sample%.*}.sup.bam # remove -F 2048 \"chimeric/non-linear alignments\" - might be interesting for translocations; possible overlap with hardclipping\n",
    "\n",
    "    # 5) Hardclipped\n",
    "    samtools view -@ $THREADS -h ${sample%.*}.sup.bam | awk '$6 ~ /H/{print}' | samtools view -bh - > ${sample%.*}.hcf.filtOut.bam & # Get only hardclipped mappings; hardclipped alignments might mean the part of the alignment that is hardclipped might map to different part of the genome - possible chimeric reads? https://www.biostars.org/p/109333/\n",
    "    samtools view -@ $THREADS -h ${sample%.*}.sup.bam | awk '$6 !~ /H/{print}' | samtools view -bh - > ${sample%.*}.hcf.bam # https://www.biostars.org/p/137461/\n",
    "\n",
    "    # 6) MAPQ10  very often repetitive alignments\n",
    "    samtools view -@ $THREADS -h ${sample%.*}.sup.bam | awk -v var=\"$MAPQ\" '$5 < var || $1 ~ /^@/' | samtools view -b - > ${sample%.*}.MAPQ${MAPQ}.filtOut.bam &\n",
    "    samtools view -@ $THREADS -h -bq $MAPQ ${sample%.*}.sup.bam > ${sample%.*}.MAPQ${MAPQ}.bam # BWA-MEM -T settings should filter all low MAPQ mappings; -F 2048 \"chimeric/non-linear alignments\" - might be interesting for translocations\n",
    "\n",
    "    # 7) MAPQ40  Only high quality alignments\n",
    "    samtools view -@ $THREADS -h ${sample%.*}.MAPQ${MAPQ}.bam | awk -v var=\"$MAPQ_FINAL\" '$5 < var || $1 ~ /^@/' | samtools view -b - > ${sample%.*}.MAPQ${MAPQ_FINAL}.filtOut.bam &\n",
    "    samtools view -@ $THREADS -h -bq $MAPQ_FINAL ${sample%.*}.MAPQ${MAPQ}.bam > ${sample%.*}.MAPQ${MAPQ_FINAL}.bam # Get only very high quality mappings\n",
    "\n",
    "    # 8) Singletons\n",
    "    # Remove reads that remained as singletons after all filtering steps - works only for paired-end sequencing!\n",
    "    samtools view -@ $THREADS ${sample%.*}.MAPQ${MAPQ_FINAL}.bam | cut -f 1 | sort -T $OUTPUT_DIR/tmp --parallel=$THREADS | uniq -u > ${sample%.*}.MAPQ${MAPQ_FINAL}.singleAfterFilt.txt\n",
    "\n",
    "    if [ -s ${sample%.*}.MAPQ${MAPQ_FINAL}.singleAfterFilt.txt ]\n",
    "    then\n",
    "        echo \"Number of singleton reads after filtering is\" `wc -l ${sample%.*}.MAPQ${MAPQ_FINAL}.singleAfterFilt.txt` \"for sample \" $sample\n",
    "        picard FilterSamReads I=${sample%.*}.MAPQ${MAPQ_FINAL}.bam O=${sample%.*}.singletons.filtOut.bam READ_LIST_FILE=${sample%.*}.MAPQ${MAPQ_FINAL}.singleAfterFilt.txt FILTER=includeReadList & # Include reads\n",
    "        picard FilterSamReads I=${sample%.*}.MAPQ${MAPQ_FINAL}.bam O=${sample%.*}.MAPQ${MAPQ_FINAL}.bam.tmp READ_LIST_FILE=${sample%.*}.MAPQ${MAPQ_FINAL}.singleAfterFilt.txt FILTER=excludeReadList # Exclude reads\n",
    "        mv ${sample%.*}.MAPQ${MAPQ_FINAL}.bam.tmp ${sample%.*}.MAPQ${MAPQ_FINAL}.bam\n",
    "    else\n",
    "        echo \"There are none singleton reads after filtering for sample \" $sample. Continue without filtering.\n",
    "    fi\n",
    "\n",
    "    samtools index -@ $THREADS ${sample%.*}.MAPQ${MAPQ_FINAL}.bam\n",
    "\n",
    "    sleep 60 # wait/sleep for one minute for the jobs to finish (if they didn't)\n",
    "\n",
    "done\n",
    "\n",
    "for sample in $OUTPUT_DIR/alignment/*.bam\n",
    "do\n",
    "    samtools index -@ $THREADS $sample\n",
    "done\n",
    "\n",
    "# Do some cleaning\n",
    "rm $OUTPUT_DIR/alignment/*.bam.tmp $OUTPUT_DIR/alignment/*.read_names_to_remove_highSoftClip.txt $OUTPUT_DIR/alignment/*.reads $OUTPUT_DIR/alignment/*.singleAfterFilt.txt $OUTPUT_DIR/alignment/*.fail.txt\n",
    "\n",
    "mkdir -p $OUTPUT_DIR/alignment/filtered/filtOut\n",
    "mkdir -p $OUTPUT_DIR/alignment/filtered/filt\n",
    "\n",
    "mv $OUTPUT_DIR/alignment/*.filtOut.bam* $OUTPUT_DIR/alignment/filtered/filtOut/\n",
    "mv $OUTPUT_DIR/alignment/* $OUTPUT_DIR/alignment/filtered/filt/\n",
    "mv $OUTPUT_DIR/alignment/filtered/filt/*.MAPQ40.bam* $OUTPUT_DIR/alignment/filtered/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Please note that the filtering applied is rather strict. We might lose some true positive mappings but in this case we rather focus on strongly supported alignments to be sure we are detecting what we want to detect."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Alignment statistics and reference genome coverage "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "One of the the quality checks after the filtering and before all the other steps is the genome coverage statistics...and of course general mapping statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Java memory size is set to 1200M\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.2.2-dev\n",
      "Built on 2017-08-28 08:37\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 32\n",
      "Max memory (Mb): 1118\n",
      "Wed Jun 19 03:22:48 CEST 2019\t\tWARNING\tOutput folder already exists, the results will be saved there\n",
      "\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 400\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 400 windows...\n",
      "Processed 100 out of 400 windows...\n",
      "Processed 150 out of 400 windows...\n",
      "Processed 200 out of 400 windows...\n",
      "Processed 250 out of 400 windows...\n",
      "Processed 300 out of 400 windows...\n",
      "Processed 350 out of 400 windows...\n",
      "Processed 400 out of 400 windows...\n",
      "Total processed windows:400\n",
      "Number of reads: 786004\n",
      "Number of valid reads: 786004\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 786004\n",
      "Num mapped first of pair: 393002\n",
      "Num mapped second of pair: 393002\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 8\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 105562664\n",
      "referenceSize: 1139569\n",
      "numberOfSequencedBases: 105559115\n",
      "numberOfAs: 24800265\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 8\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing PDF report...\n",
      "PDF file created successfully \n",
      "\n",
      "Finished\n",
      "Java memory size is set to 1200M\n",
      "Launching application...\n",
      "\n",
      "QualiMap v.2.2.2-dev\n",
      "Built on 2017-08-28 08:37\n",
      "\n",
      "Selected tool: bamqc\n",
      "Available memory (Mb): 32\n",
      "Max memory (Mb): 1118\n",
      "Wed Jun 19 03:22:58 CEST 2019\t\tWARNING\tOutput folder already exists, the results will be saved there\n",
      "\n",
      "Starting bam qc....\n",
      "Loading sam header...\n",
      "Loading locator...\n",
      "Loading reference...\n",
      "Number of windows: 400, effective number of windows: 400\n",
      "Chunk of reads size: 1000\n",
      "Number of threads: 12\n",
      "Processed 50 out of 400 windows...\n",
      "Processed 100 out of 400 windows...\n",
      "Processed 150 out of 400 windows...\n",
      "Processed 200 out of 400 windows...\n",
      "Processed 250 out of 400 windows...\n",
      "Processed 300 out of 400 windows...\n",
      "Processed 350 out of 400 windows...\n",
      "Processed 400 out of 400 windows...\n",
      "Total processed windows:400\n",
      "Number of reads: 786004\n",
      "Number of valid reads: 786004\n",
      "Number of correct strand reads:0\n",
      "\n",
      "Inside of regions...\n",
      "Num mapped reads: 786004\n",
      "Num mapped first of pair: 393002\n",
      "Num mapped second of pair: 393002\n",
      "Num singletons: 0\n",
      "Time taken to analyze reads: 8\n",
      "Computing descriptors...\n",
      "numberOfMappedBases: 105562664\n",
      "referenceSize: 1139569\n",
      "numberOfSequencedBases: 105559115\n",
      "numberOfAs: 24800265\n",
      "Computing per chromosome statistics...\n",
      "Computing histograms...\n",
      "Overall analysis time: 8\n",
      "end of bam qc\n",
      "Computing report...\n",
      "Writing HTML report...\n",
      "HTML report created successfully\n",
      "\n",
      "Finished\n",
      "(treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:45: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  configs = yaml.load(f)\n",
      "/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/lib/python2.7/site-packages/multiqc/utils/config.py:51: YAMLLoadWarning: calling yaml.load() without Loader=... is deprecated, as the default Loader is unsafe. Please read https://msg.pyyaml.org/load for full details.\n",
      "  sp = yaml.load(f)\n",
      "[INFO   ]         multiqc : This is MultiQC v1.7\n",
      "[INFO   ]         multiqc : Template    : default\n",
      "[INFO   ]         multiqc : Searching '/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/qualimap/'\n",
      "\u001b[?25lSearching 50 files..  [####################################]  100%          \u001b[?25h\n",
      "[INFO   ]        qualimap : Found 1 BamQC reports\n",
      "[INFO   ]         multiqc : Compressing plot data\n",
      "[INFO   ]         multiqc : Report      : ../../qc/qualimap/multiqc_report.html\n",
      "[INFO   ]         multiqc : Data        : ../../qc/qualimap/multiqc_data\n",
      "[INFO   ]         multiqc : MultiQC complete\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) [Wed Jun 19 03:23:17 CEST 2019] picard.sam.CreateSequenceDictionary REFERENCE=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa OUTPUT=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.dict    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 19 03:23:17 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "[Wed Jun 19 03:23:17 CEST 2019] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.00 minutes.\n",
      "Runtime.totalMemory()=514850816\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Going to process files CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "(treponema) (.env) (treponema) (.env) INFO  03:23:20,262 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  03:23:20,264 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.7-0-gcfedb67, Compiled 2016/12/12 11:21:18 \n",
      "INFO  03:23:20,264 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  03:23:20,265 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  03:23:20,265 HelpFormatter - [Wed Jun 19 03:23:20 CEST 2019] Executing on Linux 4.4.0-148-generic amd64 \n",
      "INFO  03:23:20,265 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_192-b01 \n",
      "INFO  03:23:20,268 HelpFormatter - Program Args: -T DepthOfCoverage -R /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa -I /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/filtered/input_bams.list -o /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/coverage/SS14.coverage -ct 3 -ct 5 -ct 10 \n",
      "INFO  03:23:20,277 HelpFormatter - Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01. \n",
      "INFO  03:23:20,277 HelpFormatter - Date/Time: 2019/06/19 03:23:20 \n",
      "INFO  03:23:20,277 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  03:23:20,277 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  03:23:20,287 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  03:23:20,381 GenomeAnalysisEngine - Downsampling Settings: No downsampling \n",
      "INFO  03:23:20,389 SAMDataSource$SAMReaders - Initializing SAMRecords in serial \n",
      "INFO  03:23:20,413 SAMDataSource$SAMReaders - Done initializing BAM readers: total time 0.02 \n",
      "INFO  03:23:20,602 GenomeAnalysisEngine - Preparing for traversal over 1 BAM files \n",
      "INFO  03:23:20,640 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  03:23:20,640 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  03:23:20,641 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  03:23:20,641 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  03:23:50,648 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:354865    344064.0    30.0 s      87.0 s       31.1%    96.0 s      66.0 s \n",
      "INFO  03:24:20,651 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:745681    737280.0    60.0 s      81.0 s       65.4%    91.0 s      31.0 s \n",
      "INFO  03:24:50,654 ProgressMeter - gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:1097729   1097728.0    90.0 s      81.0 s       96.3%    93.0 s       3.0 s \n",
      "INFO  03:24:53,574 DepthOfCoverage - Printing summary info \n",
      "INFO  03:24:53,592 DepthOfCoverage - Printing locus summary \n",
      "INFO  03:24:53,628 ProgressMeter -            done   1139569.0    92.0 s      81.0 s      100.0%    92.0 s       0.0 s \n",
      "INFO  03:24:53,629 ProgressMeter - Total runtime 92.99 secs, 1.55 min, 0.03 hours \n",
      "INFO  03:24:53,636 MicroScheduler - 0 reads were filtered out during the traversal out of approximately 784198 total reads (0.00%) \n",
      "INFO  03:24:53,636 MicroScheduler -   -> 0 reads (0.00% of total) failing BadCigarFilter \n",
      "INFO  03:24:53,637 MicroScheduler -   -> 0 reads (0.00% of total) failing DuplicateReadFilter \n",
      "INFO  03:24:53,637 MicroScheduler -   -> 0 reads (0.00% of total) failing FailsVendorQualityCheckFilter \n",
      "INFO  03:24:53,637 MicroScheduler -   -> 0 reads (0.00% of total) failing MalformedReadFilter \n",
      "INFO  03:24:53,637 MicroScheduler -   -> 0 reads (0.00% of total) failing NotPrimaryAlignmentFilter \n",
      "INFO  03:24:53,638 MicroScheduler -   -> 0 reads (0.00% of total) failing UnmappedReadFilter \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were no warn messages.\n",
      "------------------------------------------------------------------------------------------\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Run mapping QC\n",
    "COVER_INT=\"-ct 3 -ct 5 -ct 10\" # Intervals of coverage to make the statistics for, can be multiple values; --summaryCoverageThreshold\n",
    "\n",
    "cd $OUTPUT_DIR/alignment/filtered\n",
    "\n",
    "# Get general alignment statistics\n",
    "mkdir $OUTPUT_DIR/qc/qualimap\n",
    "\n",
    "for sample in *.MAPQ${MAPQ_FINAL}.bam\n",
    "do\n",
    "    # PDF version is better for browsing\n",
    "    qualimap bamqc -bam $sample -nt $THREADS -c -outformat PDF -outdir $OUTPUT_DIR/qc/qualimap -outfile ${sample%.*}.qualimap.pdf\n",
    "    mv $OUTPUT_DIR/qc/qualimap/genome_results.txt $OUTPUT_DIR/qc/qualimap/${sample%.*}.genome_results.txt\n",
    "    # HTML version is necessary for multiQC\n",
    "    mkdir -p $OUTPUT_DIR/qc/qualimap/html/${sample%.*}\n",
    "    qualimap bamqc -bam $sample -nt $THREADS -c -outformat HTML -outdir $OUTPUT_DIR/qc/qualimap/html/${sample%.*} # Has to be redirected to separate folder for each file\n",
    "done\n",
    "\n",
    "multiqc -o $OUTPUT_DIR/qc/qualimap $OUTPUT_DIR/qc/qualimap/\n",
    "\n",
    "# Calculate the coverage\n",
    "mkdir -p $OUTPUT_DIR/qc/coverage\n",
    "\n",
    "# Prepare indexes\n",
    "samtools faidx $REFERENCE\n",
    "rm ${REFERENCE%.*}.dict\n",
    "picard CreateSequenceDictionary R=$REFERENCE O=${REFERENCE%.*}.dict\n",
    "\n",
    "# Run GATK DepthOfCoverage\n",
    "ls *.MAPQ40.bam | tr ' ' '\\n' > $OUTPUT_DIR/alignment/filtered//input_bams.list\n",
    "\n",
    "echo \"Going to process files\" `cat $OUTPUT_DIR/alignment/filtered//input_bams.list`\n",
    "\n",
    "gatk \\\n",
    "-T DepthOfCoverage \\\n",
    "-R $REFERENCE \\\n",
    "-I $OUTPUT_DIR/alignment/filtered/input_bams.list \\\n",
    "-o $OUTPUT_DIR/qc/coverage/$(basename $REFERENCE .fa)${INPUT_BAM%.bam}.coverage \\\n",
    "$COVER_INT # --outputFormat csv # Default is readable table (rtable)\n",
    "\n",
    "rm $OUTPUT_DIR/alignment/filtered/input_bams.list\n",
    "\n",
    "pigz -p $THREADS $OUTPUT_DIR/qc/coverage/*.coverage\n",
    "\n",
    "# Count number of mapped reads - we have the same information in Qualimap report\n",
    "echo \"Number of mapped read pairs (-F 3852 -f 2 -q 40 = read mapped in proper pair ; NOT read unmapped, mate unmapped not primary alignment, read fails platform/vendor quality checks, read is PCR or optical duplicate, supplementary alignment)\" > $OUTPUT_DIR/qc/coverage/num_reads.txt\n",
    "\n",
    "for sample in *.MAPQ40.bam\n",
    "do\n",
    "    echo -ne $sample ' '\n",
    "    samtools view -@ $THREADS -F 3852 -f 2 -q 40 $sample | cut -f 1 | sort -T $OUTPUT_DIR/tmp | uniq | wc -l\n",
    "done >> $OUTPUT_DIR/qc/coverage/num_reads.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "\n",
    "### Alignment consensus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "In this phase, we can generate the mapping consensus. Please not this steps is closely related to the previous step as it will take only the filtered mappings into consideration. Here, we use a 'simple' variant call performed by `samtools`&`bcftools` and `GATK`. You might consider alternating this step with more sophisticated variant call if you think it is necessary. The difference is that the \"samtools\" version gives you lower-case letters where coverage was low and *N* where there was no coverage. Here, we convert all low-coverage positions to *n* (lower-case). The \"GAKT\" version copies the reference sequence where is no coverage. \"samtools\" version doesn't like to include indels whereas \"GATK\" version should be OK with it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Fix reference name and SAM header (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "`bcftools` and `vcfutils` might have problem with \".\" or any \"strange\" symbol in the reference name. If you haven't \"fix\" your reference name you can do it here but remember you have to replace it in the SAM/BAM header as well. Or go back to the beginning and redo the whole analysis (joke)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Replace the reference name\n",
    "sed -i 's/gi_511533127_gb_CP004011\\.1__Treponema_pallidum_subsp\\._pallidum_SS14,_complete_genome/gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome/g' $REFERENCE \n",
    "\n",
    "# Replace the SAM header\n",
    "for sample in *.filt.indelRealigned.dedup.MAPQ40.bam\n",
    "do\n",
    "    samtools view -@ $THREADS -H $sample > $sample.header.sam  # extract header only\n",
    "\n",
    "    sed -i 's/gi_511533127_gb_CP004011\\.1__Treponema_pallidum_subsp\\._pallidum_SS14,_complete_genome/gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome/g' $sample.header.sam\n",
    "\n",
    "    samtools reheader $sample.header.sam $sample > $sample.tmp\n",
    "    mv $sample.tmp $sample\n",
    "\n",
    "    rm $sample.header.sam\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### Mapping consensus ~ alignment-guided assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) [Tue Jun 25 23:15:10 CEST 2019] picard.sam.CreateSequenceDictionary REFERENCE=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa OUTPUT=/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.dict    TRUNCATE_NAMES_AT_WHITESPACE=true NUM_SEQUENCES=2147483647 VERBOSITY=INFO QUIET=false VALIDATION_STRINGENCY=STRICT COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Tue Jun 25 23:15:10 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "[Tue Jun 25 23:15:10 CEST 2019] picard.sam.CreateSequenceDictionary done. Elapsed time: 0.01 minutes.\n",
      "Runtime.totalMemory()=514850816\n",
      "(treponema) (.env) (treponema) (.env) [mpileup] 1 samples in 1 input files\n",
      "INFO  23:18:27,845 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  23:18:27,848 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.7-0-gcfedb67, Compiled 2016/12/12 11:21:18 \n",
      "INFO  23:18:27,849 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  23:18:27,849 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  23:18:27,849 HelpFormatter - [Tue Jun 25 23:18:27 CEST 2019] Executing on Linux 4.4.0-148-generic amd64 \n",
      "INFO  23:18:27,849 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_192-b01 \n",
      "INFO  23:18:27,852 HelpFormatter - Program Args: -T FastaAlternateReferenceMaker -R /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa -o /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/consensus/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.cns.gatk.fasta --variant /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/consensus/other/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.vcf \n",
      "INFO  23:18:27,864 HelpFormatter - Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01. \n",
      "INFO  23:18:27,864 HelpFormatter - Date/Time: 2019/06/25 23:18:27 \n",
      "INFO  23:18:27,865 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  23:18:27,865 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  23:18:27,874 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  23:18:27,968 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  23:18:36,244 RMDTrackBuilder - Writing Tribble index to disk for file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/consensus/other/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.vcf.idx \n",
      "INFO  23:18:37,082 GenomeAnalysisEngine - Preparing for traversal \n",
      "INFO  23:18:37,083 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  23:18:37,084 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  23:18:37,084 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  23:18:37,085 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  23:18:45,809 ProgressMeter -            done   1139569.0     8.0 s       7.0 s      100.0%     8.0 s       0.0 s \n",
      "INFO  23:18:45,810 ProgressMeter - Total runtime 8.73 secs, 0.15 min, 0.00 hours \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were no warn messages.\n",
      "------------------------------------------------------------------------------------------\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "MIN_COVER=3 # All bases bellow this coverage are converted to ns\n",
    "PLOIDY=1 # Ploidy for the bcftools\n",
    "\n",
    "mkdir -p $OUTPUT_DIR/consensus/other\n",
    "\n",
    "# Make sure we have reference index and dictionary (! have to redo if you used the previous \"fix\" step)\n",
    "samtools faidx $REFERENCE\n",
    "rm ${REFERENCE%.*}.dict\n",
    "picard CreateSequenceDictionary R=$REFERENCE O=${REFERENCE%.*}.dict\n",
    "\n",
    "for sample in *.filt.indelRealigned.dedup.MAPQ40.bam\n",
    "do\n",
    "    samtools index -@ $THREADS $sample\n",
    "    samtools mpileup --max-depth 10000 -E --min-BQ 13 --fasta-ref $REFERENCE --min-MQ 40 -g -u --max-idepth 100000 --min-ireads 5 --gap-frac 0.002 --excl-flags UNMAP,SECONDARY,QCFAIL,DUP --output-tags DP,AD,ADF,ADR,SP,INFO/AD,INFO/ADF,INFO/ADR --reference $REFERENCE $sample | bcftools call --ploidy $PLOIDY -c --keep-masked-ref --output-type v --threads $THREADS - > $OUTPUT_DIR/consensus/other/${sample%.*}.vcf\n",
    "\n",
    "    cat $OUTPUT_DIR/consensus/other/${sample%.*}.vcf | vcfutils.pl vcf2fq -d $MIN_COVER | seqtk seq -A - > $OUTPUT_DIR/consensus/${sample%.*}.cns.def.fasta\n",
    "\n",
    "    # Convert all low-coverage (=lower-case) positions to \"n\"\n",
    "    sed -e '/^>/! s/[[:lower:]]/n/g' $OUTPUT_DIR/consensus/${sample%.*}.cns.def.fasta > $OUTPUT_DIR/consensus/${sample%.*}.cns.final.fasta\n",
    "\n",
    "    gatk -T FastaAlternateReferenceMaker -R $REFERENCE -o $OUTPUT_DIR/consensus/${sample%.*}.cns.gatk.fasta --variant $OUTPUT_DIR/consensus/other/${sample%.*}.vcf\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### BAM downsampling (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "In some cases, our resulting SAM/BAM is very deep which causes problems during the subsequent analysis. An option to lower down the computational demands is to downsample to a specific depth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) INFO\t2019-06-26 00:05:24\tSortingCollection\tCreating merging iterator from 16 files\n",
      "[INFO][Biostar154220]Alloc memory for contig gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome N=1139569*sizeof(int)\n",
      "[INFO][Biostar154220]Last M70401:55:000000000-B24PV:1:2104:14484:21063;gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome;339,874 11 seconds\n",
      "[INFO][Biostar154220]. Completed. N=1,179,007. That took:18 seconds\n",
      "[INFO][SortSamRefName]done: N=786004\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "MAX_COVER=250 # Set maximum coverage to put the limit on\n",
    "\n",
    "for sample in *.filt.indelRealigned.dedup.MAPQ40.bam\n",
    "do\n",
    "    java -jar -Xmx4g $CONDA_PREFIX/bin/sortsamrefname.jar --tmpDir $OUTPUT_DIR/tmp $sample |  java -jar -Xmx4g $CONDA_PREFIX/bin/downsamplebam.jar -n $MAX_COVER | samtools sort -@ $THREADS - > ${sample%.bam*}.downsamp${MAX_COVER}.bam\n",
    "    samtools index -@ $THREADS ${sample%.bam*}.downsamp${MAX_COVER}.bam\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Variant call"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "To get the most relevant variants directly from the alignments we could either use the variants used for the consensus generation or we can use probably more suitable tools for bacteria variant call such as [`freebayes`](https://github.com/ekg/freebayes). The reason why we don't use other tools than `samtools`/`bcftools` for the generation of consensus is the compatibility of the approach. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Processing CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.bam\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Processing  CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.full.vcf\n",
      "INFO  01:18:51,060 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  01:18:51,063 HelpFormatter - The Genome Analysis Toolkit (GATK) v3.7-0-gcfedb67, Compiled 2016/12/12 11:21:18 \n",
      "INFO  01:18:51,063 HelpFormatter - Copyright (c) 2010-2016 The Broad Institute \n",
      "INFO  01:18:51,063 HelpFormatter - For support and documentation go to https://software.broadinstitute.org/gatk \n",
      "INFO  01:18:51,063 HelpFormatter - [Wed Jun 26 01:18:51 CEST 2019] Executing on Linux 4.4.0-148-generic amd64 \n",
      "INFO  01:18:51,064 HelpFormatter - OpenJDK 64-Bit Server VM 1.8.0_192-b01 \n",
      "INFO  01:18:51,067 HelpFormatter - Program Args: -T VariantFiltration -R /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa -o /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/variants/freebayes/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.filt.vcf --variant:VCF CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.full.vcf --filterExpression DP < 3 --filterName LowCoverage --filterExpression QUAL > 0 && QUAL < 20 --filterName VeryLowQual --filterExpression QUAL > 20 && QUAL < 50 --filterName LowQual --filterExpression QUAL == 0 --filterName TechnicalQual \n",
      "INFO  01:18:51,079 HelpFormatter - Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01. \n",
      "INFO  01:18:51,079 HelpFormatter - Date/Time: 2019/06/26 01:18:51 \n",
      "INFO  01:18:51,079 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  01:18:51,080 HelpFormatter - -------------------------------------------------------------------------------- \n",
      "INFO  01:18:51,097 GenomeAnalysisEngine - Strictness is SILENT \n",
      "INFO  01:18:51,186 GenomeAnalysisEngine - Downsampling Settings: Method: BY_SAMPLE, Target Coverage: 1000 \n",
      "INFO  01:18:51,588 RMDTrackBuilder - Writing Tribble index to disk for file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/variants/freebayes/CW56_S1_.SS14.filt.indelRealigned.dedup.MAPQ40.full.vcf.idx \n",
      "INFO  01:18:51,760 GenomeAnalysisEngine - Preparing for traversal \n",
      "INFO  01:18:51,761 GenomeAnalysisEngine - Done preparing for traversal \n",
      "INFO  01:18:51,761 ProgressMeter - [INITIALIZATION COMPLETE; STARTING PROCESSING] \n",
      "INFO  01:18:51,762 ProgressMeter -                 | processed |    time |    per 1M |           |   total | remaining \n",
      "INFO  01:18:51,762 ProgressMeter -        Location |     sites | elapsed |     sites | completed | runtime |   runtime \n",
      "INFO  01:18:52,779 ProgressMeter -            done      7677.0     1.0 s       2.2 m       99.9%     1.0 s       0.0 s \n",
      "INFO  01:18:52,780 ProgressMeter - Total runtime 1.02 secs, 0.02 min, 0.00 hours \n",
      "------------------------------------------------------------------------------------------\n",
      "Done. There were no warn messages.\n",
      "------------------------------------------------------------------------------------------\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "MIN_DP=$MIN_COVER # Set minimal coverage depth of the mapping\n",
    "QUAL_SET=50 # Set minimal variant quality for the most certain variants; default it 20\n",
    "\n",
    "mkdir -p $OUTPUT_DIR/variants/freebayes/stats\n",
    "\n",
    "for sample in *filt.indelRealigned.dedup.MAPQ40.bam\n",
    "do\n",
    "    echo \"Processing $sample\"\n",
    "    \n",
    "    samtools index $sample\n",
    "\n",
    "    freebayes --bam $sample --vcf $OUTPUT_DIR/variants/freebayes/${sample%.*}.full.vcf --fasta-reference $REFERENCE --theta 0.001 --ploidy $PLOIDY --min-mapping-quality 10 --min-base-quality 15 --min-alternate-fraction 0.01 --min-alternate-count 2 --genotype-qualities --report-all-haplotype-alleles # You can add \"--report-monomorphic\" to get all positions\n",
    "done\n",
    "\n",
    "# Variant post-processing and filtering\n",
    "cd $OUTPUT_DIR/variants/freebayes/\n",
    "\n",
    "for sample in *full.vcf\n",
    "do\n",
    "    echo \"Processing \" $sample\n",
    "\n",
    "    # GATK filtering\n",
    "    gatk -T VariantFiltration -R $REFERENCE -o $OUTPUT_DIR/variants/freebayes/${sample%.full*}.filt.vcf --variant:VCF $sample --filterExpression \"DP < ${MIN_DP}\" --filterName \"LowCoverage\" --filterExpression \"QUAL > 0 && QUAL < 20\" --filterName \"VeryLowQual\" --filterExpression \"QUAL > 20 && QUAL < ${QUAL_SET}\" --filterName \"LowQual\" --filterExpression \"QUAL == 0\" --filterName \"TechnicalQual\" # This might be too strict for FreeBayes\n",
    "    # gatk -T VariantFiltration -R $REFERENCE -o ${sample%.full*}.filt.vcf --variant:VCF $sample --filterExpression \"DP < ${MIN_DP}\" --filterName \"LowCoverage\" --filterExpression \"QUAL > 0 && QUAL < 20\" --filterName \"VeryLowQual\" --filterExpression \"QUAL == 0\" --filterName \"TechnicalQual\" # This would be the way FreeBayes author recommends to use the filtering\n",
    "\n",
    "    mv ${sample%.full*}.filt.vcf $sample\n",
    "    rm ${sample%.full*}.filt.vcf.idx\n",
    "    \n",
    "    # Get only variants (PASS filtering)\n",
    "    grep \"^#\" $sample > ${sample%.full*}.var.vcf # Get only header\n",
    "    awk '$7 == \"PASS\" {print $0}' $sample >> ${sample%.full*}.var.vcf\n",
    "\n",
    "    # Get only SNPs and/or indels\n",
    "#    vcffilter -f \"TYPE = snp\" ${sample%.full*}.var.vc > ${sample%.full*}.var.snp.vcf # Get only SNP\n",
    "#    vcffilter -f \"( TYPE = ins | TYPE = del )\" ${sample%.full*}.var.vcf > ${sample%.full*}.var.indel.vcf # Get only indels\n",
    "\n",
    "    # Convert vct to tab vcftools\n",
    "    cat ${sample%.full*}.var.vcf | vcf-to-tab > ${sample%.full*}.var.tsv\n",
    "\n",
    "    # Get stats of the variant calling\n",
    "    # Stats from raw and filtered SNPs\n",
    "    vcfstats $sample > $OUTPUT_DIR/variants/freebayes/stats/${sample%.vcf}.stats.txt\n",
    "    vcfstats ${sample%.full*}.var.vcf > $OUTPUT_DIR/variants/freebayes/stats/${sample%.full*}.var.stats.txt\n",
    "\n",
    "    # Compare ts/tv ration between high/low quality variants\n",
    "    echo $sample > $OUTPUT_DIR/variants/freebayes/stats/${sample%.vcf}.ratio.stats.txt\n",
    "    echo \"Low quality variants\" >> $OUTPUT_DIR/variants/freebayes/stats/${sample%.*}.ratio.stats.txt # Low quality variants\n",
    "    vcffilter -f \"QUAL < 20\" $sample | vcfstats | grep \"ts\\|bial\" >> $OUTPUT_DIR/variants/freebayes/stats/${sample%.vcf}.ratio.stats.txt\n",
    "    echo \"High quality variants\" >> $OUTPUT_DIR/variants/freebayes/stats/${sample%.vcf}.ratio.stats.txt # High quality variants\n",
    "    vcffilter -f \"QUAL > 20\" $sample | vcfstats | grep \"ts\\|bial\" >> $OUTPUT_DIR/variants/freebayes/stats/${sample%.vcf}.ratio.stats.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### *De novo* assembly - mapped reads"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "To accompany the alignment-guided assembly and to balance for the potentical pitfalls of that approach we strongly recommend to run *de novo* assembly as well. You can run *de novo* assembly on the host-cleaned data but since we assume there could be a mixture of organisms present we recommend to run the assembly only from the initial mapping step (\"rough mapping\"). "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### SAM/BAM to fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "In the preparation step, we first extract the mapped reads from the SAM/BAM."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) [Wed Jun 26 20:22:23 CEST 2019] picard.sam.FixMateInformation INPUT=[CW56_S1_.SS14.mapped.bam] OUTPUT=tmp.bam VALIDATION_STRINGENCY=LENIENT    ASSUME_SORTED=false ADD_MATE_CIGAR=true IGNORE_MISSING_MATES=true VERBOSITY=INFO QUIET=false COMPRESSION_LEVEL=5 MAX_RECORDS_IN_RAM=500000 CREATE_INDEX=false CREATE_MD5_FILE=false GA4GH_CLIENT_SECRETS=client_secrets.json\n",
      "[Wed Jun 26 20:22:23 CEST 2019] Executing as 325073@BioDA-server on Linux 4.4.0-148-generic amd64; OpenJDK 64-Bit Server VM 1.8.0_192-b01; Picard version: 2.9.2-SNAPSHOT\n",
      "INFO\t2019-06-26 20:22:23\tFixMateInformation\tSorting input into queryname order.\n",
      "INFO\t2019-06-26 20:23:12\tFixMateInformation\tSorting by queryname complete.\n",
      "INFO\t2019-06-26 20:23:12\tFixMateInformation\tOutput will be sorted by coordinate\n",
      "INFO\t2019-06-26 20:23:12\tFixMateInformation\tTraversing query name sorted records and fixing up mate pair information.\n",
      "INFO\t2019-06-26 20:23:22\tFixMateInformation\tProcessed     1,000,000 records.  Elapsed time: 00:00:09s.  Time for last 1,000,000:    9s.  Last read position: gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:50,042\n",
      "INFO\t2019-06-26 20:23:35\tFixMateInformation\tProcessed     2,000,000 records.  Elapsed time: 00:00:22s.  Time for last 1,000,000:   13s.  Last read position: gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:709,888\n",
      "INFO\t2019-06-26 20:23:48\tFixMateInformation\tProcessed     3,000,000 records.  Elapsed time: 00:00:35s.  Time for last 1,000,000:   12s.  Last read position: gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome:228,435\n",
      "INFO\t2019-06-26 20:24:00\tFixMateInformation\tFinished processing reads; re-sorting output file.\n",
      "[Wed Jun 26 20:24:40 CEST 2019] picard.sam.FixMateInformation done. Elapsed time: 2.28 minutes.\n",
      "Runtime.totalMemory()=935329792\n",
      "[M::bam2fq_mainloop_singletontrack] discarded 0 singletons\n",
      "[M::bam2fq_mainloop_singletontrack] processed 3960618 reads\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir -p $OUTPUT_DIR/alignment/fastq\n",
    "\n",
    "cd $OUTPUT_DIR/alignment\n",
    "\n",
    "for sample in *.$(basename $REFERENCE .fa).bam\n",
    "do\n",
    "    echo \"Processing file $sample\"\n",
    "\n",
    "    samtools view -H $sample > ${sample%.*}.sam\n",
    "\n",
    "    samtools view -@ $THREADS -F 4 -f 3 $sample  >> ${sample%.*}.sam\n",
    "    samtools view -@ $THREADS -b ${sample%.*}.sam | samtools sort -@ $THREADS -o ${sample%.*}.mapped.bam -\n",
    "    rm ${sample%.*}.sam\n",
    "\n",
    "    picard FixMateInformation I=${sample%.*}.mapped.bam O=tmp.bam VALIDATION_STRINGENCY=LENIENT\n",
    "    mv tmp.bam ${sample%.*}.mapped.bam\n",
    "    samtools index -@ $THREADS ${sample%.*}.mapped.bam\n",
    "\n",
    "    samtools collate -O -@ $THREADS ${sample%.*}.mapped.bam - | samtools fastq -@ $THREADS -s $OUTPUT_DIR/alignment/fastq/${sample%.*}_unpaired.mapped.fastq -1 $OUTPUT_DIR/alignment/fastq/${sample%.*}_R1.mapped.fastq -2 $OUTPUT_DIR/alignment/fastq/${sample%.*}_R2.mapped.fastq -\n",
    "done\n",
    "\n",
    "cd $OUTPUT_DIR/alignment/fastq/\n",
    "\n",
    "rm *_unpaired.mapped.fastq\n",
    "\n",
    "for sample in *.fastq\n",
    "do\n",
    "    pigz -p $THREADS $sample\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "#### *De novo* assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Now we can do the *de novo* assembly itself. In this section, we are using SPAdes but you can consider using [Unicycler](https://github.com/rrwick/Unicycler) or [metaSPADES](http://cab.spbu.ru/software/meta-spades/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) Command line: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/spades.py\t--careful\t--threads\t12\t--cov-cutoff\t5\t--tmp-dir\t/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp\t-k\t15,21,27,33,55,77,99,127\t-1\t/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz\t-2\t/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz\t-o\t/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_\t\n",
      "\n",
      "System information:\n",
      "  SPAdes version: 3.13.0\n",
      "  Python version: 2.7.15\n",
      "  OS: Linux-4.4.0-148-generic-x86_64-with-debian-stretch-sid\n",
      "\n",
      "Output dir: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_\n",
      "Mode: read error correction and assembling\n",
      "Debug mode is turned OFF\n",
      "\n",
      "Dataset parameters:\n",
      "  Multi-cell mode (you should set '--sc' flag if input data was obtained with MDA (single-cell) technology or --meta flag if processing metagenomic dataset)\n",
      "  Reads:\n",
      "    Library number: 1, library type: paired-end\n",
      "      orientation: fr\n",
      "      left reads: ['/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz']\n",
      "      right reads: ['/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz']\n",
      "      interlaced reads: not specified\n",
      "      single reads: not specified\n",
      "      merged reads: not specified\n",
      "Read error correction parameters:\n",
      "  Iterations: 1\n",
      "  PHRED offset will be auto-detected\n",
      "  Corrected reads will be compressed\n",
      "Assembly parameters:\n",
      "  k: [15, 21, 27, 33, 55, 77, 99, 127]\n",
      "  Repeat resolution is enabled\n",
      "  Mismatch careful mode is turned ON\n",
      "  MismatchCorrector will be used\n",
      "  Coverage cutoff is turned ON and threshold is 5.0\n",
      "Other parameters:\n",
      "  Dir for temp files: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp\n",
      "  Threads: 12\n",
      "  Memory limit (in Gb): 246\n",
      "\n",
      "\n",
      "======= SPAdes pipeline started. Log can be found here: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/spades.log\n",
      "\n",
      "\n",
      "===== Read error correction started. \n",
      "\n",
      "\n",
      "== Running read error correction tool: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/spades-3.13.0-0/bin/spades-hammer /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/corrected/configs/config.info\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  75)   Starting BayesHammer, built from refs/heads/spades_3.13.0, git revision 8ea46659e9b2aca35444a808db550ac333006f8b\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  76)   Loading config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/corrected/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  78)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 246 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  86)   Trying to determine PHRED offset\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  :  92)   Determined value is 33\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (hammer_tools.cpp          :  36)   Hamming graph threshold tau=1, k=21, subkmer positions = [ 0 10 ]\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (main.cpp                  : 113)   Size of aux. kmer data 24 bytes\n",
      "     === ITERATION 0 begins ===\n",
      "  0:00:00.001     4M / 4M    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83322 Gb\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 349525\n",
      "  0:00:00.032     6G / 6G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz\n",
      "  0:00:12.634     6G / 6G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 1980309 reads\n",
      "  0:00:12.634     6G / 6G    INFO   K-mer Splitting          (kmer_data.cpp             :  97)   Processing /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz\n",
      "  0:00:26.539     6G / 6G    INFO   K-mer Splitting          (kmer_data.cpp             : 107)   Processed 3960618 reads\n",
      "  0:00:26.539     6G / 6G    INFO   K-mer Splitting          (kmer_data.cpp             : 112)   Total 3960618 reads processed\n",
      "  0:00:27.099    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:27.810    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 40437878 kmers in total.\n",
      "  0:00:27.810    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:28.273    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:30.041    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:30.504    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 18759336 bytes occupied (3.71124 bits per kmer).\n",
      "  0:00:30.505    48M / 6G    INFO   K-mer Counting           (kmer_data.cpp             : 356)   Arranging kmers in hash map order\n",
      "  0:00:32.285   672M / 6G    INFO    General                 (main.cpp                  : 148)   Clustering Hamming graph.\n",
      "  0:01:32.792   672M / 6G    INFO    General                 (main.cpp                  : 155)   Extracting clusters\n",
      "  0:02:00.698   672M / 6G    INFO    General                 (main.cpp                  : 167)   Clustering done. Total clusters: 11976802\n",
      "  0:02:00.738   360M / 6G    INFO   K-mer Counting           (kmer_data.cpp             : 376)   Collecting K-mer information, this takes a while.\n",
      "  0:02:01.289     1G / 6G    INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz\n",
      "  0:02:43.681     1G / 6G    INFO   K-mer Counting           (kmer_data.cpp             : 382)   Processing /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz\n",
      "  0:03:25.794     1G / 6G    INFO   K-mer Counting           (kmer_data.cpp             : 389)   Collection done, postprocessing.\n",
      "  0:03:26.021     1G / 6G    INFO   K-mer Counting           (kmer_data.cpp             : 403)   There are 40437878 kmers in total. Among them 33670032 (83.2636%) are singletons.\n",
      "  0:03:26.021     1G / 6G    INFO    General                 (main.cpp                  : 173)   Subclustering Hamming graph\n",
      "  0:03:30.358     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 649)   Subclustering done. Total 31 non-read kmers were generated.\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 650)   Subclustering statistics:\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 651)     Total singleton hamming clusters: 9666128. Among them 232870 (2.40913%) are good\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 652)     Total singleton subclusters: 11691. Among them 11653 (99.675%) are good\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 653)     Total non-singleton subcluster centers: 2344160. Among them 2279949 (97.2608%) are good\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 654)     Average size of non-trivial subcluster: 13.127 kmers\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 655)     Average number of sub-clusters per non-singleton cluster: 1.01955\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 656)     Total solid k-mers: 2524472\n",
      "  0:03:30.359     1G / 6G    INFO   Hamming Subclustering    (kmer_cluster.cpp          : 657)     Substitution probabilities: [4,4]((0.939466,0.0301014,0.0177737,0.0126594),(0.0224659,0.954188,0.0103778,0.012968),(0.0129363,0.010343,0.954211,0.0225095),(0.0126257,0.0177377,0.0300562,0.93958))\n",
      "  0:03:30.374     1G / 6G    INFO    General                 (main.cpp                  : 178)   Finished clustering.\n",
      "  0:03:30.374     1G / 6G    INFO    General                 (main.cpp                  : 197)   Starting solid k-mers expansion in 12 threads.\n",
      "  0:04:02.852     1G / 6G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 0 produced 61120 new k-mers.\n",
      "  0:04:35.024     1G / 6G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 1 produced 5862 new k-mers.\n",
      "  0:05:07.125     1G / 6G    INFO    General                 (main.cpp                  : 218)   Solid k-mers iteration 2 produced 0 new k-mers.\n",
      "  0:05:07.125     1G / 6G    INFO    General                 (main.cpp                  : 222)   Solid k-mers finalized\n",
      "  0:05:07.125     1G / 6G    INFO    General                 (hammer_tools.cpp          : 220)   Starting read correction in 12 threads.\n",
      "  0:05:07.125     1G / 6G    INFO    General                 (hammer_tools.cpp          : 233)   Correcting pair of reads: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz and /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz\n",
      "  0:05:16.643     2G / 6G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 0 of 1200000 reads.\n",
      "  0:05:38.451     2G / 6G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 0\n",
      "  0:05:42.374     2G / 6G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 0\n",
      "  0:05:48.874     2G / 6G    INFO    General                 (hammer_tools.cpp          : 168)   Prepared batch 1 of 780309 reads.\n",
      "  0:06:02.846     2G / 6G    INFO    General                 (hammer_tools.cpp          : 175)   Processed batch 1\n",
      "  0:06:05.570     2G / 6G    INFO    General                 (hammer_tools.cpp          : 185)   Written batch 1\n",
      "  0:06:08.493     1G / 6G    INFO    General                 (hammer_tools.cpp          : 274)   Correction done. Changed 1398039 bases in 577859 reads.\n",
      "  0:06:08.493     1G / 6G    INFO    General                 (hammer_tools.cpp          : 275)   Failed to correct 0 bases out of 567644980.\n",
      "  0:06:08.664    48M / 6G    INFO    General                 (main.cpp                  : 255)   Saving corrected dataset description to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/corrected/corrected.yaml\n",
      "  0:06:08.664    48M / 6G    INFO    General                 (main.cpp                  : 262)   All done. Exiting.\n",
      "\n",
      "== Compressing corrected reads (with pigz)\n",
      "\n",
      "== Dataset description file was created: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/corrected/corrected.yaml\n",
      "\n",
      "\n",
      "===== Read error correction finished. \n",
      "\n",
      "\n",
      "===== Assembling started.\n",
      "\n",
      "\n",
      "== Running assembler: K15\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K15/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K15/configs/careful_mode.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 246 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from refs/heads/spades_3.13.0, git revision 8ea46659e9b2aca35444a808db550ac333006f8b\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/dataset.info) with K=15\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  77)   Converting reads to binary format for library #0 (takes a while)\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  78)   Converting paired reads\n",
      "  0:00:00.251   112M / 112M  INFO    General                 (binary_converter.hpp      :  93)   16384 reads processed\n",
      "  0:00:00.443   128M / 128M  INFO    General                 (binary_converter.hpp      :  93)   32768 reads processed\n",
      "  0:00:00.827   156M / 156M  INFO    General                 (binary_converter.hpp      :  93)   65536 reads processed\n",
      "  0:00:01.597   212M / 212M  INFO    General                 (binary_converter.hpp      :  93)   131072 reads processed\n",
      "  0:00:03.140   328M / 328M  INFO    General                 (binary_converter.hpp      :  93)   262144 reads processed\n",
      "  0:00:06.221   556M / 556M  INFO    General                 (binary_converter.hpp      :  93)   524288 reads processed\n",
      "  0:00:15.966   912M / 912M  INFO    General                 (binary_converter.hpp      :  93)   1048576 reads processed\n",
      "  0:00:32.206   196M / 912M  INFO    General                 (binary_converter.hpp      : 117)   1973109 reads written\n",
      "  0:00:32.400     4M / 912M  INFO    General                 (read_converter.hpp        :  87)   Converting single reads\n",
      "  0:00:32.601   152M / 912M  INFO    General                 (binary_converter.hpp      : 117)   7191 reads written\n",
      "  0:00:32.616     4M / 912M  INFO    General                 (read_converter.hpp        :  95)   Converting merged reads\n",
      "  0:00:32.764   148M / 912M  INFO    General                 (binary_converter.hpp      : 117)   0 reads written\n",
      "  0:00:32.776     4M / 912M  INFO    General                 (construction.cpp          : 111)   Max read length 151\n",
      "  0:00:32.776     4M / 912M  INFO    General                 (construction.cpp          : 117)   Average read length 143.316\n",
      "  0:00:32.776     4M / 912M  INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:32.776     4M / 912M  INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:00:32.776     4M / 912M  INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:32.776     4M / 912M  INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83322 Gb\n",
      "  0:00:32.776     4M / 912M  INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 466033\n",
      "  0:00:40.345     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 7906818 reads\n",
      "  0:00:40.345     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:00:40.830    48M / 6G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 7906818 reads\n",
      "  0:00:40.830    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:40.900    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2490520 kmers in total.\n",
      "  0:00:40.900    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:40.935    48M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:00:40.935    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:40.935    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:00:40.935    48M / 6G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:40.935    48M / 6G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83203 Gb\n",
      "  0:00:40.935    48M / 6G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 349525\n",
      "  0:00:41.153     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 2490520 kmers\n",
      "  0:00:41.154     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 2490520 kmers.\n",
      "  0:00:41.224    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:41.304    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2452827 kmers in total.\n",
      "  0:00:41.304    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:41.343    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:41.471    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:41.501    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1145648 bytes occupied (3.73658 bits per kmer).\n",
      "  0:00:41.504    52M / 6G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:00:41.636    52M / 6G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:41.637    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:00:41.637    52M / 6G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:00:41.637    52M / 6G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:00:41.877    52M / 6G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   684563 16-mers were removed by early tip clipper\n",
      "  0:00:41.877    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:00:41.878    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:00:41.994    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 104979 sequences extracted\n",
      "  0:00:42.066    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:00:42.093    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 0 loops collected\n",
      "  0:00:42.131    56M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:00:42.131    56M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:42.131    56M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:42.190    56M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1161152 bytes occupied (3.72983 bits per kmer).\n",
      "  0:00:42.196    68M / 6G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:00:52.596    68M / 6G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:00:52.696    68M / 6G    INFO    General                 (construction.cpp          : 464)   Processed 209816 edges\n",
      "  0:00:52.707    52M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:00:52.707    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 14\n",
      "  0:00:52.707    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 411\n",
      "  0:00:52.707    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 428. Coverage mad: 137.882\n",
      "  0:00:52.707    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:00:52.922    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:00:53.506    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:00:55.483    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:00:58.908    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 444.981. Fitted coverage std. dev: 134.983\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.998645\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 197\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 197\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 1104926\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 444.981\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 197\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 4.47668\n",
      "  0:00:58.918    52M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:00:58.918    52M / 6G    INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:00:58.918    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:00:58.930    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:00:58.930    52M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:00:58.930    52M / 6G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:00:58.931    52M / 6G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:00:58.931    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:00:58.931    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:58.950    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 462 times\n",
      "  0:00:58.951    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:00.659    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 17182 times\n",
      "  0:01:00.659    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.204    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 17337 times\n",
      "  0:01:01.204    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:01:01.204    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.218    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 544 times\n",
      "  0:01:01.218    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.234    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 104 times\n",
      "  0:01:01.234    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.238    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 162 times\n",
      "  0:01:01.238    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:01:01.238    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.238    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 6 times\n",
      "  0:01:01.238    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.239    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 5 times\n",
      "  0:01:01.239    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.240    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 51 times\n",
      "  0:01:01.241    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:01:01.241    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.241    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:01:01.241    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.241    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 3 times\n",
      "  0:01:01.241    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.242    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 32 times\n",
      "  0:01:01.242    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:01:01.242    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.242    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.242    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.242    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 6 times\n",
      "  0:01:01.242    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.243    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 25 times\n",
      "  0:01:01.243    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:01:01.243    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.243    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.243    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 3 times\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 16 times\n",
      "  0:01:01.244    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:01:01.244    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 15 times\n",
      "  0:01:01.245    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 19 times\n",
      "  0:01:01.245    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:01.245    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.246    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 27 times\n",
      "  0:01:01.246    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:01:01.246    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.246    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.246    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.246    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:01.246    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.247    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 41 times\n",
      "  0:01:01.247    56M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:01:01.247    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.248    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 6 times\n",
      "  0:01:01.248    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.261    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:01:01.261    56M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:01.263    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:01:01.263    52M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:01:01.263    52M / 6G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:01:01.263    52M / 6G    INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:01:01.263    52M / 6G    INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:01:01.263    52M / 6G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:01:01.263    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.264    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.264    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.277    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:01.277    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:01:01.277    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:01:01.277    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:01:01.289    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:01:01.289    52M / 6G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:01:01.292    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:01:01.293    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 2 times\n",
      "  0:01:01.293    52M / 6G    INFO    General                 (simplification.cpp        : 340)   Removing all the edges having coverage 4.47668 and less\n",
      "  0:01:01.294    52M / 6G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:01:01.295    52M / 6G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 449.916\n",
      "  0:01:01.295    52M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:01:01.295    52M / 6G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/assembly_graph_with_scaffolds.gfa\n",
      "  0:01:01.323    52M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/before_rr.fasta\n",
      "  0:01:01.365    52M / 6G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/assembly_graph.fastg\n",
      "  0:01:01.480    52M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/simplified_contigs.fasta\n",
      "  0:01:01.521    52M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/final_contigs.fasta\n",
      "  0:01:01.563    52M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:01:01.563    52M / 6G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/assembly_graph_with_scaffolds.gfa\n",
      "  0:01:01.590    52M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/before_rr.fasta\n",
      "  0:01:01.631    52M / 6G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/assembly_graph.fastg\n",
      "  0:01:01.745    52M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/simplified_contigs.fasta\n",
      "  0:01:01.786    52M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K15/final_contigs.fasta\n",
      "  0:01:01.828    52M / 6G    INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:01:01.860    48M / 6G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 1 minutes 1 seconds\n",
      "Max read length detected as 151\n",
      "\n",
      "== Running assembler: K21\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K21/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K21/configs/careful_mode.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 246 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from refs/heads/spades_3.13.0, git revision 8ea46659e9b2aca35444a808db550ac333006f8b\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/dataset.info) with K=21\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 151\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 143.316\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83322 Gb\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 466033\n",
      "  0:00:08.615     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 7906818 reads\n",
      "  0:00:08.615     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:00:09.300    48M / 6G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 7906818 reads\n",
      "  0:00:09.300    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:09.393    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2617217 kmers in total.\n",
      "  0:00:09.393    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:09.433    48M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:00:09.433    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:09.433    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:00:09.433    48M / 6G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:09.433    48M / 6G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83203 Gb\n",
      "  0:00:09.433    48M / 6G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 349525\n",
      "  0:00:09.667     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 2617217 kmers\n",
      "  0:00:09.667     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 2617217 kmers.\n",
      "  0:00:09.732    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:09.815    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2601126 kmers in total.\n",
      "  0:00:09.815    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:09.856    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:09.991    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:10.021    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1214576 bytes occupied (3.73554 bits per kmer).\n",
      "  0:00:10.024    52M / 6G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:00:10.189    52M / 6G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:10.189    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:00:10.190    52M / 6G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:00:10.190    52M / 6G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:00:10.433    52M / 6G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   1017267 22-mers were removed by early tip clipper\n",
      "  0:00:10.433    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:00:10.434    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:00:10.547    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 49393 sequences extracted\n",
      "  0:00:10.605    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:00:10.637    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 2 loops collected\n",
      "  0:00:10.658    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:00:10.658    52M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:10.658    52M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:10.725    52M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1219896 bytes occupied (3.72883 bits per kmer).\n",
      "  0:00:10.731    64M / 6G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:00:20.870    64M / 6G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:00:20.939    64M / 6G    INFO    General                 (construction.cpp          : 464)   Processed 98688 edges\n",
      "  0:00:20.952    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:00:20.953    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 14\n",
      "  0:00:20.953    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 364\n",
      "  0:00:20.953    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 405. Coverage mad: 131.951\n",
      "  0:00:20.953    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:00:21.105    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:00:21.471    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:00:23.111    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:00:26.866    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:00:29.394    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 424.208. Fitted coverage std. dev: 130.144\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.9986\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 190\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 190\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 1104484\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 424.208\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 190\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 4.26735\n",
      "  0:00:29.403    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:00:29.403    48M / 6G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:00:29.404    48M / 6G    INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:00:29.404    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:00:29.409    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:00:29.409    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:00:29.409    48M / 6G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:00:29.409    48M / 6G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:00:29.409    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:00:29.409    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.422    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 678 times\n",
      "  0:00:29.422    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.003    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 11602 times\n",
      "  0:00:30.003    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.152    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 4416 times\n",
      "  0:00:30.152    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:00:30.152    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.164    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 220 times\n",
      "  0:00:30.164    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.169    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 52 times\n",
      "  0:00:30.169    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.178    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 79 times\n",
      "  0:00:30.178    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:00:30.178    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.179    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 7 times\n",
      "  0:00:30.179    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.179    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
      "  0:00:30.179    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.180    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 22 times\n",
      "  0:00:30.180    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:00:30.180    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.180    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 2 times\n",
      "  0:00:30.180    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.181    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 5 times\n",
      "  0:00:30.181    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.181    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 13 times\n",
      "  0:00:30.181    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:00:30.181    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.181    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.181    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6 times\n",
      "  0:00:30.182    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.182    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.183    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 8 times\n",
      "  0:00:30.183    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:00:30.183    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.183    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.183    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.183    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:00:30.183    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 5 times\n",
      "  0:00:30.184    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:00:30.184    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:00:30.184    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6 times\n",
      "  0:00:30.185    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 3 times\n",
      "  0:00:30.185    52M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:00:30.185    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.186    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:00:30.186    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.187    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:00:30.187    52M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.188    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:00:30.188    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:00:30.188    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.188    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.188    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.188    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.188    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:30.188    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:00:30.189    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:00:30.189    48M / 6G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:00:30.189    48M / 6G    INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:00:30.189    48M / 6G    INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:00:30.189    48M / 6G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:00:30.189    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.189    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.189    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.190    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.190    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:30.190    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:30.190    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:30.191    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:30.191    48M / 6G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:00:30.192    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:00:30.195    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 197 times\n",
      "  0:00:30.195    48M / 6G    INFO    General                 (simplification.cpp        : 340)   Removing all the edges having coverage 4.26735 and less\n",
      "  0:00:30.195    48M / 6G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:00:30.195    48M / 6G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 421.696\n",
      "  0:00:30.195    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:00:30.195    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/assembly_graph_with_scaffolds.gfa\n",
      "  0:00:30.208    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/before_rr.fasta\n",
      "  0:00:30.224    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/assembly_graph.fastg\n",
      "  0:00:30.276    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/simplified_contigs.fasta\n",
      "  0:00:30.291    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/final_contigs.fasta\n",
      "  0:00:30.306    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:00:30.306    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/assembly_graph_with_scaffolds.gfa\n",
      "  0:00:30.319    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/before_rr.fasta\n",
      "  0:00:30.334    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/assembly_graph.fastg\n",
      "  0:00:30.384    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/simplified_contigs.fasta\n",
      "  0:00:30.399    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K21/final_contigs.fasta\n",
      "  0:00:30.415    48M / 6G    INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:00:30.419    48M / 6G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 0 minutes 30 seconds\n",
      "\n",
      "== Running assembler: K27\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K27/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K27/configs/careful_mode.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 246 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from refs/heads/spades_3.13.0, git revision 8ea46659e9b2aca35444a808db550ac333006f8b\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/dataset.info) with K=27\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 151\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 143.316\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83322 Gb\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 466033\n",
      "  0:00:08.159     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 7906818 reads\n",
      "  0:00:08.159     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:00:08.809    48M / 6G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 7906818 reads\n",
      "  0:00:08.810    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:08.921    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2691522 kmers in total.\n",
      "  0:00:08.921    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:08.964    48M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:00:08.964    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:08.964    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:00:08.964    48M / 6G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:08.964    48M / 6G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83203 Gb\n",
      "  0:00:08.964    48M / 6G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 349525\n",
      "  0:00:09.184     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 2691522 kmers\n",
      "  0:00:09.184     6G / 6G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 2691522 kmers.\n",
      "  0:00:09.250    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:09.308    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2681013 kmers in total.\n",
      "  0:00:09.308    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:09.354    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:09.485    48M / 6G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:09.518    48M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1251608 bytes occupied (3.73473 bits per kmer).\n",
      "  0:00:09.521    52M / 6G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:00:09.676    52M / 6G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:09.677    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:00:09.677    52M / 6G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:00:09.677    52M / 6G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:00:09.933    52M / 6G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   1136046 28-mers were removed by early tip clipper\n",
      "  0:00:09.933    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:00:09.934    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:00:10.039    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 34832 sequences extracted\n",
      "  0:00:10.096    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:00:10.131    52M / 6G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 2 loops collected\n",
      "  0:00:10.149    52M / 6G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:00:10.149    52M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:10.149    52M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:10.219    52M / 6G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1254456 bytes occupied (3.72861 bits per kmer).\n",
      "  0:00:10.226    64M / 6G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:00:19.898    64M / 6G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:00:19.961    64M / 6G    INFO    General                 (construction.cpp          : 464)   Processed 69584 edges\n",
      "  0:00:19.974    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:00:19.974    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 14\n",
      "  0:00:19.974    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 355\n",
      "  0:00:19.974    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 385. Coverage mad: 127.504\n",
      "  0:00:19.974    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:00:20.132    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:00:20.553    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:00:22.281    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:00:25.576    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:00:27.135    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 403.331. Fitted coverage std. dev: 125.126\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.998492\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 181\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 181\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 1103733\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 403.331\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 181\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 4.05803\n",
      "  0:00:27.144    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:00:27.144    48M / 6G    INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:00:27.144    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:00:27.150    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:00:27.150    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:00:27.150    48M / 6G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:00:27.150    48M / 6G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:00:27.150    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:00:27.150    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.164    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 786 times\n",
      "  0:00:27.164    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.517    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 7947 times\n",
      "  0:00:27.517    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.620    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2881 times\n",
      "  0:00:27.620    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:00:27.620    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.630    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 188 times\n",
      "  0:00:27.630    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.635    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 38 times\n",
      "  0:00:27.635    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.645    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 65 times\n",
      "  0:00:27.645    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:00:27.645    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 5 times\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 4 times\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 14 times\n",
      "  0:00:27.646    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:00:27.646    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 3 times\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 7 times\n",
      "  0:00:27.647    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 3 times\n",
      "  0:00:27.647    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.647    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.648    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 7 times\n",
      "  0:00:27.648    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:00:27.648    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.648    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.648    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 3 times\n",
      "  0:00:27.649    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:00:27.649    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:00:27.649    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 6 times\n",
      "  0:00:27.649    48M / 6G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.649    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.651    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.651    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:27.651    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:00:27.651    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:00:27.651    48M / 6G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:00:27.651    48M / 6G    INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:00:27.651    48M / 6G    INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:00:27.651    48M / 6G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:00:27.651    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.651    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.651    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.652    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.652    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:27.652    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:27.652    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:27.653    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:27.653    48M / 6G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:00:27.654    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:00:27.658    48M / 6G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 478 times\n",
      "  0:00:27.658    48M / 6G    INFO    General                 (simplification.cpp        : 340)   Removing all the edges having coverage 4.05803 and less\n",
      "  0:00:27.659    48M / 6G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:00:27.659    48M / 6G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 399.023\n",
      "  0:00:27.659    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:00:27.659    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/assembly_graph_with_scaffolds.gfa\n",
      "  0:00:27.672    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/before_rr.fasta\n",
      "  0:00:27.687    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/assembly_graph.fastg\n",
      "  0:00:27.737    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/simplified_contigs.fasta\n",
      "  0:00:27.752    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/final_contigs.fasta\n",
      "  0:00:27.767    48M / 6G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:00:27.767    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/assembly_graph_with_scaffolds.gfa\n",
      "  0:00:27.780    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/before_rr.fasta\n",
      "  0:00:27.795    48M / 6G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/assembly_graph.fastg\n",
      "  0:00:27.845    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/simplified_contigs.fasta\n",
      "  0:00:27.859    48M / 6G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K27/final_contigs.fasta\n",
      "  0:00:27.874    48M / 6G    INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:00:27.877    48M / 6G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 0 minutes 27 seconds\n",
      "\n",
      "== Running assembler: K33\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K33/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K33/configs/careful_mode.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 246 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from refs/heads/spades_3.13.0, git revision 8ea46659e9b2aca35444a808db550ac333006f8b\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/dataset.info) with K=33\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 151\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 143.316\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83322 Gb\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 233016\n",
      "  0:00:08.480     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 7112199 reads\n",
      "  0:00:09.452     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 289)   Processed 7906818 reads\n",
      "  0:00:09.452     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 295)   Adding contigs from previous K\n",
      "  0:00:10.420    48M / 7G    INFO    General                 (kmer_splitters.hpp        : 308)   Used 7906818 reads\n",
      "  0:00:10.420    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:10.543    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2743507 kmers in total.\n",
      "  0:00:10.543    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:10.612    48M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Extension index construction\n",
      "  0:00:10.612    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:10.612    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 192 files using 12 threads. This might take a while.\n",
      "  0:00:10.612    48M / 7G    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:10.612    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83203 Gb\n",
      "  0:00:10.612    48M / 7G    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 174762\n",
      "  0:00:10.916     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 380)   Processed 2743507 kmers\n",
      "  0:00:10.916     6G / 7G    INFO    General                 (kmer_splitters.hpp        : 385)   Used 2743507 kmers.\n",
      "  0:00:10.994    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 120)   Starting k-mer counting.\n",
      "  0:00:11.092    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 127)   K-mer counting done. There are 2735919 kmers in total.\n",
      "  0:00:11.092    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 133)   Merging temporary buckets.\n",
      "  0:00:11.171    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:11.276    48M / 7G    INFO    General                 (kmer_index_builder.hpp    : 150)   Merging final buckets.\n",
      "  0:00:11.343    48M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1277096 bytes occupied (3.73431 bits per kmer).\n",
      "  0:00:11.346    52M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build:  99)   Building k-mer extensions from k+1-mers\n",
      "  0:00:11.547    52M / 7G    INFO   DeBruijnExtensionIndexBu (kmer_extension_index_build: 103)   Building k-mer extensions from k+1-mers finished.\n",
      "  0:00:11.547    52M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Early tip clipping\n",
      "  0:00:11.547    52M / 7G    INFO    General                 (construction.cpp          : 253)   Early tip clipper length bound set as (RL - K)\n",
      "  0:00:11.547    52M / 7G    INFO   Early tip clipping       (early_simplification.hpp  : 181)   Early tip clipping\n",
      "  0:00:11.843    52M / 7G    INFO   Early tip clipping       (early_simplification.hpp  : 184)   1200726 34-mers were removed by early tip clipper\n",
      "  0:00:11.843    52M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Condensing graph\n",
      "  0:00:11.844    52M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 355)   Extracting unbranching paths\n",
      "  0:00:11.952    52M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 374)   Extracting unbranching paths finished. 27446 sequences extracted\n",
      "  0:00:12.010    52M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 310)   Collecting perfect loops\n",
      "  0:00:12.049    52M / 7G    INFO   UnbranchingPathExtractor (debruijn_graph_constructor: 343)   Collecting perfect loops finished. 1 loops collected\n",
      "  0:00:12.063    52M / 7G    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == Filling coverage indices (PHM)\n",
      "  0:00:12.063    52M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 301)   Building kmer index\n",
      "  0:00:12.063    52M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 314)   Building perfect hash indices\n",
      "  0:00:12.126    52M / 7G    INFO   K-mer Index Building     (kmer_index_builder.hpp    : 336)   Index built. Total 1278472 bytes occupied (3.72799 bits per kmer).\n",
      "  0:00:12.134    64M / 7G    INFO    General                 (construction.cpp          : 388)   Collecting k-mer coverage information from reads, this takes a while.\n",
      "  0:00:21.402    64M / 7G    INFO    General                 (construction.cpp          : 508)   Filling coverage and flanking coverage from PHM\n",
      "  0:00:21.467    64M / 7G    INFO    General                 (construction.cpp          : 464)   Processed 54821 edges\n",
      "  0:00:21.483    48M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == EC Threshold Finding\n",
      "  0:00:21.484    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 181)   Kmer coverage valley at: 19\n",
      "  0:00:21.484    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 201)   K-mer histogram maximum: 358\n",
      "  0:00:21.484    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 237)   Estimated median coverage: 367. Coverage mad: 120.091\n",
      "  0:00:21.484    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 259)   Fitting coverage model\n",
      "  0:00:21.629    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 2\n",
      "  0:00:22.029    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 4\n",
      "  0:00:23.691    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 8\n",
      "  0:00:26.785    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 295)   ... iteration 16\n",
      "  0:00:28.974    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 309)   Fitted mean coverage: 382.443. Fitted coverage std. dev: 120.018\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 334)   Probability of erroneous kmer at valley: 0.996575\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 358)   Preliminary threshold calculated as: 173\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 362)   Threshold adjusted to: 173\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (kmer_coverage_model.cpp   : 375)   Estimated genome size (ignoring repeats): 1101607\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (genomic_info_filler.cpp   : 112)   Mean coverage was calculated as 382.443\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (genomic_info_filler.cpp   : 127)   EC coverage threshold value was calculated as 173\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (genomic_info_filler.cpp   : 128)   Trusted kmer low bound: 3.8487\n",
      "  0:00:28.981    48M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Raw Simplification\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (simplification.cpp        : 128)   PROCEDURE == InitialCleaning\n",
      "  0:00:28.981    48M / 7G    INFO    General                 (graph_simplification.hpp  : 662)   Flanking coverage based disconnection disabled\n",
      "  0:00:28.981    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Self conjugate edge remover\n",
      "  0:00:28.984    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Self conjugate edge remover triggered 0 times\n",
      "  0:00:28.984    48M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification\n",
      "  0:00:28.984    48M / 7G    INFO    General                 (simplification.cpp        : 357)   Graph simplification started\n",
      "  0:00:28.984    48M / 7G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:00:28.984    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 1\n",
      "  0:00:28.984    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:28.998    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 886 times\n",
      "  0:00:28.998    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.253    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 6215 times\n",
      "  0:00:29.253    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.327    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1987 times\n",
      "  0:00:29.328    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 2\n",
      "  0:00:29.328    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.337    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 125 times\n",
      "  0:00:29.337    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.341    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 34 times\n",
      "  0:00:29.341    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.350    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 55 times\n",
      "  0:00:29.350    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 3\n",
      "  0:00:29.350    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.350    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 7 times\n",
      "  0:00:29.350    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 7 times\n",
      "  0:00:29.351    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 4\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 2 times\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 1 times\n",
      "  0:00:29.351    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 9 times\n",
      "  0:00:29.352    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 5\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:00:29.352    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 6\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 3 times\n",
      "  0:00:29.352    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 7\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.352    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 5 times\n",
      "  0:00:29.353    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 8\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 2 times\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 1 times\n",
      "  0:00:29.353    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 9\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:00:29.353    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 10\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.353    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.354    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 2 times\n",
      "  0:00:29.354    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 11\n",
      "  0:00:29.354    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.354    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 1 times\n",
      "  0:00:29.354    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.355    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.355    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:00:29.356    48M / 7G    INFO    General                 (simplification.cpp        : 362)   PROCEDURE == Simplification cycle, iteration 12\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Low coverage edge remover\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Low coverage edge remover triggered 0 times\n",
      "  0:00:29.356    48M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Simplification Cleanup\n",
      "  0:00:29.356    48M / 7G    INFO    General                 (simplification.cpp        : 196)   PROCEDURE == Post simplification\n",
      "  0:00:29.356    48M / 7G    INFO    General                 (graph_simplification.hpp  : 453)   Disconnection of relatively low covered edges disabled\n",
      "  0:00:29.356    48M / 7G    INFO    General                 (graph_simplification.hpp  : 489)   Complex tip clipping disabled\n",
      "  0:00:29.356    48M / 7G    INFO    General                 (graph_simplification.hpp  : 634)   Creating parallel br instance\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.356    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.357    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.357    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Tip clipper\n",
      "  0:00:29.357    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Tip clipper triggered 0 times\n",
      "  0:00:29.357    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Bulge remover\n",
      "  0:00:29.358    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Bulge remover triggered 0 times\n",
      "  0:00:29.358    48M / 7G    INFO    General                 (simplification.cpp        : 330)   Disrupting self-conjugate edges\n",
      "  0:00:29.359    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 165)   Running Removing isolated edges\n",
      "  0:00:29.366    48M / 7G    INFO   Simplification           (parallel_processing.hpp   : 167)   Removing isolated edges triggered 712 times\n",
      "  0:00:29.366    48M / 7G    INFO    General                 (simplification.cpp        : 340)   Removing all the edges having coverage 3.8487 and less\n",
      "  0:00:29.368    48M / 7G    INFO    General                 (simplification.cpp        : 470)   Counting average coverage\n",
      "  0:00:29.368    48M / 7G    INFO    General                 (simplification.cpp        : 476)   Average coverage = 378.311\n",
      "  0:00:29.368    48M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:00:29.368    48M / 7G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/assembly_graph_with_scaffolds.gfa\n",
      "  0:00:29.382    48M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/before_rr.fasta\n",
      "  0:00:29.397    48M / 7G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/assembly_graph.fastg\n",
      "  0:00:29.446    48M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/simplified_contigs.fasta\n",
      "  0:00:29.461    48M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/final_contigs.fasta\n",
      "  0:00:29.475    48M / 7G    INFO   StageManager             (stage.cpp                 : 132)   STAGE == Contig Output\n",
      "  0:00:29.475    48M / 7G    INFO    General                 (contig_output_stage.cpp   :  40)   Writing GFA to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/assembly_graph_with_scaffolds.gfa\n",
      "  0:00:29.487    48M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/before_rr.fasta\n",
      "  0:00:29.502    48M / 7G    INFO    General                 (contig_output_stage.cpp   :  51)   Outputting FastG graph to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/assembly_graph.fastg\n",
      "  0:00:29.551    48M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/simplified_contigs.fasta\n",
      "  0:00:29.565    48M / 7G    INFO    General                 (contig_output.hpp         :  22)   Outputting contigs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_//K33/final_contigs.fasta\n",
      "  0:00:29.579    48M / 7G    INFO    General                 (launch.hpp                : 149)   SPAdes finished\n",
      "  0:00:29.582    48M / 7G    INFO    General                 (main.cpp                  : 109)   Assembling time: 0 hours 0 minutes 29 seconds\n",
      "\n",
      "== Running assembler: K55\n",
      "\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K55/configs/config.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  74)   Loaded config from /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/K55/configs/careful_mode.info\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (memory_limit.cpp          :  49)   Memory limit set to 246 Gb\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  87)   Starting SPAdes, built from refs/heads/spades_3.13.0, git revision 8ea46659e9b2aca35444a808db550ac333006f8b\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  88)   Maximum k-mer length: 128\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  89)   Assembling dataset (/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/dataset.info) with K=55\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  90)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  51)   SPAdes started\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  58)   Starting from stage: construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  65)   Two-step RR enabled: 0\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (launch.hpp                :  76)   Will need read mapping, kmer mapper will be attached\n",
      "  0:00:00.000     4M / 4M    INFO   StageManager             (stage.cpp                 : 132)   STAGE == de Bruijn graph construction\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (read_converter.hpp        :  59)   Binary reads detected\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 111)   Max read length 151\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (construction.cpp          : 117)   Average read length 143.316\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (stage.cpp                 : 101)   PROCEDURE == k+1-mer counting\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_index_builder.hpp    : 117)   Splitting kmer instances into 144 files using 12 threads. This might take a while.\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (file_limit.hpp            :  32)   Open file limit set to 50000\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  89)   Memory available for splitting buffers: 6.83322 Gb\n",
      "  0:00:00.001     4M / 4M    INFO    General                 (kmer_splitters.hpp        :  97)   Using cell size of 233016\n",
      "  0:01:51.464    96M / 172M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_1_length_375077_cov_67.153941 processed with 0 changes in thread 0\n",
      "  0:01:51.479    48M / 172M  INFO   DatasetProcessor         (dataset_processor.cpp     : 255)   Gluing processed contigs\n",
      "  0:01:51.483    48M / 172M  INFO    General                 (main.cpp                  :  72)   Correcting time: 0 hours 1 minutes 51 seconds\n",
      "\n",
      "== Processing of scaffolds\n",
      "\n",
      "\n",
      "== Running contig polishing tool: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/spades-3.13.0-0/bin/spades-corrector-core /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/mismatch_corrector/scaffolds/configs/corrector.info /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/misc/assembled_scaffolds.fasta\n",
      "\n",
      "\n",
      "== Dataset description file was created: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/mismatch_corrector/scaffolds/configs/corrector.info\n",
      "\n",
      "/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/mismatch_corrector/scaffolds/configs/log.properties  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  58)   Starting MismatchCorrector, built from refs/heads/spades_3.13.0, git revision 8ea46659e9b2aca35444a808db550ac333006f8b\n",
      "  0:00:00.000     4M / 4M    INFO    General                 (main.cpp                  :  59)   Maximum # of threads to use (adjusted due to OMP capabilities): 12\n",
      "  0:00:00.000     4M / 4M    INFO   DatasetProcessor         (dataset_processor.cpp     : 195)   Splitting assembly...\n",
      "  0:00:00.000     4M / 4M    INFO   DatasetProcessor         (dataset_processor.cpp     : 196)   Assembly file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/misc/assembled_scaffolds.fasta\n",
      "  0:00:00.027     4M / 4M    INFO   DatasetProcessor         (dataset_processor.cpp     : 203)   Processing paired sublib of number 0\n",
      "  0:00:00.027     4M / 4M    INFO   DatasetProcessor         (dataset_processor.cpp     : 206)   /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz\n",
      "  0:00:00.027     4M / 4M    INFO   DatasetProcessor         (dataset_processor.cpp     : 140)   Running bwa index ...: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/spades-3.13.0-0/bin/spades-bwa index -a is /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/misc/assembled_scaffolds.fasta\n",
      "[bwa_index] Pack FASTA... 0.00 sec\n",
      "[bwa_index] Construct BWT for the packed sequence...\n",
      "[bwa_index] 0.28 seconds elapse.\n",
      "[bwa_index] Update BWT... 0.00 sec\n",
      "[bwa_index] Pack forward-only FASTA... 0.01 sec\n",
      "[bwa_index] Construct SA from BWT and Occ... 0.12 sec\n",
      "[main] Version: 0.7.12-r1039\n",
      "[main] CMD: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/spades-3.13.0-0/bin/spades-bwa index -a is /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/misc/assembled_scaffolds.fasta\n",
      "[main] Real time: 0.427 sec; CPU: 0.416 sec\n",
      "  0:00:00.457     4M / 4M    INFO   DatasetProcessor         (dataset_processor.cpp     : 149)   Running bwa mem ...:/mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/spades-3.13.0-0/bin/spades-bwa mem  -v 1 -t 12 /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/misc/assembled_scaffolds.fasta /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz  > /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp/corrector_q2GjOp/lib0_r8JfDh/tmp.sam\n",
      "[main] Version: 0.7.12-r1039\n",
      "[main] CMD: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/spades-3.13.0-0/bin/spades-bwa mem -v 1 -t 12 /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/misc/assembled_scaffolds.fasta /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R1.mapped.fastq.gz /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/alignment/fastq/CW56_S1_.SS14_R2.mapped.fastq.gz\n",
      "[main] Real time: 25.404 sec; CPU: 239.772 sec\n",
      "  0:00:25.886     4M / 4M    INFO   DatasetProcessor         (dataset_processor.cpp     : 209)   Adding samfile /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp/corrector_q2GjOp/lib0_r8JfDh/tmp.sam\n",
      "  0:00:30.153    44M / 44M   INFO   DatasetProcessor         (dataset_processor.cpp     : 105)   processed 1000000reads, flushing\n",
      "  0:00:34.509    44M / 44M   INFO   DatasetProcessor         (dataset_processor.cpp     : 105)   processed 2000000reads, flushing\n",
      "  0:00:38.843    44M / 44M   INFO   DatasetProcessor         (dataset_processor.cpp     : 105)   processed 3000000reads, flushing\n",
      "  0:00:43.263    44M / 44M   INFO   DatasetProcessor         (dataset_processor.cpp     : 105)   processed 4000000reads, flushing\n",
      "  0:00:47.761    44M / 44M   INFO   DatasetProcessor         (dataset_processor.cpp     : 105)   processed 5000000reads, flushing\n",
      "  0:00:50.871     8M / 44M   INFO   DatasetProcessor         (dataset_processor.cpp     : 235)   Processing contigs\n",
      "  0:00:56.953   168M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_8_length_43722_cov_59.645257 processed with 0 changes in thread 3\n",
      "  0:00:56.959   164M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_10_length_37477_cov_70.532262 processed with 0 changes in thread 5\n",
      "  0:00:57.421   160M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_9_length_43473_cov_64.821852 processed with 0 changes in thread 11\n",
      "  0:00:57.657   156M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_7_length_44530_cov_66.819134 processed with 0 changes in thread 7\n",
      "  0:00:57.812   152M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_11_length_36438_cov_81.402605 processed with 1 changes in thread 6\n",
      "  0:01:02.072   148M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_6_length_63385_cov_77.335230 processed with 0 changes in thread 9\n",
      "  0:01:03.186   140M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_5_length_76950_cov_69.043008 processed with 0 changes in thread 10\n",
      "  0:01:07.206   136M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_4_length_86646_cov_81.669911 processed with 0 changes in thread 2\n",
      "  0:01:09.284   128M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_3_length_126709_cov_62.629402 processed with 0 changes in thread 8\n",
      "  0:01:18.457   112M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_2_length_165590_cov_68.322429 processed with 0 changes in thread 0\n",
      "  0:01:50.125    92M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 251)   Contig NODE_1_length_375077_cov_67.153941 processed with 0 changes in thread 1\n",
      "  0:01:50.149    48M / 168M  INFO   DatasetProcessor         (dataset_processor.cpp     : 255)   Gluing processed contigs\n",
      "  0:01:50.154    48M / 168M  INFO    General                 (main.cpp                  :  72)   Correcting time: 0 hours 1 minutes 50 seconds\n",
      "\n",
      "===== Mismatch correction finished.\n",
      "\n",
      " * Corrected reads are in /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/corrected/\n",
      " * Assembled contigs are in /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/contigs.fasta\n",
      " * Assembled scaffolds are in /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/scaffolds.fasta\n",
      " * Assembly graph is in /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/assembly_graph.fastg\n",
      " * Assembly graph in GFA format is in /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/assembly_graph_with_scaffolds.gfa\n",
      " * Paths in the assembly graph corresponding to the contigs are in /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/contigs.paths\n",
      " * Paths in the assembly graph corresponding to the scaffolds are in /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/scaffolds.paths\n",
      "\n",
      "======= SPAdes pipeline finished.\n",
      "\n",
      "SPAdes log can be found here: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/CW56_S1_.SS14_/spades.log\n",
      "\n",
      "Thank you for using SPAdes!\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "KMERS=\"15,21,27,33,55,77,99,127\" # List of kmers for the SPAdes assembly\n",
    "NSEQ=10 # How many longest sequences we will report in short assembly fasta result\n",
    "\n",
    "mkdir -p $OUTPUT_DIR/assembly/spades\n",
    "\n",
    "for sample in *R1.mapped.fastq.gz\n",
    "do\n",
    "    FORWARD=$sample\n",
    "    extension=\"${FORWARD##*R1}\"\n",
    "    REVERSE=${FORWARD%R1*}R2${extension}\n",
    "\n",
    "    # Input file check\n",
    "    if [ ! -f ${FORWARD} ]; then\n",
    "        echo \"Input file not found! First in pair.\"; echo ${FORWARD}\n",
    "    fi\n",
    "    if [ ! -f ${REVERSE} ]; then\n",
    "        echo \"Input file not found! Second in pair.\"; echo ${REVERSE}\n",
    "    fi\n",
    "    \n",
    "    NAME=${FORWARD%R1*}\n",
    "\n",
    "    mkdir $OUTPUT_DIR/assembly/spades/$NAME\n",
    "\n",
    "    spades.py --careful --threads $THREADS --cov-cutoff 5 --tmp-dir $OUTPUT_DIR/tmp -k $KMERS -1 $FORWARD -2 $REVERSE -o $OUTPUT_DIR/assembly/spades/$NAME\n",
    "done\n",
    "\n",
    "# Assembly statistics\n",
    "cd $OUTPUT_DIR/assembly/spades\n",
    "\n",
    "dirs_tmp=$(ls -d */)\n",
    "\n",
    "mkdir scaffolds\n",
    "mkdir graphs\n",
    "\n",
    "for sample in $dirs_tmp\n",
    "do\n",
    "    cat $sample/scaffolds.fasta > scaffolds/${sample%/*}.scaffolds.fasta\n",
    "    cat $sample/assembly_graph.fastg > graphs/${sample%/*}.assembly_graph.fastg\n",
    "    cat $sample/assembly_graph_with_scaffolds.gfa > graphs/${sample%/*}.assembly_graph_with_scaffolds.gfa\n",
    "done\n",
    "\n",
    "cd scaffolds/\n",
    "\n",
    "for sample in *.fasta\n",
    "do\n",
    "    awk \"/^>/ {n++} n>$NSEQ {exit} {print}\" $sample > ${sample%.*}.top${NSEQ}.fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Assembly quality check"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Once the assembly is done we strongly recommend to run assembly quality check. We also recommend looking at the resulting graphs at [Bandage](https://rrwick.github.io/Bandage/)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) mkdir: cannot create directory /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast: File exists\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) --2019-06-26 22:07:24--  https://busco.ezlab.org/datasets/spirochaetes_odb9.tar.gz\n",
      "Resolving busco.ezlab.org (busco.ezlab.org)... 129.194.96.125\n",
      "Connecting to busco.ezlab.org (busco.ezlab.org)|129.194.96.125|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 9269492 (8.8M) [application/octet-stream]\n",
      "Saving to: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/BUSCO/spirochaetes_odb9.tar.gz\n",
      "\n",
      "/mnt/ssd/ssd_2/biod 100%[===================>]   8.84M  17.9MB/s    in 0.5s    \n",
      "\n",
      "2019-06-26 22:07:24 (17.9 MB/s) - /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/BUSCO/spirochaetes_odb9.tar.gz saved [9269492/9269492]\n",
      "\n",
      "(treponema) (.env) spirochaetes_odb9/\n",
      "spirochaetes_odb9/hmms/\n",
      "spirochaetes_odb9/hmms/POG091N038E.hmm\n",
      "spirochaetes_odb9/hmms/POG091N014E.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01G6.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02J8.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00LU.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02LM.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03AM.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03CQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00WR.hmm\n",
      "spirochaetes_odb9/hmms/POG091N002X.hmm\n",
      "spirochaetes_odb9/hmms/POG091N016U.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02PI.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02XE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01GP.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01LQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01NE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01QI.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01M7.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01MG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01JG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00HA.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02BT.hmm\n",
      "spirochaetes_odb9/hmms/POG091N036L.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01E5.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01UQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02A7.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01VY.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0073.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01JY.hmm\n",
      "spirochaetes_odb9/hmms/POG091N034Q.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02TE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01Y6.hmm\n",
      "spirochaetes_odb9/hmms/POG091N014N.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01ZZ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0017.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0180.hmm\n",
      "spirochaetes_odb9/hmms/POG091N000I.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0142.hmm\n",
      "spirochaetes_odb9/hmms/POG091N032R.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00AG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N037H.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00V5.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01GS.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02QS.hmm\n",
      "spirochaetes_odb9/hmms/POG091N030W.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02UD.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0007.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00XL.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0369.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01MK.hmm\n",
      "spirochaetes_odb9/hmms/POG091N038C.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02XQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01EQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01M4.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01QV.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00YJ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01Q3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N027F.hmm\n",
      "spirochaetes_odb9/hmms/POG091N006P.hmm\n",
      "spirochaetes_odb9/hmms/POG091N020G.hmm\n",
      "spirochaetes_odb9/hmms/POG091N000F.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02NU.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00GB.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01SE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01G2.hmm\n",
      "spirochaetes_odb9/hmms/POG091N004E.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0063.hmm\n",
      "spirochaetes_odb9/hmms/POG091N011X.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0294.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00SR.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01KQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0158.hmm\n",
      "spirochaetes_odb9/hmms/POG091N018W.hmm\n",
      "spirochaetes_odb9/hmms/POG091N006X.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01OW.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03CK.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01CJ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00S7.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00SO.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01UW.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00YW.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00PL.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02QZ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00V4.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01LO.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00OW.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00ZU.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03AJ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N039X.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01IT.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00P8.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0016.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02Q3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0087.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01BV.hmm\n",
      "spirochaetes_odb9/hmms/POG091N011Y.hmm\n",
      "spirochaetes_odb9/hmms/POG091N011W.hmm\n",
      "spirochaetes_odb9/hmms/POG091N020T.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01F3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N003Q.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00E5.hmm\n",
      "spirochaetes_odb9/hmms/POG091N034S.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02PQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03FK.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01YD.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01TC.hmm\n",
      "spirochaetes_odb9/hmms/POG091N028X.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01OB.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02Y2.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0059.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00JW.hmm\n",
      "spirochaetes_odb9/hmms/POG091N029I.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03F0.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01JZ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01HD.hmm\n",
      "spirochaetes_odb9/hmms/POG091N013M.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01Q5.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01JU.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01LR.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00G9.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00Q0.hmm\n",
      "spirochaetes_odb9/hmms/POG091N027V.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02SF.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01N0.hmm\n",
      "spirochaetes_odb9/hmms/POG091N002T.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02J9.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0227.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02ZE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01WI.hmm\n",
      "spirochaetes_odb9/hmms/POG091N000C.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00H3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02H7.hmm\n",
      "spirochaetes_odb9/hmms/POG091N016S.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01D8.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01YG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01LL.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01PU.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02AK.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01CR.hmm\n",
      "spirochaetes_odb9/hmms/POG091N029P.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01DP.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00EL.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01OE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01LH.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02AF.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00QW.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01U9.hmm\n",
      "spirochaetes_odb9/hmms/POG091N017D.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02ZK.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00LV.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01Z3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00AC.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00PC.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01D3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02FJ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00EM.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02BI.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03E0.hmm\n",
      "spirochaetes_odb9/hmms/POG091N029Z.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01RB.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01VR.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00IF.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01DC.hmm\n",
      "spirochaetes_odb9/hmms/POG091N017W.hmm\n",
      "spirochaetes_odb9/hmms/POG091N008Z.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02KQ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0089.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01OX.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01P7.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00RL.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00RY.hmm\n",
      "spirochaetes_odb9/hmms/POG091N037O.hmm\n",
      "spirochaetes_odb9/hmms/POG091N006Q.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00R3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0186.hmm\n",
      "spirochaetes_odb9/hmms/POG091N037L.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02DN.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01AT.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03CH.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0366.hmm\n",
      "spirochaetes_odb9/hmms/POG091N015Q.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01IG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00XH.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0198.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02KB.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03BZ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N023H.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01EC.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0331.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0206.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00OG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00NI.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0393.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01DF.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01CT.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01P6.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00IX.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01TB.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02AL.hmm\n",
      "spirochaetes_odb9/hmms/POG091N03CE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N025O.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01L4.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02SJ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N027H.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01TE.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01YO.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01J0.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00BX.hmm\n",
      "spirochaetes_odb9/hmms/POG091N023F.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01T0.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02OF.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0204.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01EZ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02IG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01UB.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0377.hmm\n",
      "spirochaetes_odb9/hmms/POG091N0197.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00TN.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00EN.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01Z5.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02CZ.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00PN.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01Z9.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01V2.hmm\n",
      "spirochaetes_odb9/hmms/POG091N008U.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01JK.hmm\n",
      "spirochaetes_odb9/hmms/POG091N02LH.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01T3.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01E2.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01TG.hmm\n",
      "spirochaetes_odb9/hmms/POG091N001J.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00ID.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00UR.hmm\n",
      "spirochaetes_odb9/hmms/POG091N00YX.hmm\n",
      "spirochaetes_odb9/hmms/POG091N006G.hmm\n",
      "spirochaetes_odb9/hmms/POG091N014M.hmm\n",
      "spirochaetes_odb9/hmms/POG091N01HM.hmm\n",
      "spirochaetes_odb9/ancestral\n",
      "spirochaetes_odb9/dataset.cfg\n",
      "spirochaetes_odb9/prfl/\n",
      "spirochaetes_odb9/prfl/POG091N01UQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0393.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0294.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01E2.prfl\n",
      "spirochaetes_odb9/prfl/POG091N006X.prfl\n",
      "spirochaetes_odb9/prfl/POG091N017W.prfl\n",
      "spirochaetes_odb9/prfl/POG091N006Q.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01N0.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02IG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01YD.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00IX.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01OE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00ZU.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01OB.prfl\n",
      "spirochaetes_odb9/prfl/POG091N000F.prfl\n",
      "spirochaetes_odb9/prfl/POG091N016U.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01VR.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01TE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02AF.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00YW.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00JW.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01U9.prfl\n",
      "spirochaetes_odb9/prfl/POG091N029I.prfl\n",
      "spirochaetes_odb9/prfl/POG091N011W.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02UD.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02OF.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01M7.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00NI.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00ID.prfl\n",
      "spirochaetes_odb9/prfl/POG091N002X.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01F3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01JU.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02LM.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01JY.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01CJ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01Q5.prfl\n",
      "spirochaetes_odb9/prfl/POG091N023F.prfl\n",
      "spirochaetes_odb9/prfl/POG091N037L.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03CK.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03E0.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00XL.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01E5.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01ZZ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00BX.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00EL.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01YG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01G6.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0186.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00IF.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0063.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02ZK.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02NU.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00Q0.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01DC.prfl\n",
      "spirochaetes_odb9/prfl/POG091N014N.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00SR.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0227.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02FJ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00QW.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0158.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01EC.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01EZ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01V2.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01SE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01EQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02PI.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01GP.prfl\n",
      "spirochaetes_odb9/prfl/POG091N027F.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00H3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0059.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00PN.prfl\n",
      "spirochaetes_odb9/prfl/POG091N027H.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01Z3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02KB.prfl\n",
      "spirochaetes_odb9/prfl/POG091N000C.prfl\n",
      "spirochaetes_odb9/prfl/POG091N025O.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00AC.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01GS.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00WR.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0089.prfl\n",
      "spirochaetes_odb9/prfl/POG091N008Z.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01L4.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02BT.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00GB.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01JG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01LH.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0366.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01HM.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01JK.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00EM.prfl\n",
      "spirochaetes_odb9/prfl/POG091N032R.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02XE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N018W.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03CQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01T3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00V4.prfl\n",
      "spirochaetes_odb9/prfl/POG091N003Q.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00PC.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01HD.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02AL.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01TC.prfl\n",
      "spirochaetes_odb9/prfl/POG091N023H.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03F0.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01AT.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00RY.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00PL.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01G2.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00TN.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01MG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01OW.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01DF.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00YX.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01Z9.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02BI.prfl\n",
      "spirochaetes_odb9/prfl/POG091N029Z.prfl\n",
      "spirochaetes_odb9/prfl/POG091N000I.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02KQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N020T.prfl\n",
      "spirochaetes_odb9/prfl/POG091N037O.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01J0.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01BV.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01Y6.prfl\n",
      "spirochaetes_odb9/prfl/POG091N038E.prfl\n",
      "spirochaetes_odb9/prfl/POG091N011Y.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0073.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02J9.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02A7.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0204.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02TE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02J8.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01Q3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N037H.prfl\n",
      "spirochaetes_odb9/prfl/POG091N028X.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01OX.prfl\n",
      "spirochaetes_odb9/prfl/POG091N029P.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0180.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02H7.prfl\n",
      "spirochaetes_odb9/prfl/POG091N036L.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02PQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03CE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00V5.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02QZ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03AM.prfl\n",
      "spirochaetes_odb9/prfl/POG091N017D.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00S7.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01CR.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00XH.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01YO.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02SJ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0206.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01IG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02Q3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01D3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02DN.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01UW.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0197.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01IT.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03FK.prfl\n",
      "spirochaetes_odb9/prfl/POG091N006G.prfl\n",
      "spirochaetes_odb9/prfl/POG091N015Q.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01Z5.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03AJ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N030W.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01PU.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01P7.prfl\n",
      "spirochaetes_odb9/prfl/POG091N039X.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02LH.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02CZ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01CT.prfl\n",
      "spirochaetes_odb9/prfl/POG091N034Q.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01LL.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02ZE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0142.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00E5.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02XQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00R3.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00OG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N016S.prfl\n",
      "spirochaetes_odb9/prfl/POG091N038C.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01LR.prfl\n",
      "spirochaetes_odb9/prfl/POG091N034S.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03CH.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01T0.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01RB.prfl\n",
      "spirochaetes_odb9/prfl/POG091N027V.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02Y2.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01QV.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00SO.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00LU.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01DP.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01TG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N011X.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01LQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01WI.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00YJ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01VY.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00OW.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00UR.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00HA.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01JZ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N014M.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0087.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01LO.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0016.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0331.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00AG.prfl\n",
      "spirochaetes_odb9/prfl/POG091N013M.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01P6.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0017.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0198.prfl\n",
      "spirochaetes_odb9/prfl/POG091N001J.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00G9.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01UB.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01MK.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02SF.prfl\n",
      "spirochaetes_odb9/prfl/POG091N008U.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01NE.prfl\n",
      "spirochaetes_odb9/prfl/POG091N004E.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00LV.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02QS.prfl\n",
      "spirochaetes_odb9/prfl/POG091N002T.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00P8.prfl\n",
      "spirochaetes_odb9/prfl/POG091N020G.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01KQ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00EN.prfl\n",
      "spirochaetes_odb9/prfl/POG091N014E.prfl\n",
      "spirochaetes_odb9/prfl/POG091N00RL.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01TB.prfl\n",
      "spirochaetes_odb9/prfl/POG091N006P.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01D8.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01QI.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0377.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0007.prfl\n",
      "spirochaetes_odb9/prfl/POG091N03BZ.prfl\n",
      "spirochaetes_odb9/prfl/POG091N0369.prfl\n",
      "spirochaetes_odb9/prfl/POG091N02AK.prfl\n",
      "spirochaetes_odb9/prfl/POG091N01M4.prfl\n",
      "spirochaetes_odb9/ancestral_variants\n",
      "spirochaetes_odb9/scores_cutoff\n",
      "spirochaetes_odb9/lengths_cutoff\n",
      "spirochaetes_odb9/info/\n",
      "spirochaetes_odb9/info/ogs.id.info\n",
      "spirochaetes_odb9/info/species.info\n",
      "spirochaetes_odb9/info/spirochaetes_203691_OrthoDB9_orthogroup_info.txt.gz\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) mkdir: cannot create directory /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/BUSCO/CW56_S1_.SS14_.scaffolds: File exists\n",
      "WARNING\tAn augustus species is mentioned in the config file, dataset default species (E_coli_K12) will be ignored\n",
      "INFO\t****************** Start a BUSCO 3.0.2 analysis, current time: 06/26/2019 22:07:26 ******************\n",
      "INFO\tConfiguration loaded from /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/../config/config.ini\n",
      "INFO\tInit tools...\n",
      "INFO\tCheck dependencies...\n",
      "INFO\tCheck input file...\n",
      "INFO\tTo reproduce this run: python /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/run_BUSCO.py -i CW56_S1_.SS14_.scaffolds.fasta -o busco_CW56_S1_.SS14_.scaffolds -l /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/BUSCO/spirochaetes_odb9/ -m genome -c 12 -t /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp/ -sp E_coli_K12\n",
      "INFO\tMode is: genome\n",
      "INFO\tThe lineage dataset is: spirochaetes_odb9 (prokaryota)\n",
      "INFO\tTemp directory is /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp/\n",
      "INFO\t****** Phase 1 of 2, initial predictions ******\n",
      "INFO\t****** Step 1/3, current time: 06/26/2019 22:07:26 ******\n",
      "INFO\tCreate blast database...\n",
      "INFO\t[makeblastdb]\tBuilding a new DB, current time: 06/26/2019 22:07:28\n",
      "INFO\t[makeblastdb]\tNew DB name:   /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/tmp/busco_CW56_S1_.SS14_.scaffolds_2196785295\n",
      "INFO\t[makeblastdb]\tNew DB title:  CW56_S1_.SS14_.scaffolds.fasta\n",
      "INFO\t[makeblastdb]\tSequence type: Nucleotide\n",
      "INFO\t[makeblastdb]\tKeep MBits: T\n",
      "INFO\t[makeblastdb]\tMaximum file size: 1000000000B\n",
      "INFO\t[makeblastdb]\tAdding sequences from FASTA; added 67 sequences in 0.013979 seconds.\n",
      "INFO\t[makeblastdb]\t1 of 1 task(s) completed at 06/26/2019 22:07:28\n",
      "INFO\tRunning tblastn, writing output to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/scaffolds/run_busco_CW56_S1_.SS14_.scaffolds/blast_output/tblastn_busco_CW56_S1_.SS14_.scaffolds.tsv...\n",
      "INFO\t[tblastn]\t1 of 1 task(s) completed at 06/26/2019 22:07:29\n",
      "INFO\t****** Step 2/3, current time: 06/26/2019 22:07:29 ******\n",
      "INFO\tMaximum number of candidate contig per BUSCO limited to: 3\n",
      "INFO\tGetting coordinates for candidate regions...\n",
      "INFO\tPre-Augustus scaffold extraction...\n",
      "INFO\tRunning Augustus prediction using E_coli_K12 as species:\n",
      "INFO\t[augustus]\tPlease find all logs related to Augustus errors here: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/scaffolds/run_busco_CW56_S1_.SS14_.scaffolds/augustus_output/augustus.log\n",
      "INFO\t[augustus]\t22 of 217 task(s) completed at 06/26/2019 22:07:31\n",
      "INFO\t[augustus]\t44 of 217 task(s) completed at 06/26/2019 22:07:33\n",
      "INFO\t[augustus]\t66 of 217 task(s) completed at 06/26/2019 22:07:35\n",
      "INFO\t[augustus]\t87 of 217 task(s) completed at 06/26/2019 22:07:36\n",
      "INFO\t[augustus]\t109 of 217 task(s) completed at 06/26/2019 22:07:38\n",
      "INFO\t[augustus]\t131 of 217 task(s) completed at 06/26/2019 22:07:40\n",
      "INFO\t[augustus]\t152 of 217 task(s) completed at 06/26/2019 22:07:41\n",
      "INFO\t[augustus]\t174 of 217 task(s) completed at 06/26/2019 22:07:43\n",
      "INFO\t[augustus]\t196 of 217 task(s) completed at 06/26/2019 22:07:45\n",
      "INFO\t[augustus]\t217 of 217 task(s) completed at 06/26/2019 22:07:48\n",
      "INFO\tExtracting predicted proteins...\n",
      "INFO\t****** Step 3/3, current time: 06/26/2019 22:07:49 ******\n",
      "INFO\tRunning HMMER to confirm orthology of predicted proteins:\n",
      "INFO\t[hmmsearch]\t22 of 217 task(s) completed at 06/26/2019 22:07:49\n",
      "INFO\t[hmmsearch]\t44 of 217 task(s) completed at 06/26/2019 22:07:49\n",
      "INFO\t[hmmsearch]\t109 of 217 task(s) completed at 06/26/2019 22:07:49\n",
      "INFO\t[hmmsearch]\t131 of 217 task(s) completed at 06/26/2019 22:07:49\n",
      "INFO\t[hmmsearch]\t174 of 217 task(s) completed at 06/26/2019 22:07:49\n",
      "INFO\t[hmmsearch]\t196 of 217 task(s) completed at 06/26/2019 22:07:50\n",
      "INFO\t[hmmsearch]\t217 of 217 task(s) completed at 06/26/2019 22:07:50\n",
      "INFO\tResults:\n",
      "INFO\tC:80.2%[S:80.2%,D:0.0%],F:2.5%,M:17.3%,n:237\n",
      "INFO\t190 Complete BUSCOs (C)\n",
      "INFO\t190 Complete and single-copy BUSCOs (S)\n",
      "INFO\t0 Complete and duplicated BUSCOs (D)\n",
      "INFO\t6 Fragmented BUSCOs (F)\n",
      "INFO\t41 Missing BUSCOs (M)\n",
      "INFO\t237 Total BUSCO groups searched\n",
      "INFO\t****** Phase 2 of 2, predictions using species specific training ******\n",
      "INFO\t****** Step 1/3, current time: 06/26/2019 22:07:50 ******\n",
      "INFO\tExtracting missing and fragmented buscos from the ancestral_variants file...\n",
      "INFO\tRunning tblastn, writing output to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/scaffolds/run_busco_CW56_S1_.SS14_.scaffolds/blast_output/tblastn_busco_CW56_S1_.SS14_.scaffolds_missing_and_frag_rerun.tsv...\n",
      "INFO\t[tblastn]\t1 of 1 task(s) completed at 06/26/2019 22:07:51\n",
      "ERROR\ttblastn has ended prematurely (the result file lacks the expected final line), which will produce incomplete results in the next steps ! This problem likely appeared in blast+ 2.4 and seems not fully fixed in 2.6. It happens only when using multiple cores. You can use a single core (-c 1) or downgrade to blast+ 2.2.x, a safe choice regarding this issue. See blast+ documentation for more information.\n",
      "ERROR\tBUSCO analysis failed !\n",
      "ERROR\tCheck the logs, read the user guide, if you still need technical support, then please contact mailto:support@orthodb.org\n",
      "\n",
      "(treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/quast -r /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa -t 12 --output-dir /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast CW56_S1_.SS14_.scaffolds.fasta\n",
      "\n",
      "Version: 5.0.2\n",
      "\n",
      "System information:\n",
      "  OS: Linux-4.4.0-148-generic-x86_64-with-debian-stretch-sid (linux_64)\n",
      "  Python version: 2.7.15\n",
      "  CPUs number: 70\n",
      "\n",
      "Started: 2019-06-26 22:07:51\n",
      "\n",
      "Logging to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/quast.log\n",
      "NOTICE: Output directory already exists and looks like a QUAST output dir. Existing results can be reused (e.g. previously generated alignments)!\n",
      "\n",
      "CWD: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/scaffolds\n",
      "Main parameters: \n",
      "  MODE: default, threads: 12, minimum contig length: 500, minimum alignment length: 65, \\\n",
      "  ambiguity: one, threshold for extensive misassembly size: 1000\n",
      "\n",
      "Reference:\n",
      "  /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/SS14.fa ==> SS14\n",
      "\n",
      "Contigs:\n",
      "  Pre-processing...\n",
      "  CW56_S1_.SS14_.scaffolds.fasta ==> CW56_S1_.SS14_.scaffolds\n",
      "\n",
      "2019-06-26 22:07:52\n",
      "Running Basic statistics processor...\n",
      "  Reference genome:\n",
      "    SS14.fa, length = 1139569, num fragments = 1, GC % = 52.79\n",
      "  Contig files: \n",
      "    CW56_S1_.SS14_.scaffolds\n",
      "  Calculating N50 and L50...\n",
      "    CW56_S1_.SS14_.scaffolds, N50 = 126709, L50 = 3, Total length = 1143241, GC % = 52.73, # N's per 100 kbp =  10.41\n",
      "  Drawing Nx plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/basic_stats/Nx_plot.pdf\n",
      "  Drawing NGx plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/basic_stats/NGx_plot.pdf\n",
      "  Drawing cumulative plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/basic_stats/cumulative_plot.pdf\n",
      "  Drawing GC content plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/basic_stats/GC_content_plot.pdf\n",
      "  Drawing CW56_S1_.SS14_.scaffolds GC content plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/basic_stats/CW56_S1_.SS14_.scaffolds_GC_content_plot.pdf\n",
      "  Drawing Coverage histogram (bin size: 1x)...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/basic_stats/coverage_histogram.pdf\n",
      "  Drawing CW56_S1_.SS14_.scaffolds coverage histogram (bin size: 1x)...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/basic_stats/CW56_S1_-SS14_-scaffolds_coverage_histogram.pdf\n",
      "Done.\n",
      "\n",
      "2019-06-26 22:07:54\n",
      "Running Contig analyzer...\n",
      "  CW56_S1_.SS14_.scaffolds\n",
      "  Logging to files /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/contigs_reports/contigs_report_CW56_S1_-SS14_-scaffolds.stdout and contigs_report_CW56_S1_-SS14_-scaffolds.stderr...\n",
      "  Using existing alignments... \n",
      "  Analysis is finished.\n",
      "  Creating total report...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/contigs_reports/misassemblies_report.txt, misassemblies_report.tsv, and misassemblies_report.tex\n",
      "  Transposed version of total report...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/contigs_reports/transposed_report_misassemblies.txt, transposed_report_misassemblies.tsv, and transposed_report_misassemblies.tex\n",
      "  Creating total report...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/contigs_reports/unaligned_report.txt, unaligned_report.tsv, and unaligned_report.tex\n",
      "  Drawing misassemblies by types plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/contigs_reports/misassemblies_plot.pdf\n",
      "  Drawing misassemblies FRCurve plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/contigs_reports/misassemblies_frcurve_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-06-26 22:07:55\n",
      "Running NA-NGA calculation...\n",
      "  CW56_S1_.SS14_.scaffolds, Largest alignment = 375077, NA50 = 126709, NGA50 = 126709, LA50 = 3, LGA50 = 3\n",
      "  Drawing cumulative plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/aligned_stats/cumulative_plot.pdf\n",
      "  Drawing NAx plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/aligned_stats/NAx_plot.pdf\n",
      "  Drawing NGAx plot...\n",
      "    saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/aligned_stats/NGAx_plot.pdf\n",
      "Done.\n",
      "\n",
      "2019-06-26 22:07:56\n",
      "Running Genome analyzer...\n",
      "  NOTICE: No file with genomic features were provided. Use the --features option if you want to specify it.\n",
      "\n",
      "  NOTICE: No file with operons were provided. Use the -O option if you want to specify it.\n",
      "  CW56_S1_.SS14_.scaffolds\n",
      "  Analysis is finished.\n",
      "  Skipping drawing Genome fraction, % histogram... (less than 2 columns histogram makes no sense)\n",
      "Done.\n",
      "\n",
      "NOTICE: Genes are not predicted by default. Use --gene-finding or --glimmer option to enable it.\n",
      "\n",
      "2019-06-26 22:07:56\n",
      "Creating large visual summaries...\n",
      "This may take a while: press Ctrl-C to skip this step..\n",
      "  1 of 2: Creating Icarus viewers...\n",
      "  2 of 2: Creating PDF with all tables and plots...\n",
      "Done\n",
      "\n",
      "2019-06-26 22:07:58\n",
      "RESULTS:\n",
      "  Text versions of total report are saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/report.txt, report.tsv, and report.tex\n",
      "  Text versions of transposed total report are saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/transposed_report.txt, transposed_report.tsv, and transposed_report.tex\n",
      "  HTML version (interactive tables and plots) is saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/report.html\n",
      "  PDF version (tables and plots) is saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/report.pdf\n",
      "  Icarus (contig browser) is saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/icarus.html\n",
      "  Log is saved to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/qc/assembly/quast/quast.log\n",
      "\n",
      "Finished: 2019-06-26 22:07:58\n",
      "Elapsed time: 0:00:06.594745\n",
      "NOTICEs: 4; WARNINGs: 0; non-fatal ERRORs: 0\n",
      "\n",
      "Thank you for using QUAST!\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir -p $OUTPUT_DIR/qc/assembly/BUSCO\n",
    "mkdir $OUTPUT_DIR/qc/assembly/quast\n",
    "\n",
    "#BUSCO_INI=/storage/brno2/home/opplatek/tools/anaconda3/envs/busco/share/busco-3.0.2-6/conda.config.ini\n",
    "#export BUSCO_CONFIG_FILE=$SCRATCH/busco.ini\n",
    "\n",
    "# Get BUSCO core genes annotation database\n",
    "wget https://busco.ezlab.org/datasets/spirochaetes_odb9.tar.gz -O $OUTPUT_DIR/qc/assembly/BUSCO/spirochaetes_odb9.tar.gz\n",
    "tar xvzf $OUTPUT_DIR/qc/assembly/BUSCO/spirochaetes_odb9.tar.gz -C $OUTPUT_DIR/qc/assembly/BUSCO\n",
    "\n",
    "#cp -r /storage/brno2/home/opplatek/tools/anaconda3/pkgs/augustus-3.2.3-boost1.60_0/config $SCRATCH/augustus_config\n",
    "#export AUGUSTUS_CONFIG_PATH=$SCRATCH/augustus_config\n",
    "\n",
    "# -sp XXX is one of the supported organisms by Augustus and should be as close as possible to our organism https://github.com/Gaius-Augustus/Augustus/blob/master/docs/RUNNING-AUGUSTUS.md\n",
    "\n",
    "for sample in *.scaffolds.fasta\n",
    "do\n",
    "    mkdir $OUTPUT_DIR/qc/assembly/BUSCO/${sample%.fa*}\n",
    "    \n",
    "    run_BUSCO.py -i $sample -o busco_${sample%.fa*} -l $OUTPUT_DIR/qc/assembly/BUSCO/spirochaetes_odb9 -m geno \\\n",
    "    -c $THREADS -sp E_coli_K12 -t $OUTPUT_DIR/tmp\n",
    "done\n",
    "\n",
    "mv run_busco* $OUTPUT_DIR/qc/assembly/BUSCO/\n",
    "\n",
    "quast -r $REFERENCE -t $THREADS --output-dir $OUTPUT_DIR/qc/assembly/quast *.scaffolds.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Scaffold contamination scan (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#PBS -l walltime=01:00:00\n",
    "#PBS -l select=1:ncpus=6:mem=160gb:scratch_local=200gb\n",
    "#PBS -N kraken_scan_scaffolds_P101HW18100699\n",
    "#PBS -q default\n",
    "#\n",
    "# Kraken scan fastq.gz\n",
    "# Designed for assembled scaffolds in uncompressed fasta\n",
    "# For paired-end or single-end fastq.gz please see Kraken2 manual https://ccb.jhu.edu/software/kraken2/index.shtml?t=manual#masking-of-low-complexity-sequences\n",
    "#\n",
    "#Note: Kraken2 read annotation - it would be too simple if Kraken2 would put the real taxid into the annotated fasta/fastq output\n",
    "#\tInstead, it puts there complete shit\n",
    "#\tBut, we can take a look into xxx.kraken.txt, get second and third column and from there extract the sequence by name we want from either the original fasta/fastq file or from xxx.classified-out.fasta\n",
    "#\n",
    "\n",
    "INPUT_DIR=/storage/brno2/home/opplatek/bioinf_projects/biocore/smajs_treponema/11_2018/results/assembly/P101HW18100699/all_reads/spades/scaffolds\n",
    "OUTPUT_DIR=${INPUT_DIR%/assembly*}/qc/kraken/P101HW18100699/spades_scaffolds_nonames\n",
    "\n",
    "DBNAME=/storage/plzen1/home/opplatek/genome/kraken_db/122018/kraken_db_122018\n",
    "\n",
    "SUFFIX=\".scaffolds.fasta\"\n",
    "SUFFIX1=$SUFFIX\n",
    "#SUFFIX2=\"_2.fq.gz\"\n",
    "\n",
    "THREADS=$PBS_NUM_PPN\n",
    "\n",
    "KRAKEN_BIN=/storage/brno2/home/opplatek/tools/kraken2-2.0.7-beta\n",
    "\n",
    "cp -r $DBNAME $SCRATCH/ &\n",
    "DBNAME=$(basename $DBNAME)\n",
    "\n",
    "cp $INPUT_DIR/*$SUFFIX $SCRATCH/\n",
    "\n",
    "wait\n",
    "\n",
    "cd $SCRATCH/\n",
    "\n",
    "mkdir $SCRATCH/kraken\n",
    "\n",
    "for i in *$SUFFIX1\n",
    "do\n",
    "\tFORWARD=$i\n",
    "\tREVERSE=${i%$SUFFIX1*}$SUFFIX2\n",
    "\t$KRAKEN_BIN/kraken2 --threads $THREADS --use-names --classified-out $SCRATCH/kraken/${i%$SUFFIX1*}.classified-out.fasta --unclassified-out $SCRATCH/kraken/${i%$SUFFIX1*}.unclassified-out.fasta --output $SCRATCH/kraken/${i%$SUFFIX1*}.kraken.txt --report $SCRATCH/kraken/${i%$SUFFIX1*}.kraken.report.txt --db $SCRATCH/$DBNAME $SCRATCH/$FORWARD # --gzip-compressed\n",
    "\n",
    "#\tfor a in *.unclassified*.fasta\n",
    "#\tdo\n",
    "#\t\tpigz -p $THREADS $a\n",
    "#\tdone \n",
    "\n",
    "#\tfor a in *.classified*.fasta\n",
    "#\tdo\n",
    "#\t\tpigz -p $THREADS $a\n",
    "#\tdone \n",
    "\n",
    "# Extract only the annotated reads and their taxa\n",
    "cat $SCRATCH/kraken/${i%$SUFFIX1*}.kraken.txt | cut -f2,3 $i > $SCRATCH/kraken/${i%$SUFFIX1*}.kraken.taxid.txt\n",
    "\n",
    "done\n",
    "\n",
    "\n",
    "\n",
    "mkdir -p $OUTPUT_DIR\n",
    "\n",
    "cp -r $SCRATCH/kraken/* $OUTPUT_DIR/\n",
    "\n",
    "rm -r $SCRATCH/*\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Gene prediction (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "If we are sure out assembly is correct, clean and we want to proceed to further analysis we can try gene prediction on the scaffolded assembly. We use [PROKKA](https://github.com/tseemann/prokka) but there are other options as well such as [RAST](http://rast.nmpdr.org/), [Augustus](http://bioinf.uni-greifswald.de/webaugustus/index.gsp), [funannotate](https://github.com/nextgenusfs/funannotate)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) [23:34:09] This is prokka 1.13.7\n",
      "[23:34:09] Written by Torsten Seemann <torsten.seemann@gmail.com>\n",
      "[23:34:09] Homepage is https://github.com/tseemann/prokka\n",
      "[23:34:09] Local time is Wed Jun 26 23:34:09 2019\n",
      "[23:34:09] You are 325073\n",
      "[23:34:09] Operating system is linux\n",
      "[23:34:09] You have BioPerl 1.007002\n",
      "[23:34:09] System has 70 cores.\n",
      "[23:34:09] Will use maximum of 12 cores.\n",
      "[23:34:09] Annotating as >>> Bacteria <<<\n",
      "[23:34:09] Generating locus_tag from 'CW56_S1_.SS14_.scaffolds.fasta' contents.\n",
      "[23:34:09] Setting --locustag FJBIJBEE from MD5 f3b23bee74170409c45041e87785ebc7\n",
      "[23:34:09] Creating new output folder: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction\n",
      "[23:34:09] Running: mkdir -p \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\n",
      "[23:34:09] Using filename prefix: CW56_S1_.SS14_.scaffolds.XXX\n",
      "[23:34:09] Setting HMMER_NCPU=1\n",
      "[23:34:09] Writing log to: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.log\n",
      "[23:34:09] Command: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/prokka --cpus 12 --kingdom Bacteria --outdir /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction --prefix CW56_S1_.SS14_.scaffolds CW56_S1_.SS14_.scaffolds.fasta\n",
      "[23:34:09] Appending to PATH: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin\n",
      "[23:34:09] Looking for 'aragorn' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/aragorn\n",
      "[23:34:09] Determined aragorn version is 1.2\n",
      "[23:34:09] Looking for 'barrnap' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/barrnap\n",
      "[23:34:09] Determined barrnap version is 0.8\n",
      "[23:34:09] Looking for 'blastp' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/blastp\n",
      "[23:34:09] Determined blastp version is 2.9\n",
      "[23:34:09] Looking for 'cmpress' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/cmpress\n",
      "[23:34:09] Determined cmpress version is 1.1\n",
      "[23:34:09] Looking for 'cmscan' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/cmscan\n",
      "[23:34:09] Determined cmscan version is 1.1\n",
      "[23:34:09] Looking for 'egrep' - found /bin/egrep\n",
      "[23:34:09] Looking for 'find' - found /usr/bin/find\n",
      "[23:34:09] Looking for 'grep' - found /bin/grep\n",
      "[23:34:09] Looking for 'hmmpress' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/hmmpress\n",
      "[23:34:09] Determined hmmpress version is 3.2\n",
      "[23:34:09] Looking for 'hmmscan' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/hmmscan\n",
      "[23:34:09] Determined hmmscan version is 3.2\n",
      "[23:34:09] Looking for 'java' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/java\n",
      "[23:34:09] Looking for 'less' - found /usr/bin/less\n",
      "[23:34:09] Looking for 'makeblastdb' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/makeblastdb\n",
      "[23:34:09] Determined makeblastdb version is 2.9\n",
      "[23:34:09] Looking for 'minced' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/minced-0.3.3/minced\n",
      "[23:34:09] Determined minced version is 3.3\n",
      "[23:34:09] Looking for 'parallel' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/parallel\n",
      "[23:34:09] Determined parallel version is 20190522\n",
      "[23:34:09] Looking for 'prodigal' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/prodigal\n",
      "[23:34:10] Determined prodigal version is 2.6\n",
      "[23:34:10] Looking for 'prokka-genbank_to_fasta_db' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/prokka-genbank_to_fasta_db\n",
      "[23:34:10] Looking for 'sed' - found /bin/sed\n",
      "[23:34:10] Looking for 'tbl2asn' - found /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/tbl2asn\n",
      "[23:34:10] Determined tbl2asn version is 25.6\n",
      "[23:34:10] Using genetic code table 11.\n",
      "[23:34:10] Loading and checking input file: CW56_S1_.SS14_.scaffolds.fasta\n",
      "[23:34:10] Wrote 67 contigs totalling 1153760 bp.\n",
      "[23:34:10] Predicting tRNAs and tmRNAs\n",
      "[23:34:10] Running: aragorn -l -gc11  -w \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/CW56_S1_\\.SS14_\\.scaffolds\\.fna\n",
      "[23:34:10] 1 tRNA-Arg c[22148,22222] 35 (tcg)\n",
      "[23:34:10] 2 tRNA-His c[22236,22308] 33 (gtg)\n",
      "[23:34:10] 3 tRNA-Phe [22445,22517] 34 (gaa)\n",
      "[23:34:10] 4 tRNA-Pro [54611,54683] 34 (tgg)\n",
      "[23:34:10] 5 tRNA-Arg [59000,59073] 35 (tct)\n",
      "[23:34:10] 6 tRNA-Val [67228,67299] 33 (cac)\n",
      "[23:34:10] 7 tRNA-Met c[84782,84855] 35 (cat)\n",
      "[23:34:10] 8 tRNA-Pro [92010,92083] 35 (ggg)\n",
      "[23:34:10] 9 tRNA-Lys c[166569,166643] 35 (ttt)\n",
      "[23:34:10] 10 tRNA-Leu [166786,166870] 35 (cag)\n",
      "[23:34:10] 11 tRNA-Ala [327621,327694] 34 (cgc)\n",
      "[23:34:10] 12 tRNA-Thr c[330231,330303] 34 (cgt)\n",
      "[23:34:10] 13 tRNA-Val [342320,342393] 35 (gac)\n",
      "[23:34:10] 1 tRNA-Cys [28306,28377] 33 (gca)\n",
      "[23:34:10] 2 tRNA-Asn c[45660,45734] 35 (gtt)\n",
      "[23:34:10] 3 tRNA-Gly c[86117,86188] 33 (gcc)\n",
      "[23:34:10] 4 tRNA-Leu c[86198,86280] 36 (tag)\n",
      "[23:34:10] 5 tRNA-Lys c[86339,86411] 34 (ctt)\n",
      "[23:34:10] 6 tRNA-Tyr c[103478,103560] 35 (gta)\n",
      "[23:34:10] 7 tRNA-Thr c[103570,103644] 36 (tgt)\n",
      "[23:34:10] 1 tRNA-Gln c[65653,65727] 34 (ttg)\n",
      "[23:34:10] 2 tRNA-Gly [107469,107542] 35 (tcc)\n",
      "[23:34:10] 3 tRNA-Leu [109494,109576] 35 (caa)\n",
      "[23:34:10] 1 tRNA-Ser c[38294,38381] 35 (gga)\n",
      "[23:34:10] 2 tRNA-Ser c[38438,38526] 35 (cga)\n",
      "[23:34:10] 3 tRNA-Arg c[38583,38658] 36 (gcg)\n",
      "[23:34:10] 4 tRNA-Ser c[38688,38776] 36 (gct)\n",
      "[23:34:10] 5 tRNA-Ser c[38815,38900] 35 (tga)\n",
      "[23:34:10] 1 tRNA-Val [3625,3697] 34 (tac)\n",
      "[23:34:10] 2 tRNA-Leu c[14871,14954] 36 (gag)\n",
      "[23:34:10] 3 tRNA-Met [36816,36888] 34 (cat)\n",
      "[23:34:10] 1 tRNA-Glu c[10822,10895] 35 (ctc)\n",
      "[23:34:10] 2 tRNA-Glu c[10930,11003] 35 (ttc)\n",
      "[23:34:10] 3 tRNA-Gln c[11012,11082] 33 (ctg)\n",
      "[23:34:10] 1 tRNA-Arg c[12491,12564] 35 (ccg)\n",
      "[23:34:10] 2 tRNA-Met [28279,28353] 35 (cat)\n",
      "[23:34:10] 3 tRNA-Asp [28421,28497] 36 (gtc)\n",
      "[23:34:10] 4 tRNA-Pro [28547,28621] 33 (cgg)\n",
      "[23:34:10] 5 tRNA-Arg c[28632,28704] 34 (cct)\n",
      "[23:34:10] 1 tRNA-Thr [6907,6979] 34 (ggt)\n",
      "[23:34:10] 2 tRNA-Trp [7232,7306] 35 (cca)\n",
      "[23:34:10] 3 tmRNA [41551,41902] 85,123 ANSDSFDYALAA*\n",
      "[23:34:10] 1 tRNA-Leu c[28222,28307] 35 (taa)\n",
      "[23:34:10] 2 tRNA-Ala c[32353,32429] 36 (ggc)\n",
      "[23:34:10] 1 tRNA-Leu [3,92] 35 (taa)\n",
      "[23:34:10] 2 tRNA-Met [120,196] 35 (cat)\n",
      "[23:34:10] 3 tRNA-Trp [206,281] 34 (cca)\n",
      "[23:34:10] 4 tRNA-Ile [298,374] 35 (gat)\n",
      "[23:34:10] 5 tRNA-Phe [380,455] 34 (gaa)\n",
      "[23:34:10] 6 tRNA-Met [463,537] 35 (cat)\n",
      "[23:34:10] 7 tRNA-Met [545,621] 35 (cat)\n",
      "[23:34:10] 1 tRNA-Gln c[-4,68] 33 (ctg)\n",
      "[23:34:10] 2 tRNA-Pyl c[76,152] 36 (cta)\n",
      "[23:34:10] 3 tRNA-Gly c[172,246] 33 (gcc)\n",
      "[23:34:10] 4 tRNA-Leu c[269,352] 35 (tag)\n",
      "[23:34:10] 5 tRNA-Tyr c[371,455] 35 (gta)\n",
      "[23:34:10] 6 tRNA-Thr c[457,531] 33 (tgt)\n",
      "[23:34:10] 1 tRNA-Cys c[103,177] 33 (gca)\n",
      "[23:34:10] 2 tRNA-Gly c[185,259] 33 (gcc)\n",
      "[23:34:10] 3 tRNA-Phe c[267,342] 34 (gaa)\n",
      "[23:34:10] 4 tRNA-Asp c[347,422] 34 (gtc)\n",
      "[23:34:10] 1 tRNA-Met c[79,154] 34 (cat)\n",
      "[23:34:10] 2 tRNA-Tyr c[164,248] 35 (gta)\n",
      "[23:34:10] 3 tRNA-Thr c[252,327] 34 (tgt)\n",
      "[23:34:10] 1 tRNA-Ala c[132,205] 35 (tgc)\n",
      "[23:34:10] 1 tRNA-Ile c[120,194] 35 (gat)\n",
      "[23:34:10] 1 tRNA-Glu c[-3,69] 34 (ttc)\n",
      "[23:34:10] 2 tRNA-Met c[93,168] 34 (cat)\n",
      "[23:34:10] Found 68 tRNAs\n",
      "[23:34:10] Predicting Ribosomal RNAs\n",
      "[23:34:10] Running Barrnap with 12 threads\n",
      "[23:34:12] 1 NODE_13_length_4751_cov_100.751730 209 16S ribosomal RNA (partial)\n",
      "[23:34:12] 2 NODE_13_length_4751_cov_100.751730 1587 23S ribosomal RNA\n",
      "[23:34:12] 3 NODE_13_length_4751_cov_100.751730 4590 5S ribosomal RNA\n",
      "[23:34:12] 4 NODE_15_length_1767_cov_52.501829 131 16S ribosomal RNA\n",
      "[23:34:12] 5 NODE_16_length_1757_cov_17.477914 117 16S ribosomal RNA\n",
      "[23:34:12] 6 NODE_17_length_1636_cov_9.475812 103 16S ribosomal RNA\n",
      "[23:34:12] Found 6 rRNAs\n",
      "[23:34:12] Skipping ncRNA search, enable with --rfam if desired.\n",
      "[23:34:12] Total of 73 tRNA + rRNA features\n",
      "[23:34:12] Searching for CRISPR repeats\n",
      "[23:34:13] Found 0 CRISPRs\n",
      "[23:34:13] Predicting coding sequences\n",
      "[23:34:13] Contigs total 1153760 bp, so using single mode\n",
      "[23:34:13] Running: prodigal -i \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/CW56_S1_\\.SS14_\\.scaffolds\\.fna -c -m -g 11 -p single -f sco -q\n",
      "[23:34:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_1_length_375077_cov_67.153941:22088..22294 on - strand\n",
      "[23:34:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_1_length_375077_cov_67.153941:59035..60222 on - strand\n",
      "[23:34:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_1_length_375077_cov_67.153941:165361..166593 on - strand\n",
      "[23:34:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_2_length_165590_cov_68.322429:86049..86486 on - strand\n",
      "[23:34:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_3_length_126709_cov_62.629402:65023..65760 on - strand\n",
      "[23:34:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_3_length_126709_cov_62.629402:107411..107536 on - strand\n",
      "[23:34:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_3_length_126709_cov_62.629402:109558..111099 on - strand\n",
      "[23:34:15] Excluding CDS which overlaps existing RNA (tRNA) at NODE_8_length_43722_cov_59.645257:28250..28405 on + strand\n",
      "[23:34:15] Excluding CDS which overlaps existing RNA (tRNA) at NODE_9_length_43473_cov_64.821852:6862..6996 on + strand\n",
      "[23:34:15] Excluding CDS which overlaps existing RNA (tRNA) at NODE_11_length_36439_cov_81.402605:30715..32391 on - strand\n",
      "[23:34:15] Excluding CDS which overlaps existing RNA (rRNA) at NODE_16_length_1757_cov_17.477914:232..471 on - strand\n",
      "[23:34:15] Excluding CDS which overlaps existing RNA (rRNA) at NODE_17_length_1636_cov_9.475812:137..457 on - strand\n",
      "[23:34:15] Excluding CDS which overlaps existing RNA (tRNA) at NODE_32_length_563_cov_6.174312:24..260 on + strand\n",
      "[23:34:15] Excluding CDS which overlaps existing RNA (tRNA) at NODE_37_length_435_cov_6.889610:69..248 on + strand\n",
      "[23:34:15] Found 979 CDS\n",
      "[23:34:15] Connecting features back to sequences\n",
      "[23:34:15] Not using genus-specific database. Try --usegenus to enable it.\n",
      "[23:34:15] Annotating CDS, please be patient.\n",
      "[23:34:15] Will use 12 CPUs for similarity searching.\n",
      "[23:34:15] There are still 979 unannotated CDS left (started with 979)\n",
      "[23:34:15] Will use blast to search against /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/kingdom/Bacteria/IS with 12 CPUs\n",
      "[23:34:15] Running: cat \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/IS\\.faa | parallel --gnu --plain -j 12 --block 14648 --recstart '>' --pipe blastp -query - -db /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/kingdom/Bacteria/IS -evalue 1e-30 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/IS\\.blast 2> /dev/null\n",
      "[23:34:19] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/IS.faa\n",
      "[23:34:19] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/IS.blast\n",
      "[23:34:19] There are still 979 unannotated CDS left (started with 979)\n",
      "[23:34:19] Will use blast to search against /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/kingdom/Bacteria/AMR with 12 CPUs\n",
      "[23:34:19] Running: cat \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/AMR\\.faa | parallel --gnu --plain -j 12 --block 14648 --recstart '>' --pipe blastp -query - -db /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/kingdom/Bacteria/AMR -evalue 9.99999999999999e-301 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/AMR\\.blast 2> /dev/null\n",
      "[23:34:21] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/AMR.faa\n",
      "[23:34:21] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/AMR.blast\n",
      "[23:34:22] There are still 979 unannotated CDS left (started with 979)\n",
      "[23:34:22] Will use blast to search against /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/kingdom/Bacteria/sprot with 12 CPUs\n",
      "[23:34:22] Running: cat \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/sprot\\.faa | parallel --gnu --plain -j 12 --block 14648 --recstart '>' --pipe blastp -query - -db /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/kingdom/Bacteria/sprot -evalue 1e-09 -qcov_hsp_perc 80 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/sprot\\.blast 2> /dev/null\n",
      "[23:34:30] Modify product: UPF0102 protein RPA0323 => hypothetical protein\n",
      "[23:34:30] Modify product: Probable NADH oxidase => putative NADH oxidase\n",
      "[23:34:30] Modify product: Probable peptidyl-prolyl cis-trans isomerase => putative peptidyl-prolyl cis-trans isomerase\n",
      "[23:34:30] Modify product: Uncharacterized protein TP_0956 => putative protein\n",
      "[23:34:30] Modify product: Uncharacterized protein TP_0751 => putative protein\n",
      "[23:34:30] Modify product: Uncharacterized ABC transporter ATP-binding protein YknY => putative ABC transporter ATP-binding protein YknY\n",
      "[23:34:30] Modify product: Putative DD-carboxypeptidase TP_0574 => Putative DD-carboxypeptidase\n",
      "[23:34:30] Modify product: Probable tRNA sulfurtransferase => putative tRNA sulfurtransferase\n",
      "[23:34:31] Modify product: Probable transcriptional regulatory protein aq_1575 => putative transcriptional regulatory protein\n",
      "[23:34:31] Modify product: Uncharacterized ABC transporter ATP-binding protein YbiT => putative ABC transporter ATP-binding protein YbiT\n",
      "[23:34:31] Modify product: Bifunctional protein FolD => Bifunctional protein FolD protein\n",
      "[23:34:31] Modify product: Dihydroorotate dehydrogenase B (NAD(+)), electron transfer subunit homolog => Dihydroorotate dehydrogenase B (NAD(+)), electron transfer subunit \n",
      "[23:34:31] Modify product: Outer membrane protein TP0453 => Outer membrane protein\n",
      "[23:34:31] Modify product: Probable ATP synthase YscN => putative ATP synthase YscN\n",
      "[23:34:31] Modify product: Probable peptidoglycan glycosyltransferase FtsW => putative peptidoglycan glycosyltransferase FtsW\n",
      "[23:34:31] Modify product: UPF0053 protein HI_0107 => hypothetical protein\n",
      "[23:34:31] Modify product: UPF0053 protein HI_0107 => hypothetical protein\n",
      "[23:34:31] Modify product: Cyclic di-GMP binding protein TDE_0214 => Cyclic di-GMP binding protein\n",
      "[23:34:31] Modify product: Probable dual-specificity RNA methyltransferase RlmN => putative dual-specificity RNA methyltransferase RlmN\n",
      "[23:34:31] Modify product: UPF0051 protein SA0778 => hypothetical protein\n",
      "[23:34:31] Modify product: Probable ABC transporter ATP-binding protein M6_Spy0273 => putative ABC transporter ATP-binding protein\n",
      "[23:34:31] Modify product: Putative zinc metalloprotease aq_1964 => Putative zinc metalloprotease\n",
      "[23:34:31] Modify product: 1,4-alpha-glucan branching enzyme TTHA1902 => 1,4-alpha-glucan branching enzyme\n",
      "[23:34:31] Modify product: Uncharacterized protein TP_0183 => putative protein\n",
      "[23:34:31] Modify product: Oxygen-independent coproporphyrinogen-III oxidase-like protein LL1139 => Oxygen-independent coproporphyrinogen-III oxidase-like protein\n",
      "[23:34:31] Modify product: Putative outer membrane protein assembly factor TP_0326 => Putative outer membrane protein assembly factor\n",
      "[23:34:31] Modify product: UPF0073 inner membrane protein YqfA => hypothetical protein\n",
      "[23:34:31] Modify product: Uncharacterized protein TP_0004 => putative protein\n",
      "[23:34:31] Modify product: GTP cyclohydrolase 1 type 2 homolog => GTP cyclohydrolase 1 type 2 \n",
      "[23:34:31] Modify product: Uncharacterized metal-dependent hydrolase YcfH => putative metal-dependent hydrolase YcfH\n",
      "[23:34:31] Modify product: Probable tRNA-dihydrouridine synthase => putative tRNA-dihydrouridine synthase\n",
      "[23:34:31] Modify product: UPF0056 inner membrane protein YhgN => hypothetical protein\n",
      "[23:34:31] Cleaned 32 /product names\n",
      "[23:34:31] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/sprot.faa\n",
      "[23:34:31] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/sprot.blast\n",
      "[23:34:31] There are still 554 unannotated CDS left (started with 979)\n",
      "[23:34:31] Will use hmmer3 to search against /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/hmm/HAMAP.hmm with 12 CPUs\n",
      "[23:34:31] Running: cat \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/HAMAP\\.hmm\\.faa | parallel --gnu --plain -j 12 --block 8083 --recstart '>' --pipe hmmscan --noali --notextw --acc -E 1e-09 --cpu 1 /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/db/hmm/HAMAP.hmm /dev/stdin > \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/HAMAP\\.hmm\\.hmmer3 2> /dev/null\n",
      "[23:34:38] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/HAMAP.hmm.faa\n",
      "[23:34:38] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/HAMAP.hmm.hmmer3\n",
      "[23:34:38] Labelling remaining 450 proteins as 'hypothetical protein'\n",
      "[23:34:38] Possible /pseudo 'Flagellin FlaB1' at NODE_1_length_375077_cov_67.153941 position 272233\n",
      "[23:34:38] Possible /pseudo 'Protein YhgF' at NODE_1_length_375077_cov_67.153941 position 329422\n",
      "[23:34:38] Possible /pseudo 'Flagellar basal-body rod protein FlgG' at NODE_1_length_375077_cov_67.153941 position 367173\n",
      "[23:34:38] Possible /pseudo 'Macrolide export ATP-binding/permease protein MacB' at NODE_1_length_375077_cov_67.153941 position 369417\n",
      "[23:34:38] Possible /pseudo 'L-cystine-binding protein FliY' at NODE_7_length_44530_cov_66.819134 position 2420\n",
      "[23:34:38] Found 452 unique /gene codes.\n",
      "[23:34:38] Fixed 2 duplicate /gene - ylxH_1 ylxH_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - rbn_1 rbn_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - mglB_1 mglB_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - arnB_1 arnB_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - deaD_1 deaD_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - btuD_1 btuD_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - lnt_1 lnt_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - fliG_1 fliG_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - flhB_1 flhB_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - pcrA_1 pcrA_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - lepA_1 lepA_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - rlmB_1 rlmB_2\n",
      "[23:34:38] Fixed 3 duplicate /gene - dnaJ_1 dnaJ_2 dnaJ_3\n",
      "[23:34:38] Fixed 2 duplicate /gene - ftsH_1 ftsH_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - gph_1 gph_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - ktrA_1 ktrA_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - ftsW_1 ftsW_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - atpE_1 atpE_2\n",
      "[23:34:38] Fixed 3 duplicate /gene - smc_1 smc_2 smc_3\n",
      "[23:34:38] Fixed 2 duplicate /gene - macB_1 macB_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - potA_1 potA_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - yhgF_1 yhgF_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - atpD_1 atpD_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - flaB1_1 flaB1_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - fliY_1 fliY_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - ftsY_1 ftsY_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - clpP_1 clpP_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - pgk/tpi_1 pgk/tpi_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - flgG_1 flgG_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - nth_1 nth_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - dbpA_1 dbpA_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - ntpB_1 ntpB_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - ruvB_1 ruvB_2\n",
      "[23:34:38] Fixed 3 duplicate /gene - lapB_1 lapB_2 lapB_3\n",
      "[23:34:38] Fixed 2 duplicate /gene - fusA_1 fusA_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - cmk_1 cmk_2\n",
      "[23:34:38] Fixed 2 duplicate /gene - alsT_1 alsT_2\n",
      "[23:34:38] Fixed 5 duplicate /gene - bepA_1 bepA_2 bepA_3 bepA_4 bepA_5\n",
      "[23:34:38] Fixed 38 colliding /gene names.\n",
      "[23:34:38] Adding /locus_tag identifiers\n",
      "[23:34:38] Assigned 1053 locus_tags to CDS and RNA features.\n",
      "[23:34:38] Writing outputs to /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/\n",
      "[23:34:39] Generating annotation statistics file\n",
      "[23:34:39] Generating Genbank and Sequin files\n",
      "[23:34:39] Running: tbl2asn -V b -a r10k -l paired-ends -M n -N 1 -y 'Annotated using prokka 1.13.7 from https://github.com/tseemann/prokka' -Z \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/CW56_S1_\\.SS14_\\.scaffolds\\.err -i \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/CW56_S1_\\.SS14_\\.scaffolds\\.fsa 2> /dev/null\n",
      "[23:34:42] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/errorsummary.val\n",
      "[23:34:42] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.dr\n",
      "[23:34:42] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.fixedproducts\n",
      "[23:34:42] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.ecn\n",
      "[23:34:42] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.val\n",
      "[23:34:42] Repairing broken .GBK output that tbl2asn produces...\n",
      "[23:34:42] Running: sed 's/COORDINATES: profile/COORDINATES:profile/' < \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/CW56_S1_\\.SS14_\\.scaffolds\\.gbf > \\/mnt\\/ssd\\/ssd_2\\/bioda_temp\\/honza\\/scratch_default\\/linda\\/results\\/assembly\\/gene_prediction\\/CW56_S1_\\.SS14_\\.scaffolds\\.gbk\n",
      "[23:34:42] Deleting unwanted file: /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.gbf\n",
      "[23:34:42] Output files:\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.txt\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.tsv\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.sqn\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.gbk\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.faa\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.gff\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.ffn\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.log\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.fna\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.tbl\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.err\n",
      "[23:34:42] /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/gene_prediction/CW56_S1_.SS14_.scaffolds.fsa\n",
      "[23:34:42] Annotation finished successfully.\n",
      "[23:34:42] Walltime used: 0.55 minutes\n",
      "[23:34:42] If you use this result please cite the Prokka paper:\n",
      "[23:34:42] Seemann T (2014) Prokka: rapid prokaryotic genome annotation. Bioinformatics. 30(14):2068-9.\n",
      "[23:34:42] Type 'prokka --citation' for more details.\n",
      "[23:34:42] Share and enjoy!\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# If we have a set of HIGHLY trusted proteins we can add the in genebank (gbk) format and assign GBK variable pointing to the file\n",
    "#GBK=/path/to/species_trusted_proteins.gbk\n",
    "\n",
    "for sample in *.scaffolds.fasta\n",
    "do\n",
    "    prokka --cpus $THREADS --kingdom Bacteria --outdir $OUTPUT_DIR/assembly/gene_prediction --prefix ${sample%.fa*} $sample # --proteins $GBK\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At this point, there is bug in PROKKA-minced conda installation. If you get an error with `Error: A JNI error has occurred, please check your installation and try again\n",
    "Exception in thread \"main\" java.lang.UnsupportedClassVersionError: minced has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0...` you have to downgrade minced to version 0.3.3 (or rebuild from souce). Most likely right now it's at version 0.4.0 and Java at version 1.8.0. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) minced 0.3.3\n",
      "MinCED - Mining CRISPRs in Environmental Datasets (version 0.3.3)\n",
      "Copyright 2011      Florent Angly     <florent.angly@gmail.com>\n",
      "          2013-2018 Connor Skennerton <c.skennerton@gmail.com>\n",
      "\n",
      "Minced is a modified version of CRT (www.room220.com/crt)\n",
      "Charles Bland et al., CRISPR Recognition Tool (CRT): a tool for automatic\n",
      "detection of clustered regularly interspaced palindromic repeats\n",
      "BMC Bioinformatics 8, no. 1 (2007): 209.\n",
      "Distributed under the GNU General Public License version 3\n",
      "(treponema) (.env) (treponema) (.env) lrwxrwxrwx 1 323639 domain users 30 Jun 26 22:46 /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/bin/minced -> ../share/minced-0.4.0-0/minced\n",
      "(treponema) (.env) (treponema) (.env) openjdk version \"1.8.0_192\"\n",
      "OpenJDK Runtime Environment (Zulu 8.33.0.1-linux64) (build 1.8.0_192-b01)\n",
      "OpenJDK 64-Bit Server VM (Zulu 8.33.0.1-linux64) (build 25.192-b01, mixed mode)\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) --2019-06-26 23:33:00--  https://github.com/ctSkennerton/minced/releases/download/0.3.3/minced.jar\n",
      "Resolving github.com (github.com)... 140.82.118.3\n",
      "Connecting to github.com (github.com)|140.82.118.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/14484942/fd459180-3d00-11e9-82d8-f120121f7368?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190626%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190626T213300Z&X-Amz-Expires=300&X-Amz-Signature=128d49f4b976397f57e24be3ce156071cac9bf2c5a11ad5ac86f30a9655b2682&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dminced.jar&response-content-type=application%2Foctet-stream [following]\n",
      "--2019-06-26 23:33:00--  https://github-production-release-asset-2e65be.s3.amazonaws.com/14484942/fd459180-3d00-11e9-82d8-f120121f7368?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190626%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190626T213300Z&X-Amz-Expires=300&X-Amz-Signature=128d49f4b976397f57e24be3ce156071cac9bf2c5a11ad5ac86f30a9655b2682&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dminced.jar&response-content-type=application%2Foctet-stream\n",
      "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.160.3\n",
      "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.160.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 25046 (24K) [application/octet-stream]\n",
      "Saving to: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/minced-0.3.3/minced.jar\n",
      "\n",
      "/mnt/nfs/home/32363 100%[===================>]  24.46K  --.-KB/s    in 0.1s    \n",
      "\n",
      "2019-06-26 23:33:01 (232 KB/s) - /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/minced-0.3.3/minced.jar saved [25046/25046]\n",
      "\n",
      "(treponema) (.env) --2019-06-26 23:33:01--  https://github.com/ctSkennerton/minced/releases/download/0.3.3/minced\n",
      "Resolving github.com (github.com)... 140.82.118.3\n",
      "Connecting to github.com (github.com)|140.82.118.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/14484942/fd459180-3d00-11e9-92d3-9f9d29616d87?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190626%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190626T213301Z&X-Amz-Expires=300&X-Amz-Signature=adec70000b70845ff231bc99fd8a389688ed73aaa58859c583b127200d12af64&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dminced&response-content-type=application%2Foctet-stream [following]\n",
      "--2019-06-26 23:33:01--  https://github-production-release-asset-2e65be.s3.amazonaws.com/14484942/fd459180-3d00-11e9-92d3-9f9d29616d87?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20190626%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20190626T213301Z&X-Amz-Expires=300&X-Amz-Signature=adec70000b70845ff231bc99fd8a389688ed73aaa58859c583b127200d12af64&X-Amz-SignedHeaders=host&actor_id=0&response-content-disposition=attachment%3B%20filename%3Dminced&response-content-type=application%2Foctet-stream\n",
      "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.216.160.3\n",
      "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.216.160.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 791 [application/octet-stream]\n",
      "Saving to: /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/minced-0.3.3/minced\n",
      "\n",
      "/mnt/nfs/home/32363 100%[===================>]     791  --.-KB/s    in 0s      \n",
      "\n",
      "2019-06-26 23:33:02 (25.9 MB/s) - /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/minced-0.3.3/minced saved [791/791]\n",
      "\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/minced-0.3.3/minced\n",
      "(treponema) (.env) minced 0.3.3\n",
      "MinCED - Mining CRISPRs in Environmental Datasets (version 0.3.3)\n",
      "Copyright 2011      Florent Angly     <florent.angly@gmail.com>\n",
      "          2013-2018 Connor Skennerton <c.skennerton@gmail.com>\n",
      "\n",
      "Minced is a modified version of CRT (www.room220.com/crt)\n",
      "Charles Bland et al., CRISPR Recognition Tool (CRT): a tool for automatic\n",
      "detection of clustered regularly interspaced palindromic repeats\n",
      "BMC Bioinformatics 8, no. 1 (2007): 209.\n",
      "Distributed under the GNU General Public License version 3\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Results in error\n",
    "minced --version\n",
    "# Check minced version\n",
    "ls -l $CONDA_PREFIX/bin/minced\n",
    "# Check java version\n",
    "java -version\n",
    "# Download minced-0.3.0\n",
    "mkdir $CONDA_PREFIX/share/minced-0.3.3\n",
    "wget https://github.com/ctSkennerton/minced/releases/download/0.3.3/minced.jar -O $CONDA_PREFIX/share/minced-0.3.3/minced.jar\n",
    "wget https://github.com/ctSkennerton/minced/releases/download/0.3.3/minced -O $CONDA_PREFIX/share/minced-0.3.3/minced\n",
    "chmod u+x $CONDA_PREFIX/share/minced-0.3.3/minced\n",
    "export PATH=$CONDA_PREFIX/share/minced-0.3.3:$PATH\n",
    "# Should work fine\n",
    "which minced\n",
    "minced --version"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And now you can rerun the gene prediction (PROKKA)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Whole genome alignment (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "Another thing we can do with the assembly is to look at the whole-genome similarities and try to fix the orientation of the resulting scaffolds. This can be done using whole-genome alignment of the scaffolds and the related reference."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment: File exists\n",
      "(treponema) (.env) (treponema) (.env) mkdir: cannot create directory /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds: File exists\n",
      "1: PREPARING DATA\n",
      "2,3: RUNNING mummer AND CREATING CLUSTERS\n",
      "# reading input file \"/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.ntref\" of length 1139570\n",
      "# construct suffix tree for sequence of length 1139570\n",
      "# (maximum reference length is 536870908)\n",
      "# (maximum query length is 4294967295)\n",
      "# process 11395 characters per dot\n",
      "#....................................................................................................\n",
      "# CONSTRUCTIONTIME /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/opt/mummer-3.23/mummer /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.ntref 0.30\n",
      "# reading input file \"/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/scaffolds/CW56_S1_.SS14_.scaffolds.fasta\" of length 1153826\n",
      "# matching query-file \"/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/spades/scaffolds/CW56_S1_.SS14_.scaffolds.fasta\"\n",
      "# against subject-file \"/mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.ntref\"\n",
      "# COMPLETETIME /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/opt/mummer-3.23/mummer /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.ntref 0.90\n",
      "# SPACE /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/opt/mummer-3.23/mummer /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.ntref 2.23\n",
      "4: FINISHING DATA\n",
      "Filtering alignments\n",
      "Extracting alignment coordinates\n",
      "Analyzing SNPs\n",
      "Extracting alignment breakpoints\n",
      "Generating report file\n",
      "gnuplot 5.2 patchlevel 7\n",
      "Writing filtered delta file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.filter\n",
      "Reading delta file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.filter\n",
      "Writing plot files /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.fplot, /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.rplot, /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.hplot\n",
      "Determining SNPs from sequence and alignment data\n",
      "Writing gnuplot script /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.gp\n",
      "Rendering plot /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.ps\n",
      "gnuplot 5.2 patchlevel 7\n",
      "Writing filtered delta file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.filter\n",
      "Reading delta file /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.filter\n",
      "Writing plot files /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.fplot, /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.rplot, /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.hplot\n",
      "Determining SNPs from sequence and alignment data\n",
      "Writing gnuplot script /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.gp\n",
      "Rendering plot /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/nucmer.png\n",
      "mkdir: cannot create directory /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/results/assembly/genome_alignment/CW56_S1_.SS14_.scaffolds/individual_alignments: File exists\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_1_length_375077_cov_67.153941\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_2_length_165590_cov_68.322429\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_3_length_126709_cov_62.629402\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_4_length_86646_cov_81.669911\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_5_length_76950_cov_69.043008\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_6_length_63385_cov_77.335230\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_7_length_44530_cov_66.819134\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_8_length_43722_cov_59.645257\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_9_length_43473_cov_64.821852\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_10_length_37477_cov_70.532262\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_11_length_36439_cov_81.402605\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_12_length_18555_cov_63.723681\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_13_length_4751_cov_100.751730\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_14_length_2653_cov_82.605305\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_19_length_1165_cov_58.958574\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_21_length_821_cov_49.275216\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_45_length_336_cov_51.770335\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_46_length_328_cov_48.885572\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_47_length_323_cov_94.698980\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_50_length_318_cov_52.675393\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_63_length_248_cov_34.669421\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_64_length_187_cov_25.966667\n",
      "gi_511533127_gb_CP004011_1_Treponema_pallidum_subsp_pallidum_SS14_complete_genome NODE_66_length_132_cov_100.400000\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "mkdir $OUTPUT_DIR/assembly/genome_alignment\n",
    "\n",
    "for sample in *.scaffolds.fasta\n",
    "do\n",
    "    sample_name=$(basename $sample .fasta)\n",
    "\n",
    "    mkdir $OUTPUT_DIR/assembly/genome_alignment/$sample_name\n",
    "\n",
    "    nucmer -p $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer $REFERENCE $sample\n",
    "\n",
    "    dnadiff -p $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer -d $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.delta\n",
    "\n",
    "    mummerplot -p $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer -s large --SNP --postscript --filter $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.delta\n",
    "    mummerplot -p $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer -s large --SNP --png --filter $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.delta\n",
    "\n",
    "    mkdir $OUTPUT_DIR/assembly/genome_alignment/$sample_name/individual_alignments\n",
    "\n",
    "    grep \">\" $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.delta | cut -d ' ' -f1,2 | sed 's/>//g' | while read list\n",
    "    do\n",
    "        echo $list\n",
    "        outname=`echo $list | sed 's/ /_/g'`\n",
    "        \n",
    "        show-aligns $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.delta $list > $OUTPUT_DIR/assembly/genome_alignment/$sample_name/individual_alignments/$outname.aln\n",
    "    done\n",
    "\n",
    "    # Extract unaligned scaffolds to the reference\n",
    "    cat $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.unqry | cut -f1 > $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.unqry.names\n",
    "    seqtk subseq $sample $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.unqry.names > $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.unqry.fasta\n",
    "    rm $OUTPUT_DIR/assembly/genome_alignment/$sample_name/nucmer.unqry.names\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Outputs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Variant annotation (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having variants in VCF is nice but often you might be interested what is their effect or purpose. Unless you imidiately know the important positions variant annotation might bring you some additional information. [SnpEff](http://snpeff.sourceforge.net/) allows you to build your own annotation database and apply it to your VCF. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we build the annotation database. This has to be done only once per genome. Note: there are some pre-loaded annotations in SnpEff but since we might be using different genome version it might be safer to generate our own database. You can generate the database with GFF and FASTA (hint [here](http://lab.loman.net/2012/11/16/how-to-get-snpeff-working-with-bacterial-genomes-from-ncbi/)) or you can do it from GenBank full format. The advantage of going for GenBank full is that both gene annotation and genome sequence is included in one file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2019-06-27 00:42:48--  ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Treponema_pallidum/all_assembly_versions/GCA_000410555.1_ASM41055v1/GCA_000410555.1_ASM41055v1_genomic.gbff.gz\n",
      "           => /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/GCA_000410555.1_ASM41055v1_genomic.gbff.gz\n",
      "Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.11, 2607:f220:41e:250::12\n",
      "Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.11|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /genomes/genbank/bacteria/Treponema_pallidum/all_assembly_versions/GCA_000410555.1_ASM41055v1 ... done.\n",
      "==> SIZE GCA_000410555.1_ASM41055v1_genomic.gbff.gz ... 748077\n",
      "==> PASV ... done.    ==> RETR GCA_000410555.1_ASM41055v1_genomic.gbff.gz ... done.\n",
      "Length: 748077 (731K) (unauthoritative)\n",
      "\n",
      "GCA_000410555.1_ASM 100%[===================>] 730.54K  1.36MB/s    in 0.5s    \n",
      "\n",
      "2019-06-27 00:42:50 (1.36 MB/s) - /mnt/ssd/ssd_2/bioda_temp/honza/scratch_default/linda/reference/GCA_000410555.1_ASM41055v1_genomic.gbff.gz saved [748077]\n",
      "\n",
      "(treponema) (.env) (treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "wget ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Treponema_pallidum/all_assembly_versions/GCA_000410555.1_ASM41055v1/GCA_000410555.1_ASM41055v1_genomic.gbff.gz -P $(dirname $REFERENCE)\n",
    "gunzip $(dirname $REFERENCE)/GCA_000410555.1_ASM41055v1_genomic.gbff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/snpeff-4.3.1t-2/\n",
      "(treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) (treponema) (.env) java.lang.RuntimeException: Property: 'SS14_CP004011.1.genome' not found\n",
      "\tat org.snpeff.interval.Genome.<init>(Genome.java:106)\n",
      "\tat org.snpeff.snpEffect.Config.readGenomeConfig(Config.java:681)\n",
      "\tat org.snpeff.snpEffect.Config.readConfig(Config.java:649)\n",
      "\tat org.snpeff.snpEffect.Config.init(Config.java:480)\n",
      "\tat org.snpeff.snpEffect.Config.<init>(Config.java:117)\n",
      "\tat org.snpeff.SnpEff.loadConfig(SnpEff.java:451)\n",
      "\tat org.snpeff.snpEffect.commandLine.SnpEffCmdDump.run(SnpEffCmdDump.java:183)\n",
      "\tat org.snpeff.SnpEff.run(SnpEff.java:1183)\n",
      "\tat org.snpeff.SnpEff.main(SnpEff.java:162)\n",
      "(treponema) (.env) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# Get the GenkBank full annotation - reference genome version and the gene annotation must be identical!\n",
    "#wget ftp://ftp.ncbi.nlm.nih.gov/genomes/genbank/bacteria/Treponema_pallidum/all_assembly_versions/GCA_000410555.1_ASM41055v1/GCA_000410555.1_ASM41055v1_genomic.gbff.gz -P $(dirname $REFERENCE)\n",
    "#gunzip $(dirname $REFERENCE)/GCA_000410555.1_ASM41055v1_genomic.gbff.gz\n",
    "cd $(dirname $REFERENCE)\n",
    "\n",
    "touch ${REFERENCE%.fa*}.snpEffect.config # Addition to the main snpEff.config file - be careful with this\n",
    "\n",
    "db_name=\"SS14_CP004011.1\" # How you want to call the database?\n",
    "fasta_header=`head -1 $REFERENCE | sed 's/>//g' | cut -d ' ' -f 1` # Get only the fasta header without the initial \">\" and split after first empty space\n",
    "\n",
    "# Create custom snpEff.config\n",
    "echo \"$db_name.genome : Treponema pallidum subsp. pallidum SS14\" > ${REFERENCE%.fa*}.snpEffect.config # Database name and description\n",
    "echo -e \"\\\\t$db_name.chromosomes : $fasta_header\" >> ${REFERENCE%.fa*}.snpEffect.config # Database name and fasta header\n",
    "echo -e \"\\\\t$db_name.$fasta_header.codonTable : Bacterial_and_Plant_Plastid\" >> ${REFERENCE%.fa*}.snpEffect.config # Database name, fasta header and codon table\n",
    "\n",
    "SNPEFF_DIR=$(ls -d /mnt/nfs/home/323639/000000-My_Documents/VM-home/tools/miniconda2/envs/treponema/share/snpeff-*/)\n",
    "echo $SNPEFF_DIR\n",
    "ANNOT_DIR=$SNPEFF_DIR/data/$db_name # Data directory to store the added genome annotation\n",
    "\n",
    "# Make the annotation dabase - only once per genome!\n",
    "if [ ! -d \"$ANNOT_DIR\" ]; then  # Check if the genome folder already exists\n",
    "    mkdir -p $ANNOT_DIR\n",
    "    cd $ANNOT_DIR/\n",
    "\n",
    "    cat $(dirname $REFERENCE)/GCA_000410555.1_ASM41055v1_genomic.gbff > $ANNOT_DIR/genes.gbk\n",
    "    cat ${REFERENCE%.fa*}.snpEffect.config >> $SNPEFF_DIR/snpEff.config\n",
    "\n",
    "    cd $SNPEFF_DIR/\n",
    "    snpEff build -genbank -v $db_name\n",
    "fi\n",
    "\n",
    "snpEff dump $db_name # Check if everything is OK "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "################################################################################\n",
    "### Running snpEff to evaluate the impact of SNPs\n",
    "INPUT_DIR=/home/jan/Data/projects/linda/2018_mar/results/annotation/vcf\n",
    "#INPUT_VCF=/home/jan/Data/projects/linda/2018_mar/results/annotation/vcf/test.vcf # Set the input VCF file\n",
    "cd $SNPEFF_DIR/\n",
    "\n",
    "for INPUT_VCF in $INPUT_DIR/*.vcf\n",
    "do\n",
    "\techo $INPUT_VCF\n",
    "\t$SNPEFF_RUN $(basename $REF_DIR) $INPUT_VCF > ${INPUT_VCF%.vcf}.$(basename $REF_DIR).annot.vcf # Run SnpEff\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "### Additional scaffolding (optional)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "SPAdes is quite efficient in the scaffolding and we never had a need to do additional scaffolding but you might find yourself in a situation to do so. We are using [BESST](https://github.com/ksahlin/BESST) but there are other options such as [SSPACE-basic] (https://github.com/nsoranzo/sspace_basic), [SSPACE-standard](https://www.baseclear.com/services/bioinformatics/basetools/sspace-standard/) and others. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the best functionality, you have know the insert size. You can get this information from the Qualimap report which we created above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "\n",
    "INPUT_DIR=/storage/brno2/home/opplatek/bioinf_projects/biocore/linda_treponema/2017_oct/results/ss14/alignment/mappedOnly/fastq # Input reads\n",
    "INPUT_SCAFFOLDS_DIR=/storage/brno2/home/opplatek/bioinf_projects/biocore/linda_treponema/2017_oct/results/assembly/mappedOnly/spades/scaffolds # Input scaffolds\n",
    "OUTPUT_DIR=/storage/brno2/home/opplatek/bioinf_projects/biocore/linda_treponema/2017_oct/results/assembly/mappedOnly/spades/besst\n",
    "\n",
    "# If we wan't to add this to Quast evaluation\n",
    "REF_SEQ=SS14.fa # SS14.fa; nichols.fa; bosnia_a.fa\n",
    "REF_DIR=/storage/brno2/home/opplatek/genomes/treponema\n",
    "\n",
    "THREADS=$PBS_NUM_PPN\n",
    "\n",
    "module add bwa-0.7.15\n",
    "BWA=bwa\n",
    "module add samtools-1.4\n",
    "SAMTOOLS=samtools\n",
    "BESST_BIN=/storage/brno2/home/opplatek/tools/BESST-2.2.4/bin\n",
    "QUAST=/storage/brno2/home/opplatek/tools/quast-4.5/quast.py\n",
    "#SSPACE=/storage/brno2/home/opplatek/tools/SSPACE/SSPACE-STANDARD-3.0_linux-x86_64/SSPACE_Standard_v3.0.pl # TODO\n",
    "#GAP_FILLER=/storage/brno2/home/opplatek/tools/SSPACE/GapFiller_v1-10_linux-x86_64/GapFiller.pl # Distribution along SSPACE; TODO\n",
    "\n",
    "####################################################################################################\n",
    "### Copy inputs\n",
    "cp $INPUT_SCAFFOLDS_DIR/*.scaffolds.fasta $SCRATCH/\n",
    "cp $REF_DIR/$REF_SEQ $SCRATCH/\n",
    "cp $INPUT_DIR/*.fastq.gz $SCRATCH/\n",
    "\n",
    "\n",
    "####################################################################################################\n",
    "### BESST scaffolding\n",
    "mkdir -p $SCRATCH/BESST\n",
    "\n",
    "for sample in *R1*.fastq.gz\n",
    "do\n",
    "\tFORWARD=$sample\n",
    "\textension=${FORWARD##*R1}\n",
    "\tREVERSE=${FORWARD%R1*}R2${extension}\n",
    "\n",
    "\tCONTIG=${FORWARD%_R1*}.scaffolds.fasta # SPAdes assembly contigs/scaffolds\n",
    "\n",
    "\t$BWA index $CONTIG\n",
    "\n",
    "\t$BWA mem -t $THREADS -w 0 -O 99 $CONTIG $FORWARD $REVERSE | $SAMTOOLS view -F 4 -@ $THREADS -b - | $SAMTOOLS fixmate -O bam - - | $SAMTOOLS sort -@ $THREADS - > $SCRATCH/BESST/${FORWARD%_R1*}${extension%.fastq.gz*}.bam # Make alignment for BESST; default is BWA-MEM and recommended settings are used; they have script to do it (reads_to_ctg_map.py) but it has some error - setting taken from there\n",
    "\n",
    "\t$SAMTOOLS index -@ $THREADS $SCRATCH/BESST/${FORWARD%_R1*}${extension%.fastq.gz*}.bam\n",
    "\n",
    "\tmkdir $SCRATCH/BESST/${FORWARD%%.*}\n",
    "\n",
    "\t$BESST_BIN/runBESST -c $CONTIG -f $SCRATCH/BESST/${FORWARD%_R1*}${extension%.fastq.gz*}.bam -filter_contigs 100 --orientation fr -o $SCRATCH/BESST/${FORWARD%%.*}\n",
    "done\n",
    "\n",
    "deactivate  # Deactivate virtual environment\n",
    "rm -rf $SCRATCH/python_env_BESST # Remove virtual environment\n",
    "\n",
    "####################################################################################################\n",
    "### Write and sort final scaffolds\n",
    "cd $SCRATCH/BESST\n",
    "\n",
    "mkdir -p $SCRATCH/BESST_scaffolds\n",
    "\n",
    "for i in $(ls -d */)\n",
    "do\n",
    "\tcat $i/BESST_output/pass1/Scaffolds_pass1.fa > $SCRATCH/BESST_scaffolds/${i%/*}.scaffolds.fasta\n",
    "done\n",
    "\n",
    "# Sort by length https://www.biostars.org/p/153999/\n",
    "cd $SCRATCH/BESST_scaffolds/\n",
    "\n",
    "for i in *.fasta\n",
    "do\n",
    "\tcat $i | awk '/^>/ {printf(\"%s%s\\t\",(N>0?\"\\n\":\"\"),$0);N++;next;} {printf(\"%s\",$0);} END {printf(\"\\n\");}' | awk -F '\\t' '{printf(\"%d\\t%s\\n\",length($2),$0);}' | sort -k1,1nr | cut -f 2- | tr \"\\t\" \"\\n\" > $i.tmp # Longest first; shortest first change sort 'sort -k1,1nr' to 'sort -k1,1n'\n",
    "\tmv $i.tmp $i\n",
    "done\n",
    "\n",
    "$QUAST --threads $THREADS -R $SCRATCH/$REF_SEQ $(ls *.scaffolds.fasta) # -s -> scaffolds\n",
    "\n",
    "############################################################################################\n",
    "### Copy results\n",
    "mkdir -p $OUTPUT_DIR\n",
    "\n",
    "cp -r $SCRATCH/BESST_scaffolds/* $OUTPUT_DIR/\n",
    "\n",
    "rm -rf $SCRATCH/*\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
