# Grillova et al. 2019 - DpnI enrichment 
This repository contains all the steps necessary to reproduce the analyses done in Grillova et al. 2019 - DpnI enrichment publication. 

This workflow can be used for analysis of **any** bacterial genome and is specially designed for analysis of highly contaminated deep sequencing samples. It can work with contamination by the host sample as well as any other organism (both known and unknown). It does not depend on predefined database of the contaminants and avoids missing unknown organisms contamnation. It is highly sentistive and specific at the same time and combines alignment-base assembly and *de novo* assembly.

The workflow is available as Jupyter notebook with conda recipes (?) to allow easy installation and reproducibility.

# The wofkflow schema
TODO - add schema of the workflow

# The wofkflow schema

# Setting up the environment

# Running the workflow - example

# Outputs
